// @datadog/datadog-ci-base/dist/commands/junit/upload.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JunitUploadCommand = void 0;
const clipanion_1 = require("clipanion");
const t = __importStar(require("typanion"));
const datadog_ci_base_1 = require("@datadog/datadog-ci-base");
const validation = __importStar(require("@datadog/datadog-ci-base/helpers/validation"));
const plugin_1 = require("../../helpers/plugin");
class JunitUploadCommand extends datadog_ci_base_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.basePaths = clipanion_1.Option.Rest({ required: 1 });
        this.verbose = clipanion_1.Option.Boolean('--verbose', false);
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false);
        this.env = clipanion_1.Option.String('--env');
        this.logs = clipanion_1.Option.String('--logs', 'false', {
            env: 'DD_CIVISIBILITY_LOGS_ENABLED',
            tolerateBoolean: true,
            validator: t.isBoolean(),
        });
        this.maxConcurrency = clipanion_1.Option.String('--max-concurrency', '20', { validator: validation.isInteger() });
        this.measures = clipanion_1.Option.Array('--measures');
        this.service = clipanion_1.Option.String('--service', { env: 'DD_SERVICE' });
        this.tags = clipanion_1.Option.Array('--tags');
        this.reportTags = clipanion_1.Option.Array('--report-tags');
        this.reportMeasures = clipanion_1.Option.Array('--report-measures');
        this.rawXPathTags = clipanion_1.Option.Array('--xpath-tag');
        this.gitRepositoryURL = clipanion_1.Option.String('--git-repository-url');
        this.skipGitMetadataUpload = clipanion_1.Option.String('--skip-git-metadata-upload', 'false', {
            validator: t.isBoolean(),
            tolerateBoolean: true,
        });
        this.automaticReportsDiscovery = clipanion_1.Option.String('--auto-discovery', 'false', {
            validator: t.isBoolean(),
            tolerateBoolean: true,
        });
        this.ignoredPaths = clipanion_1.Option.String('--ignored-paths');
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.JunitUploadCommand = JunitUploadCommand;
JunitUploadCommand.paths = [['junit', 'upload']];
JunitUploadCommand.usage = clipanion_1.Command.Usage({
    category: 'CI Visibility',
    description: 'Upload jUnit XML test reports files to Datadog.',
    details: `
      This command will upload to jUnit XML test reports files to Datadog.\n
      See README for details.
    `,
    examples: [
        ['Upload all jUnit XML test report files in current directory', 'datadog-ci junit upload --service my-service .'],
        [
            'Discover and upload all jUnit XML test report files doing recursive search in current directory',
            'datadog-ci junit upload --service my-service --auto-discovery .',
        ],
        [
            'Discover and upload all jUnit XML test report files doing recursive search in current directory, ignoring src/ignored-module-a and src/ignored-module-b',
            'datadog-ci junit upload --service my-service --ignored-paths src/ignored-module-a,src/ignored-module-b --auto-discovery .',
        ],
        [
            'Upload all jUnit XML test report files in src/unit-test-reports and src/acceptance-test-reports',
            'datadog-ci junit upload --service my-service src/unit-test-reports src/acceptance-test-reports',
        ],
        [
            'Upload all jUnit XML test report files in current directory and add extra tags globally',
            'datadog-ci junit upload --service my-service --tags key1:value1 --tags key2:value2 .',
        ],
        [
            'Upload all jUnit XML test report files in current directory and add extra measures globally',
            'datadog-ci junit upload --service my-service --measures key1:123 --measures key2:321 .',
        ],
        [
            'Upload all jUnit XML test report files in current directory to the datadoghq.eu site',
            'DD_SITE=datadoghq.eu datadog-ci junit upload --service my-service .',
        ],
        [
            'Upload all jUnit XML test report files in current directory while also collecting logs',
            'datadog-ci junit upload --service my-service --logs .',
        ],
        [
            'Upload all jUnit XML test report files in current directory customizing test suite with xpath',
            'datadog-ci junit upload --service my-service --xpath-tag test.suite=/testcase/@classname .',
        ],
        [
            'Upload all jUnit XML test report files in current directory adding a custom tag from property with xpath',
            "datadog-ci junit upload --service my-service --xpath-tag custom_tag=/testcase/..//property[@name='property-name'] .",
        ],
        [
            'Upload all jUnit XML test report files in current directory with extra verbosity',
            'datadog-ci junit upload --verbose --service my-service .',
        ],
    ],
});
