// @datadog/datadog-ci-base/dist/commands/trace/helper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomSpanCommand = void 0;
const crypto_1 = __importDefault(require("crypto"));
const chalk_1 = __importDefault(require("chalk"));
const clipanion_1 = require("clipanion");
const datadog_ci_base_1 = require("@datadog/datadog-ci-base");
const constants_1 = require("@datadog/datadog-ci-base/constants");
const api_1 = require("@datadog/datadog-ci-base/helpers/api");
const ci_1 = require("@datadog/datadog-ci-base/helpers/ci");
const env_1 = require("@datadog/datadog-ci-base/helpers/env");
const fips_1 = require("@datadog/datadog-ci-base/helpers/fips");
const format_git_span_data_1 = require("@datadog/datadog-ci-base/helpers/git/format-git-span-data");
const retry_1 = require("@datadog/datadog-ci-base/helpers/retry");
const tags_1 = require("@datadog/datadog-ci-base/helpers/tags");
const user_provided_git_1 = require("@datadog/datadog-ci-base/helpers/user-provided-git");
const api_2 = require("./api");
const interfaces_1 = require("./interfaces");
class CustomSpanCommand extends datadog_ci_base_1.BaseCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.measures = clipanion_1.Option.Array('--measures');
        this.dryRun = clipanion_1.Option.Boolean('--dry-run');
        this.tags = clipanion_1.Option.Array('--tags');
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.config = {
            apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
            envVarTags: process.env.DD_TAGS,
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
    }
    generateSpanId() {
        return crypto_1.default.randomBytes(5).toString('hex');
    }
    tryEnableFips() {
        (0, fips_1.enableFips)(this.fips || this.config.fips, this.fipsIgnoreError || this.config.fipsIgnoreError);
    }
    executeReportCustomSpan(id, startTime, endTime, extraTags) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const provider = (0, ci_1.getCIProvider)();
            if (!interfaces_1.SUPPORTED_PROVIDERS.includes(provider)) {
                this.context.stdout.write(`Unsupported CI provider "${provider}". Supported providers are: ${interfaces_1.SUPPORTED_PROVIDERS.join(', ')}\n`);
                return 1;
            }
            const realGithubJobName = (0, ci_1.getGithubJobNameFromLogs)(this.context);
            const realGithubJobID = (_a = process.env.JOB_CHECK_RUN_ID) !== null && _a !== void 0 ? _a : (0, ci_1.getGithubJobIDFromLogs)(this.context);
            const ciSpanTags = (0, ci_1.getCISpanTags)(realGithubJobName, realGithubJobID);
            const envVarTags = this.config.envVarTags ? (0, tags_1.parseTags)(this.config.envVarTags.split(',')) : {};
            const cliTags = this.tags ? (0, tags_1.parseTags)(this.tags) : {};
            const cliMeasures = this.measures ? (0, tags_1.parseTags)(this.measures) : {};
            const measures = Object.entries(cliMeasures).reduce((acc, [key, value]) => {
                const parsedValue = parseFloat(value);
                if (!isNaN(parsedValue)) {
                    return Object.assign(Object.assign({}, acc), { [key]: parsedValue });
                }
                return acc;
            }, {});
            const gitSpanTags = yield (0, format_git_span_data_1.getGitMetadata)();
            const userGitSpanTags = (0, user_provided_git_1.getUserGitSpanTags)();
            yield this.reportCustomSpan({
                start_time: startTime.toISOString(),
                end_time: endTime.toISOString(),
                ci_provider: provider,
                span_id: id,
                tags: Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, gitSpanTags), ciSpanTags), userGitSpanTags), cliTags), envVarTags),
                measures,
                command: extraTags.command,
                name: extraTags.name,
                error_message: extraTags.error_message,
                exit_code: extraTags.exit_code,
            });
            return 0;
        });
    }
    getApiHelper() {
        if (!this.config.apiKey) {
            this.context.stdout.write(`Neither ${chalk_1.default.red.bold('DATADOG_API_KEY')} nor ${chalk_1.default.red.bold('DD_API_KEY')} is in your environment.\n`);
            throw new Error('API key is missing');
        }
        return (0, api_2.apiConstructor)(this.getBaseIntakeUrl(), this.config.apiKey);
    }
    getBaseIntakeUrl() {
        return (0, api_1.getApiUrl)();
    }
    reportCustomSpan(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.dryRun) {
                this.context.stdout.write(`${chalk_1.default.green.bold('[DRY-RUN]')} Reporting custom span: ${JSON.stringify(payload)}\n`);
                return;
            }
            const api = this.getApiHelper();
            try {
                yield (0, retry_1.retryRequest)(() => api.reportCustomSpan(payload), {
                    onRetry: (e, attempt) => {
                        this.context.stderr.write(chalk_1.default.yellow(`[attempt ${attempt}] Could not report custom span. Retrying...: ${e.message}\n`));
                    },
                    retries: 5,
                });
            }
            catch (error) {
                this.handleError(error);
            }
        });
    }
    handleError(error) {
        this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} Failed to report custom span: ` +
            `${error.response ? JSON.stringify(error.response.data, undefined, 2) : ''}\n`);
    }
}
exports.CustomSpanCommand = CustomSpanCommand;
