// @datadog/datadog-ci-base/dist/commands/trace/test-utils.js
"use strict";
// Only the function is exported, not the test such that there is no test duplication
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCIProviderTests = void 0;
const fs_1 = __importDefault(require("fs"));
const ci = __importStar(require("@datadog/datadog-ci-base/helpers/ci"));
/* eslint-disable jest/no-export */
const makeCIProviderTests = (runCLI, runCLIArgs) => {
    describe('execute', () => {
        beforeEach(() => {
            jest.spyOn(ci, 'getGithubJobIDFromLogs').mockReturnValue(undefined);
        });
        afterEach(() => {
            jest.resetAllMocks();
        });
        test('should fail if no CI is detected', () => __awaiter(void 0, void 0, void 0, function* () {
            process.env = {};
            const { context, code } = yield runCLI(runCLIArgs);
            expect(code).toBe(1);
            expect(context.stdout.toString()).toContain('Unsupported CI provider "unknown"');
        }));
        test('should detect the circleci environment', () => __awaiter(void 0, void 0, void 0, function* () {
            const { context, code } = yield runCLI(runCLIArgs, {
                CIRCLECI: 'true',
                CIRCLE_WORKFLOW_ID: 'test',
                CIRCLE_BUILD_NUM: '10',
                NON_CIRCLE_ENV: 'bar',
            });
            expect(code).toBe(0);
            const dryRunOutput = context.stdout.toString();
            expect(dryRunOutput).toContain('\\"CIRCLE_WORKFLOW_ID\\":\\"test\\"');
            expect(dryRunOutput).toContain('\\"CIRCLE_BUILD_NUM\\":\\"10\\"');
        }));
        test('should detect the jenkins environment', () => __awaiter(void 0, void 0, void 0, function* () {
            const { context, code } = yield runCLI(runCLIArgs, {
                DD_CUSTOM_TRACE_ID: 'abc',
                DD_CUSTOM_PARENT_ID: 'xyz',
                JENKINS_HOME: '/root',
                JENKINS_URL: 'http://jenkins',
                NON_JENKINS_ENV: 'bar',
                WORKSPACE: 'def',
            });
            expect(code).toBe(0);
            const dryRunOutput = context.stdout.toString();
            expect(dryRunOutput).toContain('\\"DD_CUSTOM_TRACE_ID\\":\\"abc\\"');
            expect(dryRunOutput).toContain('\\"DD_CUSTOM_PARENT_ID\\":\\"xyz\\"');
        }));
        test('should detect the github environment', () => __awaiter(void 0, void 0, void 0, function* () {
            const { context, code } = yield runCLI(runCLIArgs, {
                GITHUB_ACTIONS: 'true',
                GITHUB_SERVER_URL: 'http://github',
                GITHUB_REPOSITORY: 'test/test',
                GITHUB_RUN_ID: '10',
                GITHUB_RUN_ATTEMPT: '1',
                GITHUB_JOB: 'jobname',
                DD_GITHUB_JOB_NAME: 'custom_jobname',
            });
            expect(code).toBe(0);
            const dryRunOutput = context.stdout.toString();
            expect(dryRunOutput).toContain('\\"GITHUB_SERVER_URL\\":\\"http://github\\"');
            expect(dryRunOutput).toContain('\\"GITHUB_REPOSITORY\\":\\"test/test\\"');
            expect(dryRunOutput).toContain('\\"GITHUB_RUN_ID\\":\\"10\\"');
            expect(dryRunOutput).toContain('\\"GITHUB_RUN_ATTEMPT\\":\\"1\\"');
            expect(dryRunOutput).toContain('\\"DD_GITHUB_JOB_NAME\\":\\"custom_jobname\\"');
            expect(dryRunOutput).toContain('"ci.job.name":"jobname"');
            expect(dryRunOutput).not.toContain('Determining GitHub job name');
        }));
        test('should detect the gitlab environment', () => __awaiter(void 0, void 0, void 0, function* () {
            const { context, code } = yield runCLI(runCLIArgs, {
                GITLAB_CI: 'true',
                CI_PROJECT_URL: 'http://gitlab',
                CI_PIPELINE_ID: '10',
                CI_JOB_ID: '50',
            });
            expect(code).toBe(0);
            const dryRunOutput = context.stdout.toString();
            expect(dryRunOutput).toContain('\\"CI_PROJECT_URL\\":\\"http://gitlab\\"');
            expect(dryRunOutput).toContain('\\"CI_PIPELINE_ID\\":\\"10\\"');
            expect(dryRunOutput).toContain('\\"CI_JOB_ID\\":\\"50\\"');
        }));
        test('should detect the azure environment', () => __awaiter(void 0, void 0, void 0, function* () {
            const { context, code } = yield runCLI(runCLIArgs, {
                TF_BUILD: 'true',
                SYSTEM_TEAMPROJECTID: 'test',
                BUILD_BUILDID: '10',
                SYSTEM_JOBID: '3acfg',
            });
            expect(code).toBe(0);
            const dryRunOutput = context.stdout.toString();
            expect(dryRunOutput).toContain('\\"SYSTEM_TEAMPROJECTID\\":\\"test\\"');
            expect(dryRunOutput).toContain('\\"BUILD_BUILDID\\":\\"10\\"');
            expect(dryRunOutput).toContain('\\"SYSTEM_JOBID\\":\\"3acfg\\"');
        }));
        test('should detect the aws codepipeline environment', () => __awaiter(void 0, void 0, void 0, function* () {
            const { context, code } = yield runCLI(runCLIArgs, {
                CODEBUILD_INITIATOR: 'codepipeline-abc',
                DD_PIPELINE_EXECUTION_ID: 'def-234',
                CODEBUILD_BUILD_ARN: 'arn:aws:codebuild:us-west-2:123456789012:build/MyProjectName:6a8f0d8a',
            });
            expect(code).toBe(0);
            const dryRunOutput = context.stdout.toString();
            expect(dryRunOutput).toContain('\\"DD_PIPELINE_EXECUTION_ID\\":\\"def-234\\"');
            expect(dryRunOutput).toContain('\\"CODEBUILD_BUILD_ARN\\":\\"arn:aws:codebuild:us-west-2:123456789012:build/MyProjectName:6a8f0d8a\\"');
        }));
        test('should detect the buildkite environment', () => __awaiter(void 0, void 0, void 0, function* () {
            const { context, code } = yield runCLI(runCLIArgs, {
                BUILDKITE: 'true',
                BUILDKITE_BUILD_ID: 'abc',
                BUILDKITE_JOB_ID: 'def',
            });
            expect(code).toBe(0);
            const dryRunOutput = context.stdout.toString();
            expect(dryRunOutput).toContain('\\"BUILDKITE_BUILD_ID\\":\\"abc\\"');
            expect(dryRunOutput).toContain('\\"BUILDKITE_JOB_ID\\":\\"def\\"');
        }));
        test('should try to infer github job name', () => __awaiter(void 0, void 0, void 0, function* () {
            jest.spyOn(fs_1.default, 'readdirSync').mockReturnValue([
                {
                    name: 'Worker_1.log',
                    isFile: () => true,
                    isDirectory: () => false,
                    isBlockDevice: () => false,
                    isCharacterDevice: () => false,
                    isSymbolicLink: () => false,
                    isFIFO: () => false,
                    isSocket: () => false,
                    parentPath: '',
                    path: '',
                },
            ]);
            jest.spyOn(fs_1.default, 'readFileSync').mockReturnValue(`{"jobDisplayName": "real job name", "job": {"v": 12345}}`);
            const { context, code } = yield runCLI(runCLIArgs, {
                GITHUB_ACTIONS: 'true',
                GITHUB_SERVER_URL: 'http://github',
                GITHUB_REPOSITORY: 'test/test',
                GITHUB_RUN_ID: '10',
                GITHUB_RUN_ATTEMPT: '1',
                GITHUB_JOB: 'jobname',
            });
            expect(code).toBe(0);
            const dryRunOutput = context.stdout.toString();
            expect(dryRunOutput).toContain('\\"DD_GITHUB_JOB_NAME\\":\\"real job name\\"');
            expect(dryRunOutput).toContain('Determining GitHub job name');
        }));
    });
};
exports.makeCIProviderTests = makeCIProviderTests;
