// @datadog/datadog-ci-plugin-deployment/dist/commands/gate.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginCommand = void 0;
const gate_1 = require("@datadog/datadog-ci-base/commands/deployment/gate");
const constants_1 = require("@datadog/datadog-ci-base/constants");
const api_1 = require("@datadog/datadog-ci-base/helpers/api");
const env_1 = require("@datadog/datadog-ci-base/helpers/env");
const fips_1 = require("@datadog/datadog-ci-base/helpers/fips");
const formatting_1 = require("@datadog/datadog-ci-base/helpers/formatting");
const logger_1 = require("@datadog/datadog-ci-base/helpers/logger");
const retry_1 = require("@datadog/datadog-ci-base/helpers/retry");
const utils_1 = require("@datadog/datadog-ci-base/helpers/utils");
const axios_1 = require("axios");
const chalk_1 = __importDefault(require("chalk"));
const api_2 = require("../api");
class PluginCommand extends gate_1.DeploymentGateCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.config = {
            apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
            appKey: process.env.DATADOG_APP_KEY || process.env.DD_APP_KEY,
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
        this.logger = new logger_1.Logger((s) => this.context.stdout.write(s), logger_1.LogLevel.INFO);
        this.evaluationRequestTimeout = 60000; // 1 minute
        this.pollingInterval = 15000; // 15 seconds
        this.startTime = Date.now();
    }
    execute() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            (0, fips_1.enableFips)(this.fips || this.config.fips, this.fipsIgnoreError || this.config.fipsIgnoreError);
            if (!this.service) {
                this.logger.error(chalk_1.default.red(`${formatting_1.ICONS.FAILED} Missing required parameter: ${chalk_1.default.bold('--service')}`));
                return 1;
            }
            if (!this.env) {
                this.logger.error(chalk_1.default.red(`${formatting_1.ICONS.FAILED} Missing required parameter: ${chalk_1.default.bold('--env')}`));
                return 1;
            }
            const timeoutSeconds = parseInt(this.timeout, 10);
            if (isNaN(timeoutSeconds) || timeoutSeconds <= 0) {
                this.logger.error(chalk_1.default.red(`${formatting_1.ICONS.FAILED} Invalid ${chalk_1.default.bold('--timeout')} value. Must be a positive integer.`));
                return 1;
            }
            const timeoutMilliseconds = timeoutSeconds * 1000;
            if (!this.config.apiKey) {
                this.logger.error(`${formatting_1.ICONS.FAILED} Neither ${chalk_1.default.red.bold('DATADOG_API_KEY')} nor ${chalk_1.default.red.bold('DD_API_KEY')} are in your environment.`);
                return 1;
            }
            if (!this.config.appKey) {
                this.logger.error(`${formatting_1.ICONS.FAILED} Neither ${chalk_1.default.red.bold('DATADOG_APP_KEY')} nor ${chalk_1.default.red.bold('DD_APP_KEY')} are in your environment.`);
                return 1;
            }
            this.logger.info('Starting deployment gate evaluation with parameters:');
            this.logger.info(`\tService: ${this.service}`);
            this.logger.info(`\tEnvironment: ${this.env}`);
            if (this.identifier) {
                this.logger.info(`\tIdentifier: ${this.identifier}`);
            }
            if (this.version) {
                this.logger.info(`\tVersion: ${this.version}`);
            }
            if (this.apmPrimaryTag) {
                this.logger.info(`\tAPM Primary Tag: ${this.apmPrimaryTag}`);
            }
            this.logger.info(`\tTimeout: ${timeoutSeconds} seconds`);
            this.logger.info(`\tFail on error: ${this.failOnError ? 'true' : 'false'}\n`);
            let result;
            try {
                const api = this.getApiHelper(this.config.apiKey, this.config.appKey);
                const evaluationRequest = this.buildEvaluationRequest();
                const evaluationId = yield this.requestGateEvaluation(api, evaluationRequest, this.evaluationRequestTimeout);
                result = yield this.pollForEvaluationResults(api, evaluationId, timeoutMilliseconds);
            }
            catch (error) {
                const errorMessage = error instanceof Error ? error.message : String(error);
                this.logger.error(`Deployment gate evaluation failed due to a non-retryable error: ${errorMessage}`);
                if ((0, axios_1.isAxiosError)(error) && ((_a = error.response) === null || _a === void 0 ? void 0 : _a.status)) {
                    if (error.response.status >= 400 && error.response.status < 500) {
                        this.logger.error(`${formatting_1.ICONS.FAILED} Request failed with client error, exiting with status 1`);
                        return 1;
                    }
                }
                result = this.getResultForDatadogError();
            }
            return result === 'PASS' ? 0 : 1;
        });
    }
    getApiHelper(apiKey, appKey) {
        return (0, api_2.apiConstructor)(`https://${(0, utils_1.getApiHostForSite)((0, api_1.getDatadogSite)())}`, apiKey, appKey);
    }
    buildEvaluationRequest() {
        const request = {
            service: this.service || '',
            env: this.env || '',
        };
        if (this.identifier) {
            request.identifier = this.identifier;
        }
        if (this.version) {
            request.version = this.version;
        }
        if (this.apmPrimaryTag) {
            request.apm_primary_tag = this.apmPrimaryTag;
        }
        if (this.monitorsQueryVariable) {
            request.monitors_query_variable = this.monitorsQueryVariable;
        }
        return request;
    }
    requestGateEvaluation(api, request, timeout) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.info('Requesting gate evaluation...');
            const doRequest = () => __awaiter(this, void 0, void 0, function* () {
                var _a;
                try {
                    const response = yield api.requestGateEvaluation(request);
                    const id = response.data.data.attributes.evaluation_id;
                    this.logger.info(chalk_1.default.green(`Gate evaluation started successfully. Evaluation ID: ${id}\n`));
                    return id;
                }
                catch (error) {
                    if ((0, axios_1.isAxiosError)(error) && ((_a = error.response) === null || _a === void 0 ? void 0 : _a.status)) {
                        this.logger.error(`Request failed with error: ${error.response.status} ${error.response.statusText}`);
                    }
                    else {
                        const errorMessage = error instanceof Error ? error.message : String(error);
                        this.logger.error(`Could not start gate evaluation with unknown error: ${errorMessage}`);
                    }
                    throw error;
                }
            });
            const evaluationId = yield (0, retry_1.retryRequest)(doRequest, Object.assign(Object.assign({}, this.getRetryOptions(timeout)), { onRetry: (e, attempt) => {
                    this.logger.info(`Retrying gate evaluation request (${attempt} attempts)...`);
                } }));
            return evaluationId;
        });
    }
    pollForEvaluationResults(api, evaluationId, timeout) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.info('Waiting for gate evaluation results...');
            let timePassed = Date.now() - this.startTime;
            let result;
            while (timePassed < timeout) {
                const remainingTime = timeout - timePassed;
                const waitTime = Math.min(this.pollingInterval, remainingTime);
                const waitTimeInSeconds = Math.floor(waitTime / 1000);
                result = yield this.getEvaluationResult(api, evaluationId);
                if (result) {
                    return result;
                }
                this.logger.info(`\tRetrying in ${waitTimeInSeconds}s...`);
                yield new Promise((resolve) => setTimeout(resolve, waitTime));
                timePassed = Date.now() - this.startTime;
            }
            // The block above may not have run the last time, so we need to check again
            result = yield this.getEvaluationResult(api, evaluationId);
            if (result) {
                return result;
            }
            this.logger.warn(`${formatting_1.ICONS.WARNING} Timeout reached (${timeout / 1000} seconds). Gate evaluation did not complete in time.`);
            return this.getResultForDatadogError();
        });
    }
    getEvaluationResult(api, evaluationId) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield api.getGateEvaluationResult(evaluationId);
                const status = response.data.data.attributes.gate_status;
                switch (status) {
                    case 'pass':
                        this.renderEvaluationSummary(response.data);
                        return 'PASS';
                    case 'fail':
                        this.renderEvaluationSummary(response.data);
                        return 'FAIL';
                    case 'in_progress': {
                        const rules = response.data.data.attributes.rules;
                        const totalRules = rules.length;
                        const completedRules = rules.filter((rule) => rule.status !== 'in_progress').length;
                        this.logger.info(`\tGate evaluation in progress (${completedRules}/${totalRules} rules completed)`);
                        break;
                    }
                    default:
                        this.logger.warn(`Unknown gate evaluation status: ${status}`);
                }
            }
            catch (error) {
                if ((0, axios_1.isAxiosError)(error) && ((_a = error.response) === null || _a === void 0 ? void 0 : _a.status)) {
                    const status = error.response.status;
                    const statusText = error.response.statusText;
                    if (status === 404 || status >= 500) {
                        this.logger.error(`Error polling for gate evaluation results: ${status} ${statusText}`);
                        // We want to retry in this case, so we return undefined to trigger the retry
                        return;
                    }
                    else {
                        this.logger.error(`Error polling for gate evaluation results: ${status} ${statusText}`);
                    }
                }
                else {
                    const errorMessage = error instanceof Error ? error.message : String(error);
                    this.logger.error(`Error polling for gate evaluation results: ${errorMessage}`);
                }
                throw error;
            }
        });
    }
    renderEvaluationSummary(result) {
        const attributes = result.data.attributes;
        if (attributes.gate_status === 'pass') {
            this.logger.info(chalk_1.default.green(`\n${formatting_1.ICONS.SUCCESS} Gate evaluation passed`));
        }
        else if (attributes.gate_status === 'fail') {
            this.logger.info(chalk_1.default.red(`\n${formatting_1.ICONS.FAILED} Gate evaluation failed`));
        }
        this.logger.info(`   Evaluation mode: ${attributes.dry_run ? 'Dry run' : 'Active'}`);
        this.logger.info(`   Evaluation URL: ${attributes.evaluation_url}`);
        this.logger.info(`   Rules evaluated:`);
        attributes.rules.forEach((rule) => {
            var _a;
            const ruleName = `    - Rule: ${rule.name}`;
            const evaluationMode = `\n      Evaluation mode: ${rule.dry_run ? 'Dry run' : 'Active'}`;
            const status = `\n      Status: ${rule.status.toUpperCase()}`;
            const reason = rule.status === 'fail' ? `\n      Reason: ${(_a = rule.reason) !== null && _a !== void 0 ? _a : 'Unknown'}` : '';
            this.logger.info(`${ruleName}${evaluationMode}${status}${reason}`);
        });
        this.logger.info('\n');
    }
    getResultForDatadogError() {
        if (this.failOnError) {
            this.logger.warn('Unexpected error happened, exiting with status 1 (fail) because --fail-on-error is enabled');
            return 'FAIL';
        }
        this.logger.warn('Unexpected error happened, exiting with status 0 (pass)');
        this.logger.info('You can override this behavior by setting --fail-on-error to true');
        return 'PASS';
    }
    getRetryOptions(timeout) {
        const timeElapsed = Date.now() - this.startTime;
        const maxRetryTime = timeout - timeElapsed;
        return {
            forever: true,
            randomize: false,
            factor: 1.5,
            minTimeout: 5000,
            maxTimeout: 30000,
            maxRetryTime,
        };
    }
}
exports.PluginCommand = PluginCommand;
