// @datadog/datadog-ci-plugin-deployment/dist/commands/mark.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginCommand = void 0;
const mark_1 = require("@datadog/datadog-ci-base/commands/deployment/mark");
const tag_1 = require("@datadog/datadog-ci-base/commands/tag/tag");
const constants_1 = require("@datadog/datadog-ci-base/constants");
const env_1 = require("@datadog/datadog-ci-base/helpers/env");
const fips_1 = require("@datadog/datadog-ci-base/helpers/fips");
const constants_2 = require("../constants");
class PluginCommand extends mark_1.DeploymentMarkCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.config = {
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            (0, fips_1.enableFips)(this.fips || this.config.fips, this.fipsIgnoreError || this.config.fipsIgnoreError);
            const tagJobCommand = new tag_1.TagCommand();
            tagJobCommand.setLevel('job');
            tagJobCommand.setTags(this.createJobDeploymentTags());
            tagJobCommand.context = this.context;
            tagJobCommand.setSilent(false);
            tagJobCommand.setDryRun(this.dryRun);
            const tagPipelineCommand = new tag_1.TagCommand();
            tagPipelineCommand.setLevel('pipeline');
            tagPipelineCommand.setTags(this.createPipelineDeploymentTags());
            tagPipelineCommand.context = this.context;
            tagPipelineCommand.setSilent(!this.dryRun);
            tagPipelineCommand.setDryRun(this.dryRun);
            if (this.noFail) {
                tagJobCommand.setNoFail(true);
                tagPipelineCommand.setNoFail(true);
            }
            const tagJobCommandExitCode = yield tagJobCommand.execute();
            if (tagJobCommandExitCode === 0) {
                return tagPipelineCommand.execute();
            }
            else {
                return tagJobCommandExitCode;
            }
        });
    }
    createJobDeploymentTags() {
        const tags = [constants_2.IS_DEPLOYMENT_TAG];
        if (this.env) {
            tags.push(constants_2.ENV_TAG + this.env);
        }
        if (this.revision) {
            tags.push(constants_2.REVISION_TAG + this.revision);
        }
        if (this.service) {
            tags.push(constants_2.SERVICE_TAG + this.service);
        }
        if (this.isRollback) {
            tags.push(constants_2.IS_ROLLBACK_TAG);
        }
        if (this.tags) {
            tags.push(constants_2.CUSTOM_TAGS_TAG + this.tags.join(','));
        }
        return tags;
    }
    createPipelineDeploymentTags() {
        return [constants_2.CONTAINS_DEPLOYMENT_TAG];
    }
}
exports.PluginCommand = PluginCommand;
