// @datadog/datadog-ci-base/dist/commands/cloud-run/flare.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudRunFlareCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const __1 = require("../..");
class CloudRunFlareCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.isDryRun = clipanion_1.Option.Boolean('-d,--dry,--dry-run', false, {
            description: 'Preview data that will be sent to Datadog support.',
        });
        this.withLogs = clipanion_1.Option.Boolean('--with-logs', false, {
            description: 'Collect recent logs for the specified service.',
        });
        this.service = clipanion_1.Option.String('-s,--service', { description: 'The name of the Cloud Run service.' });
        this.project = clipanion_1.Option.String('-p,--project', {
            description: 'The name of the Google Cloud project where the Cloud Run service is hosted.',
        });
        this.region = clipanion_1.Option.String('-r,--region,-l,--location', {
            description: 'The region where the Cloud Run service is hosted.',
        });
        this.caseId = clipanion_1.Option.String('-c,--case-id', { description: 'The Datadog case ID to send the files to.' });
        this.email = clipanion_1.Option.String('-e,--email', { description: 'The email associated with the specified case ID.' });
        this.start = clipanion_1.Option.String('--start', {
            description: 'Only gather logs after the time in milliseconds since Unix Epoch. (`--with-logs` must be specified.)',
        });
        this.end = clipanion_1.Option.String('--end', {
            description: 'Only gather logs before the time in milliseconds since Unix Epoch. (`--with-logs` must be specified.)',
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.CloudRunFlareCommand = CloudRunFlareCommand;
CloudRunFlareCommand.paths = [['cloud-run', 'flare']];
CloudRunFlareCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Gather Cloud Run service configuration and sends it to Datadog.',
});
