// @datadog/datadog-ci-base/dist/commands/cloud-run/instrument.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudRunInstrumentCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const constants_1 = require("../../helpers/serverless/constants");
const __1 = require("../..");
const DEFAULT_SIDECAR_IMAGE = 'gcr.io/datadoghq/serverless-init:latest';
class CloudRunInstrumentCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // protected configPath = Option.String('--config') implement if requested by customers
        this.dryRun = clipanion_1.Option.Boolean('-d,--dry,--dry-run', false, {
            description: 'Run the command in dry-run mode, without making any changes. Preview the changes that running the command would apply.',
        });
        this.extraTags = clipanion_1.Option.String('--extra-tags,--extraTags', {
            description: 'Add custom tags to your Cloud Run service in Datadog. Must be a list of `<key:><value>` separated by commas.',
        });
        this.envVars = clipanion_1.Option.Array('-e,--env-vars', {
            description: 'Additional environment variables to set for the Cloud Run service. Can specify multiple variables in the format `--env-vars VAR1=VALUE1 --env-vars VAR2=VALUE2`.',
        });
        this.project = clipanion_1.Option.String('-p,--project', {
            description: 'The name of the Google Cloud project where the Cloud Run service is hosted.',
        });
        this.services = clipanion_1.Option.Array('-s,--service,--services', [], {
            description: 'Cloud Run service(s) to instrument',
        });
        this.interactive = clipanion_1.Option.Boolean('-i,--interactive', false, {
            description: 'Interactively choose which service gets instrumented. No need for other flags.',
        });
        this.region = clipanion_1.Option.String('-r,--region', {
            description: 'The region where the Cloud Run service is hosted.',
        });
        this.logLevel = clipanion_1.Option.String('--log-level,--logLevel', { description: 'Specify your Datadog log level.' });
        this.sourceCodeIntegration = clipanion_1.Option.Boolean('--source-code-integration,--sourceCodeIntegration', true, {
            description: 'Whether to enable the Datadog Source Code integration. This tags your service(s) with the Git repository and the latest commit hash of the local directory. Specify `--no-source-code-integration` to disable.',
        });
        this.uploadGitMetadata = clipanion_1.Option.Boolean('--upload-git-metadata,--uploadGitMetadata', true, {
            description: "Whether to enable Git metadata uploading, as a part of the source code integration. Git metadata uploading is only required if you don't have the Datadog GitHub integration installed. Specify `--no-upload-git-metadata` to disable.",
        });
        this.tracing = clipanion_1.Option.String('--tracing', {
            description: 'Enables tracing of your application if the tracer is installed. Disable tracing by setting `--tracing false`.',
        });
        this.serviceTag = clipanion_1.Option.String('--service-tag,--serviceTag', {
            description: 'The value for the service tag. Use this to group related Cloud Run services belonging to similar workloads. For example, `my-service`. If not provided, the Cloud Run service name is used.',
        });
        this.version = clipanion_1.Option.String('--version', {
            description: 'The value for the version tag. Use this to correlate spikes in latency, load, or errors to new versions. For example, `1.0.0`.',
        });
        this.environment = clipanion_1.Option.String('--env', {
            description: 'The value for the env tag. Use this to separate your staging, development, and production environments. For example, `prod`.',
        });
        this.llmobs = clipanion_1.Option.String('--llmobs', {
            description: 'If specified, enables LLM Observability for the instrumented service(s) with the provided ML application name.',
        });
        this.healthCheckPort = clipanion_1.Option.String('--port,--health-check-port,--healthCheckPort');
        this.sidecarImage = clipanion_1.Option.String('--image,--sidecar-image', DEFAULT_SIDECAR_IMAGE, {
            description: `The image to use for the sidecar container. Defaults to '${DEFAULT_SIDECAR_IMAGE}'`,
        });
        this.sidecarName = clipanion_1.Option.String('--sidecar-name', constants_1.DEFAULT_SIDECAR_NAME, {
            description: `(Not recommended) The name to use for the sidecar container. Defaults to '${constants_1.DEFAULT_SIDECAR_NAME}'`,
        });
        this.sharedVolumeName = clipanion_1.Option.String('--shared-volume-name', constants_1.DEFAULT_VOLUME_NAME, {
            description: `(Not recommended) Specify a custom shared volume name. Defaults to '${constants_1.DEFAULT_VOLUME_NAME}'`,
        });
        this.sharedVolumePath = clipanion_1.Option.String('--shared-volume-path', constants_1.DEFAULT_VOLUME_PATH, {
            description: `(Not recommended) Specify a custom shared volume path. Defaults to '${constants_1.DEFAULT_VOLUME_PATH}'`,
        });
        this.logsPath = clipanion_1.Option.String('--logs-path', constants_1.DEFAULT_LOGS_PATH, {
            description: `(Not recommended) Specify a custom log file path. Must begin with the shared volume path. Defaults to '${constants_1.DEFAULT_LOGS_PATH}'`,
        });
        this.sidecarCpus = clipanion_1.Option.String('--sidecar-cpus', '1', {
            description: `The number of CPUs to allocate to the sidecar container. Defaults to 1.`,
        });
        this.sidecarMemory = clipanion_1.Option.String('--sidecar-memory', '512Mi', {
            description: `The amount of memory to allocate to the sidecar container. Defaults to '512Mi'.`,
        });
        this.language = clipanion_1.Option.String('--language', {
            description: `Set the language used in your container or function for advanced log parsing. Sets the DD_SOURCE env var. Possible values: "nodejs", "python", "go", "java", "csharp", "ruby", or "php".`,
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.CloudRunInstrumentCommand = CloudRunInstrumentCommand;
CloudRunInstrumentCommand.paths = [['cloud-run', 'instrument']];
CloudRunInstrumentCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Apply Datadog instrumentation to a Cloud Run app.',
});
