// @datadog/datadog-ci-base/dist/commands/cloud-run/uninstrument.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudRunUninstrumentCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const constants_1 = require("../../helpers/serverless/constants");
const __1 = require("../..");
class CloudRunUninstrumentCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.dryRun = clipanion_1.Option.Boolean('-d,--dry,--dry-run', false, {
            description: 'Run the command in dry-run mode, without making any changes. Preview the changes that running the command would apply.',
        });
        this.project = clipanion_1.Option.String('-p,--project', {
            description: 'The name of the Google Cloud project where the Cloud Run service is hosted.',
        });
        this.services = clipanion_1.Option.Array('-s,--service,--services', [], {
            description: 'Cloud Run service(s) to revert instrumentation',
        });
        this.interactive = clipanion_1.Option.Boolean('-i,--interactive', false, {
            description: 'Interactively choose which service gets instrumented. No need for other flags.',
        });
        this.region = clipanion_1.Option.String('-r,--region', {
            description: 'The region where the Cloud Run service is hosted.',
        });
        this.sidecarName = clipanion_1.Option.String('--sidecar-name', constants_1.DEFAULT_SIDECAR_NAME, {
            description: `The name of the sidecar container to remove. Specify if you have a different sidecar name. Defaults to '${constants_1.DEFAULT_SIDECAR_NAME}'`,
        });
        this.sharedVolumeName = clipanion_1.Option.String('--shared-volume-name', constants_1.DEFAULT_VOLUME_NAME, {
            description: `The name of the shared volume to remove. Specify if you have a different shared volume name. Defaults to '${constants_1.DEFAULT_VOLUME_NAME}'`,
        });
        this.envVars = clipanion_1.Option.Array('-e,--env-vars', {
            description: 'Additional environment variables to remove from the Cloud Run service. Can specify multiple variables in the format `--env-vars VAR1=VALUE1 --env-vars VAR2=VALUE2`.',
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.CloudRunUninstrumentCommand = CloudRunUninstrumentCommand;
CloudRunUninstrumentCommand.paths = [['cloud-run', 'uninstrument']];
CloudRunUninstrumentCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Revert Datadog instrumentation in a Cloud Run app.',
});
