// @datadog/datadog-ci-base/dist/commands/dsyms/upload.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DsymsUploadCommand = void 0;
const fs_1 = require("fs");
const chalk_1 = __importDefault(require("chalk"));
const clipanion_1 = require("clipanion");
const upath_1 = __importDefault(require("upath"));
const datadog_ci_base_1 = require("@datadog/datadog-ci-base");
const constants_1 = require("@datadog/datadog-ci-base/constants");
const api_1 = require("@datadog/datadog-ci-base/helpers/api");
const apikey_1 = require("@datadog/datadog-ci-base/helpers/apikey");
const concurrency_1 = require("@datadog/datadog-ci-base/helpers/concurrency");
const env_1 = require("@datadog/datadog-ci-base/helpers/env");
const errors_1 = require("@datadog/datadog-ci-base/helpers/errors");
const fips_1 = require("@datadog/datadog-ci-base/helpers/fips");
const glob_1 = require("@datadog/datadog-ci-base/helpers/glob");
const metrics_1 = require("@datadog/datadog-ci-base/helpers/metrics");
const upload_1 = require("@datadog/datadog-ci-base/helpers/upload");
const utils_1 = require("@datadog/datadog-ci-base/helpers/utils");
const validation = __importStar(require("@datadog/datadog-ci-base/helpers/validation"));
const validation_1 = require("@datadog/datadog-ci-base/helpers/validation");
const version_1 = require("@datadog/datadog-ci-base/version");
const interfaces_1 = require("./interfaces");
const renderer_1 = require("./renderer");
const utils_2 = require("./utils");
class DsymsUploadCommand extends datadog_ci_base_1.BaseCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.basePath = clipanion_1.Option.String({ required: true });
        this.configPath = clipanion_1.Option.String('--config');
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false);
        this.maxConcurrency = clipanion_1.Option.String('--max-concurrency', '20', { validator: validation.isInteger() });
        this.cliVersion = version_1.cliVersion;
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.fipsConfig = {
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
        this.config = {
            datadogSite: 'datadoghq.com',
        };
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            (0, fips_1.enableFips)(this.fips || this.fipsConfig.fips, this.fipsIgnoreError || this.fipsConfig.fipsIgnoreError);
            this.basePath = upath_1.default.normalize(this.basePath);
            this.context.stdout.write((0, renderer_1.renderCommandInfo)(this.basePath, this.maxConcurrency, this.dryRun));
            yield this.loadConfig();
            const { metricsLogger, apiKeyValidator } = this.createServices();
            const initialTime = Date.now();
            try {
                const tmpDirectory = yield (0, utils_2.createUniqueTmpDirectory)();
                try {
                    const results = yield this.processAndUploadDsyms(tmpDirectory, metricsLogger, apiKeyValidator);
                    const totalTime = (Date.now() - initialTime) / 1000;
                    this.context.stdout.write((0, renderer_1.renderSuccessfulCommand)(results, totalTime, this.dryRun));
                    metricsLogger.logger.gauge('duration', totalTime);
                    return 0;
                }
                finally {
                    yield (0, utils_2.deleteDirectory)(tmpDirectory);
                }
            }
            catch (error) {
                if (error instanceof errors_1.InvalidConfigurationError) {
                    this.context.stdout.write((0, renderer_1.renderConfigurationError)(error));
                    return 1;
                }
                throw error;
            }
            finally {
                try {
                    yield metricsLogger.flush();
                }
                catch (err) {
                    this.context.stdout.write(`WARN: ${err}\n`);
                }
            }
        });
    }
    loadConfig() {
        return __awaiter(this, void 0, void 0, function* () {
            this.config = yield (0, utils_1.resolveConfigFromFileAndEnvironment)(this.config, {
                apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
                datadogSite: (0, api_1.getDatadogSiteFromEnv)(),
            }, {
                configPath: this.configPath,
                defaultConfigPaths: ['datadog-ci.json', '../datadog-ci.json'],
                configFromFileCallback: (configFromFile) => {
                    (0, validation_1.checkAPIKeyOverride)(process.env.DATADOG_API_KEY || process.env.DD_API_KEY, configFromFile.apiKey, this.context.stdout);
                },
            });
        });
    }
    createServices() {
        const metricsLogger = (0, metrics_1.getMetricsLogger)({
            apiKey: this.config.apiKey,
            datadogSite: this.config.datadogSite,
            defaultTags: [`cli_version:${this.cliVersion}`],
            prefix: 'datadog.ci.dsyms.',
        });
        const apiKeyValidator = (0, apikey_1.newApiKeyValidator)({
            apiKey: this.config.apiKey,
            datadogSite: this.config.datadogSite,
            metricsLogger: metricsLogger.logger,
        });
        return { metricsLogger, apiKeyValidator };
    }
    processAndUploadDsyms(tmpDirectory, metricsLogger, apiKeyValidator) {
        return __awaiter(this, void 0, void 0, function* () {
            const intermediateDirectory = (0, utils_1.buildPath)(tmpDirectory, 'datadog-ci', 'dsyms', 'intermediate');
            const uploadDirectory = (0, utils_1.buildPath)(tmpDirectory, 'datadog-ci', 'dsyms', 'upload');
            this.context.stdout.write((0, renderer_1.renderCommandDetail)(intermediateDirectory, uploadDirectory));
            const searchDirectory = yield this.prepareSearchDirectory(tmpDirectory);
            const dsyms = yield this.findDsyms(searchDirectory);
            const thinDsyms = yield this.processDsyms(dsyms, intermediateDirectory);
            const compressedDsyms = yield this.compressDsyms(thinDsyms, uploadDirectory);
            const requestBuilder = this.createRequestBuilder();
            const uploadFunction = this.createUploadFunction(requestBuilder, metricsLogger, apiKeyValidator);
            return (0, concurrency_1.doWithMaxConcurrency)(this.maxConcurrency, compressedDsyms, uploadFunction);
        });
    }
    prepareSearchDirectory(tmpDirectory) {
        return __awaiter(this, void 0, void 0, function* () {
            if (yield (0, utils_2.isZipFile)(this.basePath)) {
                yield (0, utils_2.unzipArchiveToDirectory)(this.basePath, tmpDirectory);
                return tmpDirectory;
            }
            return this.basePath;
        });
    }
    findDsyms(directoryPath) {
        return __awaiter(this, void 0, void 0, function* () {
            const dsymPaths = (0, glob_1.globSync)((0, utils_1.buildPath)(directoryPath, '**/*.dSYM'));
            const results = yield Promise.all(dsymPaths.map((bundle) => __awaiter(this, void 0, void 0, function* () {
                try {
                    const { stdout } = yield (0, utils_2.executeDwarfdump)(bundle);
                    const dwarf = this.parseDwarfdumpOutput(stdout);
                    return [{ bundle, dwarf }];
                }
                catch (_a) {
                    this.context.stdout.write((0, renderer_1.renderInvalidDsymWarning)(bundle));
                    return [];
                }
            })));
            return results.flat();
        });
    }
    /**
     * Parses the output of `dwarfdump --uuid` command (ref.: https://www.unix.com/man-page/osx/1/dwarfdump/).
     * It returns one or many DWARF UUID and arch read from the output.
     *
     * Example `dwarfdump --uuid` output:
     * ```
     * $ dwarfdump --uuid DDTest.framework.dSYM
     * UUID: C8469F85-B060-3085-B69D-E46C645560EA (armv7) DDTest.framework.dSYM/Contents/Resources/DWARF/DDTest
     * UUID: 06EE3D68-D605-3E92-B92D-2F48C02A505E (arm64) DDTest.framework.dSYM/Contents/Resources/DWARF/DDTest
     * ```
     */
    parseDwarfdumpOutput(output) {
        const lineRegexp = /UUID: ([0-9A-F]{8}-(?:[0-9A-F]{4}-){3}[0-9A-F]{12}) \(([a-z0-9_]+)\) (.+)/;
        return output
            .split('\n')
            .map((line) => {
            const match = line.match(lineRegexp);
            return match ? [{ uuid: match[1], arch: match[2], object: match[3] }] : [];
        })
            .flat();
    }
    /**
     * It takes `N` dSYMs and returns `N` or more dSYMs. If a dSYM includes more than one arch slice,
     * it will be thinned by extracting each arch to a new dSYM in `output`.
     */
    processDsyms(dsyms, output) {
        return __awaiter(this, void 0, void 0, function* () {
            yield fs_1.promises.mkdir(output, { recursive: true });
            const results = yield Promise.all(dsyms.map((dsym) => __awaiter(this, void 0, void 0, function* () {
                // Reduce dSYMs size by extracting single UUIDs and arch slices from fat dSYMs to separate
                // single-arch dSYMs in intermediate location. This is to avoid exceeding intake limit whenever possible.
                return dsym.dwarf.length > 1 ? this.thinDsym(dsym, output) : [dsym];
            })));
            return results.flat();
        });
    }
    /**
     * It takes fat dSYM as input and returns multiple dSYMs by extracting **each arch**
     * to separate dSYM file. New files are saved to `output` and named by their object uuid (`<uuid>.dSYM`).
     *
     * For example, given `<source path>/Foo.dSYM/Contents/Resources/DWARF/Foo` dSYM with two arch slices: `arm64` (uuid1)
     * and `x86_64` (uuid2), it will:
     * - create `<intermediate path>/<uuid1>.dSYM/Contents/Resources/DWARF/Foo` for `arm64`,
     * - create `<intermediate path>/<uuid2>.dSYM/Contents/Resources/DWARF/Foo` for `x86_64`.
     */
    thinDsym(dsym, output) {
        return __awaiter(this, void 0, void 0, function* () {
            const results = yield Promise.all(dsym.dwarf.map((dwarf) => __awaiter(this, void 0, void 0, function* () {
                try {
                    const bundle = (0, utils_1.buildPath)(output, `${dwarf.uuid}.dSYM`);
                    const object = (0, utils_1.buildPath)(bundle, upath_1.default.relative(dsym.bundle, dwarf.object));
                    yield fs_1.promises.mkdir(upath_1.default.dirname(object), { recursive: true });
                    try {
                        // Attempt to extract the single arch object file.
                        // We could use the `lipo -archs` command to get the list of archs,
                        // but it's more straightforward to just attempt to extract
                        // the single arch.
                        yield (0, utils_2.executeLipo)(dwarf.object, dwarf.arch, object);
                    }
                    catch (_a) {
                        // The thinning using `lipo` failed, meaning the dSYM is already
                        // a single arch dSYM. Copy the object file to the new dSYM.
                        yield fs_1.promises.copyFile(dwarf.object, object);
                    }
                    yield this.copyInfoPlist(dsym.bundle, bundle);
                    return [{ bundle, dwarf: [Object.assign(Object.assign({}, dwarf), { object })] }];
                }
                catch (error) {
                    this.context.stdout.write((0, renderer_1.renderDSYMSlimmingFailure)(dsym, dwarf, error));
                    return [];
                }
            })));
            return results.flat();
        });
    }
    copyInfoPlist(src, dst) {
        return __awaiter(this, void 0, void 0, function* () {
            const infoPlistPaths = (0, glob_1.globSync)((0, utils_1.buildPath)(src, '**/Info.plist'));
            if (infoPlistPaths.length === 0) {
                return;
            }
            const infoPlistPath = infoPlistPaths[0];
            const newInfoPlistPath = (0, utils_1.buildPath)(dst, upath_1.default.relative(src, infoPlistPath));
            yield fs_1.promises.mkdir(upath_1.default.dirname(newInfoPlistPath), { recursive: true });
            yield fs_1.promises.copyFile(infoPlistPath, newInfoPlistPath);
        });
    }
    compressDsyms(dsyms, output) {
        return __awaiter(this, void 0, void 0, function* () {
            yield fs_1.promises.mkdir(output, { recursive: true });
            return Promise.all(dsyms.map((dsym) => __awaiter(this, void 0, void 0, function* () {
                const archivePath = (0, utils_1.buildPath)(output, `${dsym.dwarf[0].uuid}.zip`);
                yield (0, utils_2.zipDirectoryToArchive)(dsym.bundle, archivePath);
                return new interfaces_1.CompressedDsym(archivePath, dsym);
            })));
        });
    }
    createRequestBuilder() {
        if (!this.config.apiKey) {
            throw new errors_1.InvalidConfigurationError(`Missing ${chalk_1.default.bold('DATADOG_API_KEY')} or ${chalk_1.default.bold('DD_API_KEY')} in your environment.`);
        }
        return (0, utils_1.getRequestBuilder)({
            apiKey: this.config.apiKey,
            baseUrl: (0, utils_2.getBaseIntakeUrl)(this.config.datadogSite),
            headers: new Map([
                ['DD-EVP-ORIGIN', 'datadog-ci_dsyms'],
                ['DD-EVP-ORIGIN-VERSION', this.cliVersion],
            ]),
            overrideUrl: 'api/v2/srcmap',
        });
    }
    createUploadFunction(requestBuilder, metricsLogger, apiKeyValidator) {
        return (dsym) => __awaiter(this, void 0, void 0, function* () {
            const payload = dsym.asMultipartPayload();
            if (this.dryRun) {
                this.context.stdout.write(`[DRYRUN] ${(0, renderer_1.renderUpload)(dsym)}`);
                return upload_1.UploadStatus.Success;
            }
            return (0, upload_1.upload)(requestBuilder)(payload, {
                apiKeyValidator,
                onError: (e) => {
                    this.context.stdout.write((0, renderer_1.renderFailedUpload)(dsym, e.message));
                    metricsLogger.logger.increment('failed', 1);
                },
                onRetry: (e, attempts) => {
                    this.context.stdout.write((0, renderer_1.renderRetriedUpload)(dsym, e.message, attempts));
                    metricsLogger.logger.increment('retries', 1);
                },
                onUpload: () => {
                    this.context.stdout.write((0, renderer_1.renderUpload)(dsym));
                },
                retries: 5,
            });
        });
    }
}
exports.DsymsUploadCommand = DsymsUploadCommand;
DsymsUploadCommand.paths = [['dsyms', 'upload']];
DsymsUploadCommand.usage = clipanion_1.Command.Usage({
    category: 'RUM',
    description: 'Upload dSYM files to Datadog.',
    details: `
      This command will upload all dSYM files to Datadog in order to symbolicate crash reports received by Datadog.\n
      See README for details.
    `,
    examples: [
        ['Upload all dSYM files in Derived Data path', 'datadog-ci dsyms upload ~/Library/Developer/Xcode/DerivedData'],
        [
            'Upload all dSYM files in a zip file (this is usually the case if your app has Bitcode enabled)',
            'datadog-ci dsyms upload /path/to/folder/my_file.zip',
        ],
    ],
});
