// @datadog/datadog-ci-base/dist/commands/measure/measure.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeasureCommand = exports.parseMeasures = void 0;
const chalk_1 = __importDefault(require("chalk"));
const clipanion_1 = require("clipanion");
const datadog_ci_base_1 = require("@datadog/datadog-ci-base");
const constants_1 = require("@datadog/datadog-ci-base/constants");
const api_1 = require("@datadog/datadog-ci-base/helpers/api");
const ci_1 = require("@datadog/datadog-ci-base/helpers/ci");
const env_1 = require("@datadog/datadog-ci-base/helpers/env");
const fips_1 = require("@datadog/datadog-ci-base/helpers/fips");
const retry_1 = require("@datadog/datadog-ci-base/helpers/retry");
const tags_1 = require("@datadog/datadog-ci-base/helpers/tags");
const utils_1 = require("@datadog/datadog-ci-base/helpers/utils");
const parseMeasures = (measures) => measures.reduce((acc, keyValue) => {
    if (!keyValue.includes(':')) {
        throw new Error(`invalid measures key value pair "${keyValue}"`);
    }
    const [key, value] = keyValue.split(':', 2);
    const floatVal = parseFloat(value);
    if (isNaN(floatVal)) {
        throw new Error('value is not numeric');
    }
    return Object.assign(Object.assign({}, acc), { [key]: floatVal });
}, {});
exports.parseMeasures = parseMeasures;
class MeasureCommand extends datadog_ci_base_1.BaseCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.level = clipanion_1.Option.String('--level');
        this.measures = clipanion_1.Option.Array('--measures');
        this.measuresFile = clipanion_1.Option.String('--measures-file');
        this.noFail = clipanion_1.Option.Boolean('--no-fail');
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false);
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.config = {
            apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            (0, fips_1.enableFips)(this.fips || this.config.fips, this.fipsIgnoreError || this.config.fipsIgnoreError);
            if (this.level !== 'pipeline' && this.level !== 'job') {
                this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} Level must be one of [pipeline, job]\n`);
                return 1;
            }
            const cliMeasures = this.measures;
            if (!cliMeasures && !this.measuresFile) {
                this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} --measures or --measures-file is required\n`);
                return 1;
            }
            const [measuresFromFile, valid] = (0, tags_1.parseMeasuresFile)(this.context, this.measuresFile);
            if (!valid) {
                // we should fail if attempted to read measures from a file and failed
                return 1;
            }
            const measures = Object.assign(Object.assign({}, (cliMeasures ? (0, exports.parseMeasures)(cliMeasures) : {})), measuresFromFile);
            if (Object.keys(measures).length === 0) {
                // This can happen for example if the measures file is provided but is empty
                this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} No measures found\n`);
                return 1;
            }
            try {
                const { provider, ciEnv } = (0, ci_1.getCIEnv)();
                if (this.level !== 'pipeline') {
                    const jobName = (0, ci_1.getGithubJobNameFromLogs)(this.context);
                    if (jobName) {
                        ciEnv[ci_1.envDDGithubJobName] = jobName;
                    }
                }
                const exitStatus = yield this.sendMeasures(ciEnv, this.level === 'pipeline' ? 0 : 1, provider, measures);
                if (exitStatus !== 0 && this.noFail) {
                    this.context.stderr.write(`${chalk_1.default.yellow.bold('[WARNING]')} sending measures failed but continuing due to --no-fail\n`);
                    return 0;
                }
                else if (exitStatus === 0 && !this.dryRun) {
                    this.context.stdout.write('Measures sent\n');
                }
                return exitStatus;
            }
            catch (error) {
                this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} ${error.message}\n`);
                return 1;
            }
        });
    }
    sendMeasures(ciEnv, level, provider, measures) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.config.apiKey) {
                this.context.stdout.write(`Neither ${chalk_1.default.red.bold('DATADOG_API_KEY')} nor ${chalk_1.default.red.bold('DD_API_KEY')} is in your environment.\n`);
                throw new Error('API key is missing');
            }
            const baseAPIURL = `https://${(0, utils_1.getApiHostForSite)((0, api_1.getDatadogSite)())}`;
            if (this.dryRun) {
                this.context.stdout.write(`[DRYRUN] Measure request: ${JSON.stringify(this.buildMeasureRequest(ciEnv, level, provider, measures), undefined, 2)}\n`);
                return 0;
            }
            const request = (0, utils_1.getRequestBuilder)({ baseUrl: baseAPIURL, apiKey: this.config.apiKey });
            const doRequest = () => request({
                data: this.buildMeasureRequest(ciEnv, level, provider, measures),
                method: 'post',
                url: 'api/v2/ci/pipeline/metrics',
            });
            try {
                yield (0, retry_1.retryRequest)(doRequest, {
                    onRetry: (e, attempt) => {
                        this.context.stderr.write(chalk_1.default.yellow(`[attempt ${attempt}] Could not send measures. Retrying...: ${e.message}\n`));
                    },
                    retries: 5,
                });
            }
            catch (error) {
                this.handleError(error);
                return 1;
            }
            return 0;
        });
    }
    handleError(error) {
        this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} Could not send measures: ` +
            `${error.response ? JSON.stringify(error.response.data, undefined, 2) : ''}\n`);
    }
    buildMeasureRequest(ciEnv, level, provider, measures) {
        return {
            data: {
                attributes: {
                    ci_env: ciEnv,
                    ci_level: level,
                    metrics: measures,
                    provider,
                },
                type: 'ci_custom_metric',
            },
        };
    }
}
exports.MeasureCommand = MeasureCommand;
MeasureCommand.paths = [['measure']];
MeasureCommand.usage = clipanion_1.Command.Usage({
    category: 'CI Visibility',
    description: 'Add measures to a CI Pipeline trace pipeline or job span in Datadog.',
    details: `
            This command when run from a supported CI provider sends an arbitrary set of key:value
            numeric tags to Datadog to include in the CI Visibility traces.
    `,
    examples: [
        ['Add a binary size to the current pipeline', 'datadog-ci measure --level pipeline --measures binary.size:500'],
        [
            'Tag the current CI job with a command runtime',
            'datadog-ci measure --level job --measures command.runtime:67.1',
        ],
        ['Add measures in bulk using a JSON file', 'datadog-ci measure --level job --measures-file my_measures.json'],
    ],
});
