// @datadog/datadog-ci-base/dist/commands/git-metadata/git.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseGitDiff = exports.getGitFileHash = exports.getGitDiff = exports.getMergeBase = exports.getCommitInfo = exports.gitTrackedFiles = exports.stripCredentials = exports.newSimpleGit = void 0;
const url_1 = require("url");
const simpleGit = __importStar(require("simple-git"));
const simple_git_1 = require("simple-git");
const ci_1 = require("../../helpers/ci");
const get_git_data_1 = require("../../helpers/git/get-git-data");
const interfaces_1 = require("./interfaces");
// Returns a configured SimpleGit.
const newSimpleGit = () => __awaiter(void 0, void 0, void 0, function* () {
    const currentDir = process.cwd();
    const options = {
        baseDir: currentDir,
        binary: 'git',
        maxConcurrentProcesses: 1,
    };
    const git = simpleGit.simpleGit(options);
    const isDocker = (yield import('is-docker')).default;
    if ((0, ci_1.isCI)() || isDocker()) {
        try {
            // In some CI envs repo may be checked out as a different user than the one running the command.
            // To be able to run git commands, we need to add the current directory as a safe directory.
            yield git.addConfig('safe.directory', currentDir, true, simple_git_1.GitConfigScope.global);
        }
        catch (e) {
            // Ignore the error
        }
    }
    // Attempt to set the baseDir to the root of the repository so the 'git ls-files' command
    // returns the tracked files paths relative to the root of the repository.
    const root = yield git.revparse('--show-toplevel');
    options.baseDir = root;
    return simpleGit.simpleGit(options);
});
exports.newSimpleGit = newSimpleGit;
// StripCredentials removes credentials from a remote HTTP url.
const stripCredentials = (remote) => {
    try {
        const url = new url_1.URL(remote);
        url.username = '';
        url.password = '';
        return url.toString();
    }
    catch (_a) {
        return remote;
    }
};
exports.stripCredentials = stripCredentials;
// Returns the hash of the current repository.
const gitHash = (git) => __awaiter(void 0, void 0, void 0, function* () { return git.revparse('HEAD'); });
// Returns the tracked files of the current repository.
const gitTrackedFiles = (git) => __awaiter(void 0, void 0, void 0, function* () {
    const files = yield git.raw('ls-files');
    return files.split(/\r\n|\r|\n/).filter((s) => s !== '');
});
exports.gitTrackedFiles = gitTrackedFiles;
// Returns the current hash, remote URL and tracked files paths.
const getCommitInfo = (git, repositoryURL) => __awaiter(void 0, void 0, void 0, function* () {
    // Invoke git commands to retrieve the remote, hash and tracked files.
    // We're using Promise.all instead of Promise.allSettled since we want to fail early if
    // any of the promises fails.
    let remote;
    let hash;
    let trackedFiles;
    if (repositoryURL) {
        ;
        [hash, trackedFiles] = yield Promise.all([gitHash(git), (0, exports.gitTrackedFiles)(git)]);
        remote = repositoryURL;
    }
    else {
        ;
        [remote, hash, trackedFiles] = yield Promise.all([(0, get_git_data_1.gitRemote)(git), gitHash(git), (0, exports.gitTrackedFiles)(git)]);
    }
    return new interfaces_1.CommitInfo(hash, remote, trackedFiles);
});
exports.getCommitInfo = getCommitInfo;
const getMergeBase = (git, target, source) => __awaiter(void 0, void 0, void 0, function* () {
    const result = yield git.raw(['merge-base', target, source]);
    return result.trim();
});
exports.getMergeBase = getMergeBase;
const getGitDiff = (git, from, to) => __awaiter(void 0, void 0, void 0, function* () {
    const fromResolved = yield git.revparse(from);
    const toResolved = yield git.revparse(to);
    const rawDiff = yield git.diff([
        '--unified=0',
        '--no-color',
        '--no-ext-diff',
        '--no-renames',
        '--diff-algorithm=minimal',
        `${fromResolved}..${toResolved}`,
    ]);
    return {
        head_sha: toResolved,
        base_sha: fromResolved,
        files: (0, exports.parseGitDiff)(rawDiff),
    };
});
exports.getGitDiff = getGitDiff;
const getGitFileHash = (git, path) => __awaiter(void 0, void 0, void 0, function* () {
    return git.revparse([`HEAD:${path}`]);
});
exports.getGitFileHash = getGitFileHash;
const diffHeaderRegex = /^diff --git a\/.+ b\/(.+)$/;
const hunkHeaderRegex = /^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@/;
const parseGitDiff = (diff) => {
    var _a, _b;
    const root = {};
    let currentPath;
    let currentLines;
    for (const line of diff.split(/\r?\n/)) {
        const diffHeader = diffHeaderRegex.exec(line);
        if (diffHeader) {
            if (currentPath && currentLines) {
                root[currentPath] = {
                    added_lines: base64Encode(currentLines.addedLines),
                    removed_lines: base64Encode(currentLines.removedLines),
                };
            }
            currentPath = diffHeader[1];
            currentLines = { addedLines: new Set(), removedLines: new Set() };
            continue;
        }
        if (!currentLines) {
            // still before first diff-header
            continue;
        }
        const hunkHeader = hunkHeaderRegex.exec(line);
        if (hunkHeader) {
            const oldStart = Number(hunkHeader[1]);
            const oldLen = Number((_a = hunkHeader[2]) !== null && _a !== void 0 ? _a : 1);
            const newStart = Number(hunkHeader[3]);
            const newLen = Number((_b = hunkHeader[4]) !== null && _b !== void 0 ? _b : 1);
            for (let i = 0; i < oldLen; i++) {
                currentLines.removedLines.add(oldStart + i);
            }
            for (let i = 0; i < newLen; i++) {
                currentLines.addedLines.add(newStart + i);
            }
        }
    }
    if (currentPath && currentLines) {
        root[currentPath] = {
            added_lines: base64Encode(currentLines.addedLines),
            removed_lines: base64Encode(currentLines.removedLines),
        };
    }
    return root;
};
exports.parseGitDiff = parseGitDiff;
const base64Encode = (set) => {
    const maxBit = set.size ? Math.max(...set) : 0;
    if (maxBit === 0) {
        return '';
    }
    const bytes = new Uint8Array(Math.ceil(maxBit / 8));
    for (const n of set) {
        // eslint-disable-next-line no-bitwise
        bytes[n >> 3] |= 1 << (n & 7);
    }
    return Buffer.from(bytes).toString('base64');
};
