// @datadog/datadog-ci-base/dist/commands/git-metadata/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommitInfo = void 0;
class CommitInfo {
    constructor(hash, remote, trackedFiles) {
        this.repositoryPayload = () => JSON.stringify({
            data: [
                {
                    files: this.trackedFiles,
                    hash: this.hash,
                    repository_url: this.remote,
                },
            ],
            // Make sure to update the version if the format of the JSON payloads changes in any way.
            version: 1,
        });
        this.hash = hash;
        this.remote = remote;
        this.trackedFiles = trackedFiles;
    }
    asMultipartPayload(cliVersion) {
        return {
            content: new Map([
                ['event', this.getMetadataPayload(cliVersion)],
                [
                    'repository',
                    {
                        type: 'string',
                        options: {
                            contentType: 'application/json',
                            filename: 'repository',
                        },
                        value: this.repositoryPayload(),
                    },
                ],
            ]),
        };
    }
    getMetadataPayload(cliVersion) {
        return {
            type: 'string',
            options: {
                contentType: 'application/json',
                filename: 'event',
            },
            value: JSON.stringify({
                cli_version: cliVersion,
                git_commit_sha: this.hash,
                git_repository_url: this.remote,
                type: 'repository',
            }),
        };
    }
}
exports.CommitInfo = CommitInfo;
