// @datadog/datadog-ci-base/dist/commands/git-metadata/renderer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderCommandInfo = exports.renderDryRunWarning = exports.renderSuccessfulCommand = exports.renderRetriedUpload = exports.renderFailedUpload = exports.renderConfigurationError = void 0;
const chalk_1 = __importDefault(require("chalk"));
const formatting_1 = require("../../helpers/formatting");
const renderConfigurationError = (error) => `${formatting_1.ICONS.FAILED} Configuration error: ${error}.`;
exports.renderConfigurationError = renderConfigurationError;
const renderFailedUpload = (errorMessage) => `${formatting_1.ICONS.FAILED} Failed upload: ${errorMessage}`;
exports.renderFailedUpload = renderFailedUpload;
const renderRetriedUpload = (errorMessage, attempt) => `[attempt ${attempt}] Retrying upload: ${errorMessage}`;
exports.renderRetriedUpload = renderRetriedUpload;
const renderSuccessfulCommand = (duration, dryRun) => {
    if (dryRun) {
        return chalk_1.default.green(`${formatting_1.ICONS.SUCCESS} [DRYRUN] Handled in ${duration} seconds.`);
    }
    else {
        return chalk_1.default.green(`${formatting_1.ICONS.SUCCESS} Uploaded in ${duration} seconds.`);
    }
};
exports.renderSuccessfulCommand = renderSuccessfulCommand;
const renderDryRunWarning = () => `${formatting_1.ICONS.WARNING} DRY-RUN MODE ENABLED. WILL NOT UPLOAD`;
exports.renderDryRunWarning = renderDryRunWarning;
const renderCommandInfo = (commit) => `Reporting commit ${commit.hash} from repository ${commit.remote}.
${commit.trackedFiles.length} tracked file paths will be reported.`;
exports.renderCommandInfo = renderCommandInfo;
