// @datadog/datadog-ci/dist/cli.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cli = exports.BETA_COMMANDS = void 0;
const cli_1 = require("@datadog/datadog-ci-base/cli");
const version_1 = require("@datadog/datadog-ci-base/version");
const clipanion_1 = require("clipanion");
const package_json_1 = require("@datadog/datadog-ci/package.json");
exports.BETA_COMMANDS = new Set(['deployment', 'elf-symbols']);
const betaCommandsEnabled = process.env.DD_BETA_COMMANDS_ENABLED === '1' || process.env.DD_BETA_COMMANDS_ENABLED === 'true';
const onError = (err) => {
    console.log(err);
    process.exitCode = 1;
};
process.on('uncaughtException', onError);
process.on('unhandledRejection', onError);
const cli = new clipanion_1.Cli({
    binaryLabel: 'Datadog CI',
    binaryName: 'datadog-ci',
    binaryVersion: version_1.cliVersion,
});
exports.cli = cli;
cli.register(clipanion_1.Builtins.HelpCommand);
cli.register(clipanion_1.Builtins.VersionCommand);
Object.entries(cli_1.commands).forEach(([scope, commands]) => {
    if (!betaCommandsEnabled && exports.BETA_COMMANDS.has(scope)) {
        return;
    }
    commands.forEach((command) => cli.register(command));
});
const builtinPlugins = Object.keys(package_json_1.dependencies).filter((plugin) => plugin.startsWith('@datadog/datadog-ci-plugin-'));
if (require.main === module) {
    void cli.runExit(process.argv.slice(2), {
        stderr: process.stderr,
        stdin: process.stdin,
        stdout: process.stdout,
        builtinPlugins,
    });
}
