// @datadog/datadog-ci-base/dist/commands/sourcemaps/upload.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourcemapsUploadCommand = void 0;
const fs_1 = __importDefault(require("fs"));
const url_1 = require("url");
const chalk_1 = __importDefault(require("chalk"));
const clipanion_1 = require("clipanion");
const upath_1 = __importDefault(require("upath"));
const datadog_ci_base_1 = require("@datadog/datadog-ci-base");
const constants_1 = require("@datadog/datadog-ci-base/constants");
const api_1 = require("@datadog/datadog-ci-base/helpers/api");
const apikey_1 = require("@datadog/datadog-ci-base/helpers/apikey");
const base_intake_url_1 = require("@datadog/datadog-ci-base/helpers/base-intake-url");
const concurrency_1 = require("@datadog/datadog-ci-base/helpers/concurrency");
const env_1 = require("@datadog/datadog-ci-base/helpers/env");
const errors_1 = require("@datadog/datadog-ci-base/helpers/errors");
const fips_1 = require("@datadog/datadog-ci-base/helpers/fips");
const format_git_sourcemaps_data_1 = require("@datadog/datadog-ci-base/helpers/git/format-git-sourcemaps-data");
const glob_1 = require("@datadog/datadog-ci-base/helpers/glob");
const metrics_1 = require("@datadog/datadog-ci-base/helpers/metrics");
const upload_1 = require("@datadog/datadog-ci-base/helpers/upload");
const utils_1 = require("@datadog/datadog-ci-base/helpers/utils");
const validation = __importStar(require("@datadog/datadog-ci-base/helpers/validation"));
const version_1 = require("@datadog/datadog-ci-base/version");
const interfaces_1 = require("./interfaces");
const renderer_1 = require("./renderer");
const utils_2 = require("./utils");
const validation_1 = require("./validation");
class SourcemapsUploadCommand extends datadog_ci_base_1.BaseCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.basePath = clipanion_1.Option.String({ required: true });
        this.disableGit = clipanion_1.Option.Boolean('--disable-git');
        this.quiet = clipanion_1.Option.Boolean('--quiet', false);
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false);
        this.maxConcurrency = clipanion_1.Option.String('--max-concurrency', '20', { validator: validation.isInteger() });
        this.minifiedPathPrefix = clipanion_1.Option.String('--minified-path-prefix');
        this.projectPath = clipanion_1.Option.String('--project-path');
        this.releaseVersion = clipanion_1.Option.String('--release-version');
        this.repositoryURL = clipanion_1.Option.String('--repository-url');
        this.commitSha = clipanion_1.Option.String('--commit-sha');
        this.service = clipanion_1.Option.String('--service');
        this.cliVersion = version_1.cliVersion;
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.config = {
            apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
            commitSha: process.env.DD_GIT_COMMIT_SHA,
            datadogSite: (0, api_1.getDatadogSite)(),
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
            repositoryURL: process.env.DD_GIT_REPOSITORY_URL,
        };
        // Fills the 'repository' field of each payload with data gathered using git or provided overrides.
        this.addRepositoryDataToPayloads = (payloads) => __awaiter(this, void 0, void 0, function* () {
            var _c, _d;
            const userProvidedRepoUrl = (_c = this.repositoryURL) !== null && _c !== void 0 ? _c : this.config.repositoryURL;
            const userProvidedCommitSha = (_d = this.commitSha) !== null && _d !== void 0 ? _d : this.config.commitSha;
            // If both user-provided repository URL and commit SHA are provided, use alternative approach without git.
            if (userProvidedRepoUrl && userProvidedCommitSha) {
                payloads.forEach((payload) => {
                    const repoPayload = this.getRepositoryPayloadWithoutGit(userProvidedCommitSha, userProvidedRepoUrl, payload.sourcemapPath);
                    payload.addRepositoryData({
                        gitCommitSha: userProvidedCommitSha,
                        gitRepositoryPayload: repoPayload,
                        gitRepositoryURL: userProvidedRepoUrl,
                    });
                });
                return;
            }
            try {
                const repositoryData = yield (0, format_git_sourcemaps_data_1.getRepositoryData)(yield (0, format_git_sourcemaps_data_1.newSimpleGit)(), userProvidedRepoUrl, userProvidedCommitSha);
                yield Promise.all(payloads.map((payload) => __awaiter(this, void 0, void 0, function* () {
                    const repositoryPayload = this.getRepositoryPayload(repositoryData, payload.sourcemapPath);
                    payload.addRepositoryData({
                        gitCommitSha: repositoryData.hash,
                        gitRepositoryPayload: repositoryPayload,
                        gitRepositoryURL: repositoryData.remote,
                    });
                })));
            }
            catch (e) {
                this.context.stdout.write((0, renderer_1.renderGitWarning)(e) + '\n');
            }
        });
        // Looks for the sourcemaps and minified files on disk and returns
        // the associated payloads.
        this.getMatchingSourcemapFiles = () => __awaiter(this, void 0, void 0, function* () {
            const jsFiles = (0, glob_1.globSync)((0, utils_1.buildPath)(this.basePath, '**/*.js'));
            const sourcemaps = (yield (0, concurrency_1.doWithMaxConcurrency)(this.maxConcurrency, jsFiles, (minifiedFilePath) => __awaiter(this, void 0, void 0, function* () {
                try {
                    const lastLine = yield (0, utils_2.readLastLine)(minifiedFilePath);
                    // Look for sourceMappingURL comment
                    const sourceMappingMatch = lastLine.match(/\/\/# sourceMappingURL=(.+\.map)/);
                    if (sourceMappingMatch) {
                        // mert: nextjs/turbopack uses url-percent encoding
                        const sourcemapUrl = decodeURIComponent(sourceMappingMatch[1].trim());
                        // Skip absolute URLs and absolute paths — they can't be resolved to local files
                        if (sourcemapUrl.includes('://') || upath_1.default.isAbsolute(sourcemapUrl)) {
                            return undefined;
                        }
                        // Join the sourcemap path relative to the minified file's directory
                        const minifiedFileDir = upath_1.default.dirname(minifiedFilePath);
                        const sourcemapPath = upath_1.default.join(minifiedFileDir, sourcemapUrl);
                        const [minifiedURL, relativePath] = this.getMinifiedURLAndRelativePath(minifiedFilePath);
                        return new interfaces_1.Sourcemap(minifiedFilePath, minifiedURL, sourcemapPath, relativePath, this.minifiedPathPrefix);
                    }
                }
                catch (error) {
                    return undefined;
                }
                return undefined;
            }))).filter((sourcemap) => sourcemap !== undefined);
            // Fall back to legacy method if no sourcemaps were found
            if (sourcemaps.length === 0) {
                return this.getLegacyMatchingSourcemapFiles();
            }
            return sourcemaps;
        });
        // Looks for the sourcemaps and minified files on disk and returns
        // the associated payloads.
        this.getLegacyMatchingSourcemapFiles = () => __awaiter(this, void 0, void 0, function* () {
            const sourcemapFiles = (0, glob_1.globSync)((0, utils_1.buildPath)(this.basePath, '**/*js.map'));
            return Promise.all(sourcemapFiles.map((sourcemapPath) => __awaiter(this, void 0, void 0, function* () {
                const minifiedFilePath = (0, utils_2.getMinifiedFilePath)(sourcemapPath);
                const [minifiedURL, relativePath] = this.getMinifiedURLAndRelativePath(minifiedFilePath);
                return new interfaces_1.Sourcemap(minifiedFilePath, minifiedURL, sourcemapPath, relativePath, this.minifiedPathPrefix);
            })));
        });
        this.getPayloadsToUpload = (useGit) => __awaiter(this, void 0, void 0, function* () {
            const payloads = yield this.getMatchingSourcemapFiles();
            if (!useGit) {
                return payloads;
            }
            yield this.addRepositoryDataToPayloads(payloads);
            return payloads;
        });
        // GetRepositoryPayload generates the repository payload for a specific sourcemap.
        // It specifically looks for the list of tracked files that are associated to the source paths
        // declared inside the sourcemap.
        this.getRepositoryPayload = (repositoryData, sourcemapPath) => {
            const onSourcesNotFound = () => {
                this.context.stdout.write((0, renderer_1.renderSourcesNotFoundWarning)(sourcemapPath));
            };
            let repositoryPayload;
            try {
                const files = repositoryData.trackedFilesMatcher.matchSourcemap(sourcemapPath, onSourcesNotFound);
                if (files) {
                    repositoryPayload = JSON.stringify({
                        data: [
                            {
                                files,
                                hash: repositoryData.hash,
                                repository_url: repositoryData.remote,
                            },
                        ],
                        // Make sure to update the version if the format of the JSON payloads changes in any way.
                        version: 1,
                    });
                }
                return repositoryPayload;
            }
            catch (error) {
                this.context.stdout.write((0, renderer_1.renderGitDataNotAttachedWarning)(sourcemapPath, error.message));
                return undefined;
            }
        };
        // Builds the repository payload for a sourcemap using its 'sources' field directly,
        // without relying on git. Used when both --repository-url and --commit-sha are provided.
        this.getRepositoryPayloadWithoutGit = (commitSha, repositoryURL, sourcemapPath) => {
            try {
                const buff = fs_1.default.readFileSync(sourcemapPath, 'utf8');
                const srcmapObj = JSON.parse(buff);
                if (!srcmapObj.sources || srcmapObj.sources.length === 0) {
                    return undefined;
                }
                const files = (0, format_git_sourcemaps_data_1.normalizeSources)(srcmapObj.sources, upath_1.default.dirname(sourcemapPath), process.cwd(), (source) => this.context.stdout.write((0, renderer_1.renderAbsolutePathWarning)(source)));
                if (files.length === 0) {
                    this.context.stdout.write((0, renderer_1.renderSourcesNotFoundWarning)(sourcemapPath));
                    return undefined;
                }
                return JSON.stringify({
                    data: [{ files, hash: commitSha, repository_url: repositoryURL }],
                    version: 1,
                });
            }
            catch (error) {
                this.context.stdout.write((0, renderer_1.renderGitDataNotAttachedWarning)(sourcemapPath, error.message));
                return undefined;
            }
        };
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            (0, fips_1.enableFips)(this.fips || this.config.fips, this.fipsIgnoreError || this.config.fipsIgnoreError);
            if (!this.releaseVersion) {
                this.context.stderr.write('Missing release version\n');
                return 1;
            }
            if (!this.service) {
                this.context.stderr.write('Missing service\n');
                return 1;
            }
            if (!this.minifiedPathPrefix) {
                this.context.stderr.write('Missing minified path prefix\n');
                return 1;
            }
            if (!this.isMinifiedPathPrefixValid()) {
                this.context.stdout.write(renderer_1.renderInvalidPrefix);
                return 1;
            }
            // Normalizing the basePath to resolve .. and .
            this.basePath = upath_1.default.normalize(this.basePath);
            this.context.stdout.write((0, renderer_1.renderCommandInfo)(this.basePath, this.minifiedPathPrefix, this.projectPath, this.releaseVersion, this.service, this.maxConcurrency, this.dryRun));
            const metricsLogger = (0, metrics_1.getMetricsLogger)({
                datadogSite: (0, api_1.getDatadogSiteFromEnv)(),
                defaultTags: [`version:${this.releaseVersion}`, `service:${this.service}`, `cli_version:${this.cliVersion}`],
                prefix: 'datadog.ci.sourcemaps.',
            });
            const apiKeyValidator = (0, apikey_1.newApiKeyValidator)({
                apiKey: this.config.apiKey,
                datadogSite: this.config.datadogSite,
                metricsLogger: metricsLogger.logger,
            });
            const useGit = this.disableGit === undefined || !this.disableGit;
            const initialTime = Date.now();
            const payloads = yield this.getPayloadsToUpload(useGit);
            const requestBuilder = this.getRequestBuilder();
            const uploadMultipart = this.upload(requestBuilder, metricsLogger, apiKeyValidator);
            try {
                const results = yield (0, concurrency_1.doWithMaxConcurrency)(this.maxConcurrency, payloads, uploadMultipart);
                const totalTime = (Date.now() - initialTime) / 1000;
                this.context.stdout.write((0, renderer_1.renderSuccessfulCommand)(results, totalTime, this.dryRun));
                metricsLogger.logger.gauge('duration', totalTime);
                return 0;
            }
            catch (error) {
                if (error instanceof errors_1.InvalidConfigurationError) {
                    this.context.stdout.write((0, renderer_1.renderConfigurationError)(error));
                    return 1;
                }
                // Otherwise unknown error, let's propagate the exception
                throw error;
            }
            finally {
                try {
                    yield metricsLogger.flush();
                }
                catch (err) {
                    this.context.stdout.write(`WARN: ${err}\n`);
                }
            }
        });
    }
    getMinifiedURLAndRelativePath(minifiedFilePath) {
        const relativePath = minifiedFilePath.replace(this.basePath, '');
        return [(0, utils_1.buildPath)(this.minifiedPathPrefix, relativePath), relativePath];
    }
    getRequestBuilder() {
        if (!this.config.apiKey) {
            throw new errors_1.InvalidConfigurationError(`Missing ${chalk_1.default.bold('DATADOG_API_KEY')} or ${chalk_1.default.bold('DD_API_KEY')} in your environment.`);
        }
        return (0, utils_1.getRequestBuilder)({
            apiKey: this.config.apiKey,
            baseUrl: (0, base_intake_url_1.getBaseSourcemapIntakeUrl)(this.config.datadogSite),
            headers: new Map([
                ['DD-EVP-ORIGIN', 'datadog-ci_sourcemaps'],
                ['DD-EVP-ORIGIN-VERSION', this.cliVersion],
            ]),
            overrideUrl: 'api/v2/srcmap',
        });
    }
    isMinifiedPathPrefixValid() {
        let host;
        try {
            const objUrl = new url_1.URL(this.minifiedPathPrefix);
            host = objUrl.host;
        }
        catch (_a) {
            // Do nothing.
        }
        if (!host && !this.minifiedPathPrefix.startsWith('/')) {
            return false;
        }
        return true;
    }
    upload(requestBuilder, metricsLogger, apiKeyValidator) {
        return (sourcemap) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            try {
                (0, validation_1.validatePayload)(sourcemap, this.context.stdout);
            }
            catch (error) {
                if (error instanceof validation_1.InvalidPayload) {
                    this.context.stdout.write((0, renderer_1.renderFailedUpload)(sourcemap, error.message));
                    metricsLogger.logger.increment('skipped_sourcemap', 1, [`reason:${error.reason}`]);
                }
                else {
                    this.context.stdout.write((0, renderer_1.renderFailedUpload)(sourcemap, `Skipping sourcemap ${sourcemap.sourcemapPath} because of error: ${error.message}`));
                    metricsLogger.logger.increment('skipped_sourcemap', 1, ['reason:unknown']);
                }
                return upload_1.UploadStatus.Skipped;
            }
            const payload = sourcemap.asMultipartPayload(this.cliVersion, this.service, this.releaseVersion, (_a = this.projectPath) !== null && _a !== void 0 ? _a : '');
            if (this.dryRun) {
                this.context.stdout.write(`[DRYRUN] ${(0, renderer_1.renderUpload)(sourcemap)}`);
                return upload_1.UploadStatus.Success;
            }
            return (0, upload_1.upload)(requestBuilder)(payload, {
                apiKeyValidator,
                onError: (e) => {
                    this.context.stdout.write((0, renderer_1.renderFailedUpload)(sourcemap, e.message));
                    metricsLogger.logger.increment('failed', 1);
                },
                onRetry: (e, attempts) => {
                    this.context.stdout.write((0, renderer_1.renderRetriedUpload)(sourcemap, e.message, attempts));
                    metricsLogger.logger.increment('retries', 1);
                },
                onUpload: () => {
                    if (this.quiet) {
                        return;
                    }
                    this.context.stdout.write((0, renderer_1.renderUpload)(sourcemap));
                },
                retries: 5,
                useGzip: true,
            });
        });
    }
}
exports.SourcemapsUploadCommand = SourcemapsUploadCommand;
SourcemapsUploadCommand.paths = [['sourcemaps', 'upload']];
SourcemapsUploadCommand.usage = clipanion_1.Command.Usage({
    category: 'RUM',
    description: 'Upload JavaScript sourcemaps to Datadog.',
    details: `
      This command will upload all JavaScript sourcemaps and their corresponding JavaScript file to Datadog in order to un-minify front-end stack traces received by Datadog.\n
      See README for details.
    `,
    examples: [
        [
            'Upload all sourcemaps in current directory',
            'datadog-ci sourcemaps upload . --service my-service --minified-path-prefix https://static.datadog.com --release-version 1.234',
        ],
        [
            'Upload all sourcemaps in /home/users/ci with 50 concurrent uploads',
            'datadog-ci sourcemaps upload /home/users/ci --service my-service --minified-path-prefix https://static.datadog.com --release-version 1.234 --max-concurrency 50',
        ],
    ],
});
