// @datadog/datadog-ci-base/dist/commands/ci-env/read.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CiEnvReadCommand = void 0;
const clipanion_1 = require("clipanion");
const t = __importStar(require("typanion"));
const datadog_ci_base_1 = require("@datadog/datadog-ci-base");
const ci_1 = require("../../helpers/ci");
const tags_1 = require("../../helpers/tags");
const user_provided_git_1 = require("../../helpers/user-provided-git");
/**
 * Explicit mapping from internal tag keys to DD_* environment variable names.
 */
const TAG_TO_ENV_VAR = {
    // CI tags
    [tags_1.CI_JOB_ID]: 'DD_CI_JOB_ID',
    [tags_1.CI_JOB_NAME]: 'DD_CI_JOB_NAME',
    [tags_1.CI_JOB_URL]: 'DD_CI_JOB_URL',
    [tags_1.CI_NODE_LABELS]: 'DD_CI_NODE_LABELS',
    [tags_1.CI_NODE_NAME]: 'DD_CI_NODE_NAME',
    [tags_1.CI_PIPELINE_ID]: 'DD_CI_PIPELINE_ID',
    [tags_1.CI_PIPELINE_NAME]: 'DD_CI_PIPELINE_NAME',
    [tags_1.CI_PIPELINE_NUMBER]: 'DD_CI_PIPELINE_NUMBER',
    [tags_1.CI_PIPELINE_URL]: 'DD_CI_PIPELINE_URL',
    [tags_1.CI_PROVIDER_NAME]: 'DD_CI_PROVIDER_NAME',
    [tags_1.CI_STAGE_NAME]: 'DD_CI_STAGE_NAME',
    [tags_1.CI_WORKSPACE_PATH]: 'DD_CI_WORKSPACE_PATH',
    [tags_1.CI_ENV_VARS]: 'DD_CI_ENV_VARS',
    // Git tags
    [tags_1.GIT_BRANCH]: 'DD_GIT_BRANCH',
    [tags_1.GIT_COMMIT_AUTHOR_DATE]: 'DD_GIT_COMMIT_AUTHOR_DATE',
    [tags_1.GIT_COMMIT_AUTHOR_EMAIL]: 'DD_GIT_COMMIT_AUTHOR_EMAIL',
    [tags_1.GIT_COMMIT_AUTHOR_NAME]: 'DD_GIT_COMMIT_AUTHOR_NAME',
    [tags_1.GIT_COMMIT_COMMITTER_DATE]: 'DD_GIT_COMMIT_COMMITTER_DATE',
    [tags_1.GIT_COMMIT_COMMITTER_EMAIL]: 'DD_GIT_COMMIT_COMMITTER_EMAIL',
    [tags_1.GIT_COMMIT_COMMITTER_NAME]: 'DD_GIT_COMMIT_COMMITTER_NAME',
    [tags_1.GIT_COMMIT_MESSAGE]: 'DD_GIT_COMMIT_MESSAGE',
    [tags_1.GIT_HEAD_SHA]: 'DD_GIT_COMMIT_HEAD_SHA',
    [tags_1.GIT_PULL_REQUEST_BASE_BRANCH]: 'DD_GIT_PULL_REQUEST_BASE_BRANCH',
    [tags_1.GIT_PULL_REQUEST_BASE_BRANCH_HEAD_SHA]: 'DD_GIT_PULL_REQUEST_BASE_BRANCH_HEAD_SHA',
    [tags_1.GIT_PULL_REQUEST_BASE_BRANCH_SHA]: 'DD_GIT_PULL_REQUEST_BASE_BRANCH_SHA',
    [tags_1.GIT_REPOSITORY_URL]: 'DD_GIT_REPOSITORY_URL',
    [tags_1.GIT_SHA]: 'DD_GIT_COMMIT_SHA',
    [tags_1.GIT_TAG]: 'DD_GIT_TAG',
    // PR tags
    [tags_1.PR_NUMBER]: 'DD_PR_NUMBER',
};
/**
 * Converts a tag key to an environment variable name using explicit mapping.
 * Falls back to the tag key itself if no explicit mapping exists.
 */
const tagKeyToEnvVar = (tagKey) => {
    return TAG_TO_ENV_VAR[tagKey] || tagKey;
};
/**
 * Escapes a value for use in bash single-quoted strings.
 * Uses single quotes to avoid variable expansion and command injection.
 * Inside single quotes, nothing is interpreted by bash ($VAR, $(cmd), etc. stay literal).
 */
const escapeBashValue = (value) => {
    // Replace single quotes with '\'' which ends the quote, adds an escaped quote, and starts a new quote
    return value.replace(/'/g, "'\\''");
};
/**
 * Escapes a value for use in double-quoted tag values.
 */
const escapeTagValue = (value) => {
    // Escape backslashes first, then double quotes
    return value.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
};
/**
 *  Output for different formats.
 */
const formatOutput = (tags, format) => {
    const entries = Object.entries(tags).map(([key, value]) => {
        const envVarName = tagKeyToEnvVar(key);
        return { key, envVarName, value };
    });
    if (format === 'tags') {
        return entries
            .map(({ key, value }) => {
            const escapedValue = escapeTagValue(value);
            return `${key}:"${escapedValue}"`;
        })
            .join('\n');
    }
    if (format === 'json') {
        const jsonObj = {};
        for (const { envVarName, value } of entries) {
            jsonObj[envVarName] = value;
        }
        return JSON.stringify(jsonObj, undefined, 2);
    }
    return entries
        .map(({ envVarName, value }) => {
        const escapedValue = escapeBashValue(value);
        return `${envVarName}='${escapedValue}'`;
    })
        .join('\n');
};
class CiEnvReadCommand extends datadog_ci_base_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.format = clipanion_1.Option.String('--format', 'bash', {
            description: 'Output format: bash (default), json, or tags',
            validator: t.isEnum(['bash', 'json', 'tags']),
        });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            // Merge CI-detected tags with user-provided tags (user-provided takes precedence)
            const ciTags = (0, ci_1.getCISpanTags)();
            const userCITags = (0, user_provided_git_1.getUserCISpanTags)();
            const userGitTags = (0, user_provided_git_1.getUserGitSpanTags)();
            const tags = Object.assign(Object.assign(Object.assign({}, ciTags), userCITags), userGitTags);
            if (!tags || Object.keys(tags).length === 0) {
                this.context.stderr.write('Warning: No CI environment detected or no CI tags found. ' +
                    'This command should be run in a CI environment.\n');
                return 1;
            }
            const output = formatOutput(tags, this.format);
            this.context.stdout.write(output + '\n');
            return 0;
        });
    }
}
exports.CiEnvReadCommand = CiEnvReadCommand;
CiEnvReadCommand.paths = [['ci-env', 'read']];
CiEnvReadCommand.usage = clipanion_1.Command.Usage({
    category: 'CI Visibility',
    description: 'Extract and normalize CI environment variables into `DD_CI_*` and `DD_GIT_*` format.',
    details: `
      This command reads CI-specific environment variables from various CI providers
      and outputs them as normalized \`DD_CI_*\` and \`DD_GIT_*\` environment variables.
      Supports bash (default), json, and tags output formats.
    `,
    examples: [
        ['Set variables in current shell', 'eval "$(datadog-ci ci-env read)"'],
        ['Save as .env file for Docker', 'datadog-ci ci-env read > .env'],
        ['Use with shell scripts', 'datadog-ci ci-env read > ci.env && source ci.env'],
        ['Get JSON output', 'datadog-ci ci-env read --format json'],
        ['Get Datadog tags format', 'datadog-ci ci-env read --format tags'],
    ],
});
