// @datadog/datadog-ci-base/dist/commands/pe-symbols/breakpad.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBreakpadSymMetadata = void 0;
const fs_1 = __importDefault(require("fs"));
const readline_1 = __importDefault(require("readline"));
const pe_constants_1 = require("./pe-constants");
const cpuToArchitecture = (cpu) => {
    const normalized = cpu.toLowerCase();
    if (normalized === 'x86') {
        return pe_constants_1.MachineArchitecture.x86;
    }
    if (normalized === 'x86_64' || normalized === 'amd64' || normalized === 'x64') {
        return pe_constants_1.MachineArchitecture.x64;
    }
    if (normalized === 'arm64') {
        return pe_constants_1.MachineArchitecture.Arm64;
    }
    if (normalized === 'arm' || normalized === 'armv7' || normalized === 'arm32') {
        return pe_constants_1.MachineArchitecture.Arm32;
    }
    return pe_constants_1.MachineArchitecture.unknown;
};
const formatGuid = (guidHex) => `${guidHex.slice(0, 8)}-${guidHex.slice(8, 12)}-${guidHex.slice(12, 16)}-${guidHex.slice(16, 20)}-${guidHex.slice(20)}`;
const parseModuleHeader = (line) => {
    const trimmed = line.trim();
    if (!trimmed.length) {
        return undefined;
    }
    if (!trimmed.startsWith('MODULE ')) {
        return undefined;
    }
    const tokens = trimmed.split(/\s+/);
    if (tokens.length < 5) {
        throw new Error('Invalid MODULE header in Breakpad symbol file');
    }
    const [, os, cpu, id, ...rest] = tokens;
    const name = rest.join(' ');
    return { os, cpu, id, name };
};
const isAscii = (line) => {
    for (let i = 0; i < line.length; i++) {
        if (line.charCodeAt(i) > 127) {
            return false;
        }
    }
    return true;
};
const analyzeBreakpadFile = (pathname) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, e_1, _b, _c;
    // heuristic to figure out if we have debug info or not
    const stream = fs_1.default.createReadStream(pathname, { encoding: 'utf8' });
    const rl = readline_1.default.createInterface({ input: stream, crlfDelay: Infinity });
    let header;
    let hasFileRecords = false;
    try {
        try {
            for (var _d = true, rl_1 = __asyncValues(rl), rl_1_1; rl_1_1 = yield rl_1.next(), _a = rl_1_1.done, !_a; _d = true) {
                _c = rl_1_1.value;
                _d = false;
                const line = _c;
                if (!header) {
                    const headerLine = line.trim();
                    if (!headerLine) {
                        continue;
                    }
                    if (!isAscii(line)) {
                        throw new Error('Unsupported symbol file: Breakpad .sym files must be ASCII encoded');
                    }
                    if (!headerLine.startsWith('MODULE ')) {
                        throw new Error('Unsupported symbol file: first non-empty line must be a Breakpad MODULE header');
                    }
                    const parsed = parseModuleHeader(line);
                    if (parsed) {
                        header = parsed;
                    }
                    continue;
                }
                const dataLine = line.trim();
                if (!hasFileRecords && dataLine.startsWith('FILE ')) {
                    hasFileRecords = true;
                    break;
                }
                // we hit the functions, so we won't find any more file records
                if (!hasFileRecords && (dataLine.startsWith('FUNC ') || dataLine.startsWith('PUBLIC '))) {
                    break;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (!_d && !_a && (_b = rl_1.return)) yield _b.call(rl_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    finally {
        rl.close();
        stream.close();
    }
    if (!header) {
        throw new Error('Breakpad symbol file is missing MODULE header');
    }
    return { header, hasFileRecords };
});
const getBreakpadSymMetadata = (pathname) => __awaiter(void 0, void 0, void 0, function* () {
    const { header, hasFileRecords } = yield analyzeBreakpadFile(pathname);
    const identifier = header.id.toUpperCase();
    if (identifier.length <= 32) {
        throw new Error('Breakpad MODULE identifier is malformed');
    }
    if (!/^[0-9A-F]+$/.test(identifier)) {
        throw new Error('Breakpad MODULE identifier must be hexadecimal');
    }
    const guidHex = identifier.slice(0, 32);
    const ageHex = identifier.slice(32);
    if (guidHex.length !== 32 || ageHex.length === 0) {
        throw new Error('Breakpad MODULE identifier must contain GUID and age');
    }
    const pdbAge = Number.parseInt(ageHex, 16);
    if (Number.isNaN(pdbAge)) {
        throw new Error('Breakpad MODULE age is not hexadecimal');
    }
    const metadata = {
        filename: pathname,
        isPE: false,
        arch: cpuToArchitecture(header.cpu),
        hasPdbInfo: true,
        pdbAge,
        pdbSig: formatGuid(guidHex),
        pdbFilename: header.name,
        sourceType: 'breakpad_sym',
        symbolPath: pathname,
        symbolSource: hasFileRecords ? 'debug_info' : 'symbol_table',
        moduleOs: header.os,
    };
    return metadata;
});
exports.getBreakpadSymMetadata = getBreakpadSymMetadata;
