// @datadog/datadog-ci-base/dist/helpers/serverless/common.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInstrumentedTemplate = exports.byName = exports.generateConfigDiff = exports.sortedEqual = exports.getBaseEnvVars = exports.collectAsyncIterator = exports.parseEnvVars = void 0;
const node_util_1 = require("node:util");
const chalk_1 = __importDefault(require("chalk"));
const jest_diff_1 = require("jest-diff");
const constants_1 = require("../../constants");
const api_1 = require("../api");
const constants_2 = require("./constants");
/**
 * Parses environment variables from array format (KEY=VALUE) to object format.
 * @param envVars - Array of environment variables in KEY=VALUE format
 * @returns Object with parsed environment variables
 */
const parseEnvVars = (envVars) => {
    const result = {};
    envVars === null || envVars === void 0 ? void 0 : envVars.forEach((e) => {
        const match = e.match(constants_2.ENV_VAR_REGEX);
        if (match) {
            const [, key, value] = match;
            result[key] = value;
        }
    });
    return result;
};
exports.parseEnvVars = parseEnvVars;
/**
 * Collects all items from a paged async iterator into an array.
 * @param it - Paged async iterator
 * @returns Array of all items
 */
const collectAsyncIterator = (it) => { var _a, it_1, it_1_1; return __awaiter(void 0, void 0, void 0, function* () {
    var _b, e_1, _c, _d;
    const arr = [];
    try {
        for (_a = true, it_1 = __asyncValues(it); it_1_1 = yield it_1.next(), _b = it_1_1.done, !_b; _a = true) {
            _d = it_1_1.value;
            _a = false;
            const x = _d;
            arr.push(x);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (!_a && !_b && (_c = it_1.return)) yield _c.call(it_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return arr;
}); };
exports.collectAsyncIterator = collectAsyncIterator;
/**
 * Builds base environment variables for serverless instrumentation.
 * @param config - Configuration options
 * @returns Base environment variables object
 */
const getBaseEnvVars = (config) => {
    const envVars = Object.assign({ DD_API_KEY: process.env.DD_API_KEY, DD_SITE: (0, api_1.getDatadogSite)(), DD_SERVICE: config.service }, (0, exports.parseEnvVars)(config.envVars));
    if (config.environment) {
        envVars.DD_ENV = config.environment;
    }
    if (config.version) {
        envVars.DD_VERSION = config.version;
    }
    if (config.logPath) {
        envVars[constants_2.LOGS_PATH_ENV_VAR] = config.logPath;
    }
    if (config.extraTags) {
        envVars.DD_TAGS = config.extraTags;
    }
    return envVars;
};
exports.getBaseEnvVars = getBaseEnvVars;
/**
 * Recursively sort object keys to ensure consistent ordering
 */
const sortObject = (obj) => {
    if (!obj) {
        return obj;
    }
    if (Array.isArray(obj)) {
        return obj.map(sortObject).sort((a, b) => {
            return JSON.stringify(a).localeCompare(JSON.stringify(b));
        });
    }
    if (typeof obj === 'object') {
        const sorted = {};
        Object.keys(obj)
            .sort()
            .forEach((key) => {
            sorted[key] = sortObject(obj[key]);
        });
        return sorted;
    }
    return obj;
};
const sortedEqual = (a, b) => {
    const sortedA = sortObject(a);
    const sortedB = sortObject(b);
    return (0, node_util_1.isDeepStrictEqual)(sortedA, sortedB);
};
exports.sortedEqual = sortedEqual;
/**
 * Obfuscate sensitive values in a line if it contains a key-like pattern
 */
const obfuscateSensitiveValues = (line) => {
    // Match hex strings of 16, 32, or 64 characters (common API key/token lengths)
    return line
        .replace(/("[0-9a-fA-F]{16}"|"[0-9a-fA-F]{32}"|"[0-9a-fA-F]{64}")/g, '"***"')
        .replace(/('[0-9a-fA-F]{16}'|'[0-9a-fA-F]{32}'|'[0-9a-fA-F]{64}')/g, "'***'");
};
/**
 * Generate a git diff-style comparison between two configurations
 * @param original The original configuration object
 * @param updated The updated configuration object
 * @returns A formatted diff string with colors
 */
const generateConfigDiff = (original, updated) => {
    // Sort keys consistently before comparison
    const sortedOriginal = sortObject(original);
    const sortedUpdated = sortObject(updated);
    const originalJson = JSON.stringify(sortedOriginal, undefined, 2);
    const updatedJson = JSON.stringify(sortedUpdated, undefined, 2);
    const configDiff = (0, jest_diff_1.diff)(originalJson, updatedJson, {
        aColor: chalk_1.default.red,
        bColor: chalk_1.default.green,
        omitAnnotationLines: true,
        expand: false,
    });
    if (!configDiff || configDiff.includes('no visual difference')) {
        return chalk_1.default.gray('No changes detected.');
    }
    return configDiff.split('\n').map(obfuscateSensitiveValues).join('\n');
};
exports.generateConfigDiff = generateConfigDiff;
const byName = (xs) => {
    return Object.fromEntries(xs.filter((x) => x.name).map((x) => [x.name, x]));
};
exports.byName = byName;
const DEFAULT_ENV_VARS_BY_NAME = (0, exports.byName)([
    { name: constants_2.SITE_ENV_VAR, value: constants_1.DATADOG_SITE_US1 },
    { name: constants_2.LOGS_INJECTION_ENV_VAR, value: 'true' },
    { name: constants_2.DD_TRACE_ENABLED_ENV_VAR, value: 'true' },
    { name: constants_2.HEALTH_PORT_ENV_VAR, value: constants_2.DEFAULT_HEALTH_CHECK_PORT.toString() },
]);
/**
 * Given the configuration, an app template, the base sidecar configuration, and base shared volume,
 */
const createInstrumentedTemplate = (template, baseSidecar, sharedVolumeOptions, envVarsByName) => {
    var _a;
    const sharedVolumeMount = {
        [sharedVolumeOptions.volumeMountNameKey]: sharedVolumeOptions.name,
        mountPath: sharedVolumeOptions.mountPath,
    };
    const containers = template.containers || [];
    const hasSidecarContainer = containers.some((c) => c.name === baseSidecar.name);
    const newSidecarContainer = Object.assign(Object.assign({}, baseSidecar), { env: Object.values(Object.assign(Object.assign(Object.assign({}, DEFAULT_ENV_VARS_BY_NAME), (0, exports.byName)((_a = baseSidecar.env) !== null && _a !== void 0 ? _a : [])), envVarsByName)), volumeMounts: [sharedVolumeMount] });
    // Update all app containers to add volume mounts and env vars if they don't have them
    const updatedContainers = containers.map((container) => {
        var _a;
        if (container.name === baseSidecar.name) {
            return newSidecarContainer;
        }
        const existingVolumeMounts = container.volumeMounts || [];
        const hasSharedVolumeMount = existingVolumeMounts.some((mount) => mount[sharedVolumeOptions.volumeMountNameKey] === sharedVolumeOptions.name);
        const updatedVolumeMounts = existingVolumeMounts.map((mount) => mount[sharedVolumeOptions.volumeMountNameKey] === sharedVolumeOptions.name ? sharedVolumeMount : mount);
        if (!hasSharedVolumeMount) {
            updatedVolumeMounts.push(sharedVolumeMount);
        }
        return Object.assign(Object.assign({}, container), { volumeMounts: updatedVolumeMounts, env: Object.values(Object.assign(Object.assign(Object.assign({}, DEFAULT_ENV_VARS_BY_NAME), (0, exports.byName)((_a = container.env) !== null && _a !== void 0 ? _a : [])), envVarsByName)) });
    });
    // Add sidecar if it doesn't exist
    if (!hasSidecarContainer) {
        updatedContainers.push(newSidecarContainer);
    }
    const volumes = template.volumes || [];
    // Add shared volume if it doesn't exist
    const sharedVolume = Object.assign(Object.assign({}, sharedVolumeOptions.mountOptions), { name: sharedVolumeOptions.name });
    const hasSharedVolume = volumes.some((volume) => volume.name === sharedVolumeOptions.name);
    const updatedVolumes = volumes.map((volume) => (volume.name === sharedVolumeOptions.name ? sharedVolume : volume));
    if (!hasSharedVolume) {
        updatedVolumes.push(sharedVolume);
    }
    return Object.assign(Object.assign({}, template), { containers: updatedContainers, volumes: updatedVolumes });
};
exports.createInstrumentedTemplate = createInstrumentedTemplate;
