// @datadog/datadog-ci-base/dist/helpers/serverless/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENV_VAR_REGEX = exports.DEFAULT_HEALTH_CHECK_PORT = exports.DEFAULT_LOGS_PATH = exports.DEFAULT_VOLUME_PATH = exports.DEFAULT_VOLUME_NAME = exports.DEFAULT_SIDECAR_NAME = exports.SIDECAR_PORT = exports.SIDECAR_IMAGE = exports.SIDECAR_CONTAINER_NAME = exports.FLARE_PROJECT_FILES = exports.FLARE_ENDPOINT_PATH = exports.INSIGHTS_FILE_NAME = exports.ADDITIONAL_FILES_DIRECTORY = exports.PROJECT_FILES_DIRECTORY = exports.LOGS_DIRECTORY = exports.FLARE_OUTPUT_DIRECTORY = exports.EXTRA_TAGS_REG_EXP = exports.VERSION_ENV_VAR = exports.ENVIRONMENT_ENV_VAR = exports.SERVICE_ENV_VAR = exports.DD_SOURCE_ENV_VAR = exports.DD_TAGS_ENV_VAR = exports.DD_LLMOBS_AGENTLESS_ENABLED_ENV_VAR = exports.DD_LLMOBS_ML_APP_ENV_VAR = exports.DD_LLMOBS_ENABLED_ENV_VAR = exports.DD_TRACE_ENABLED_ENV_VAR = exports.DD_LOG_LEVEL_ENV_VAR = exports.HEALTH_PORT_ENV_VAR = exports.LOGS_PATH_ENV_VAR = exports.LOGS_INJECTION_ENV_VAR = exports.SITE_ENV_VAR = exports.CI_SITE_ENV_VAR = exports.CI_API_KEY_ENV_VAR = exports.API_KEY_ENV_VAR = void 0;
// Environment variables for Lambda and Cloud Run
exports.API_KEY_ENV_VAR = 'DD_API_KEY';
exports.CI_API_KEY_ENV_VAR = 'DATADOG_API_KEY';
exports.CI_SITE_ENV_VAR = 'DATADOG_SITE';
exports.SITE_ENV_VAR = 'DD_SITE';
exports.LOGS_INJECTION_ENV_VAR = 'DD_LOGS_INJECTION';
exports.LOGS_PATH_ENV_VAR = 'DD_SERVERLESS_LOG_PATH';
exports.HEALTH_PORT_ENV_VAR = 'DD_HEALTH_PORT';
exports.DD_LOG_LEVEL_ENV_VAR = 'DD_LOG_LEVEL';
exports.DD_TRACE_ENABLED_ENV_VAR = 'DD_TRACE_ENABLED';
exports.DD_LLMOBS_ENABLED_ENV_VAR = 'DD_LLMOBS_ENABLED';
exports.DD_LLMOBS_ML_APP_ENV_VAR = 'DD_LLMOBS_ML_APP';
exports.DD_LLMOBS_AGENTLESS_ENABLED_ENV_VAR = 'DD_LLMOBS_AGENTLESS_ENABLED';
exports.DD_TAGS_ENV_VAR = 'DD_TAGS';
exports.DD_SOURCE_ENV_VAR = 'DD_SOURCE'; // Tagging env vars
exports.SERVICE_ENV_VAR = 'DD_SERVICE';
exports.ENVIRONMENT_ENV_VAR = 'DD_ENV';
exports.VERSION_ENV_VAR = 'DD_VERSION';
/*
 * DD_TAGS Regular Expression
 * This RegExp ensures that the --extra-tags string
 * matches a list of <key>:<value> separated by commas
 * such as layer:api,team:intake
 */
exports.EXTRA_TAGS_REG_EXP = /^(([a-zA-Z]+)[\w\-/.]*:[^,]+)+((,)([a-zA-Z]+)[\w\-/.]*:[^,]+)*$/g;
// Flare constants
exports.FLARE_OUTPUT_DIRECTORY = '.datadog-ci';
exports.LOGS_DIRECTORY = 'logs';
exports.PROJECT_FILES_DIRECTORY = 'project_files';
exports.ADDITIONAL_FILES_DIRECTORY = 'additional_files';
exports.INSIGHTS_FILE_NAME = 'INSIGHTS.md';
exports.FLARE_ENDPOINT_PATH = '/api/ui/support/serverless/flare';
// Project files to search for in Flare
exports.FLARE_PROJECT_FILES = [
    // Datadog CloudFormation Template
    'datadog-cloudfomation-macro.yaml',
    'datadog-cloudfomation-macro.yml',
    'datadog-cloudformation-macro.json',
    // Node.js
    'package.json',
    'package-lock.json',
    'yarn.lock',
    '.nvmrc',
    // Python
    'requirements.txt',
    'Pipfile',
    'Pipfile.lock',
    'pyproject.toml',
    // Java
    'pom.xml',
    'build.gradle',
    'gradlew',
    'gradlew.bat',
    // Go
    'Makefile',
    'go.mod',
    'go.sum',
    'Gopkg.toml',
    'gomod.sh',
    // Ruby
    'Gemfile',
    'Gemfile.lock',
    // .NET
    'project.json',
    'packages.config',
    'PackageReference',
    'global.json',
    // Docker
    'Dockerfile',
    'docker-compose.yaml',
    'docker-compose.yml',
    // Webpack, bundlers
    'webpack.config.js',
    '.babelrc',
    'tsconfig.json',
    'esbuild.config.js',
];
/**
 * Shared constants for serverless instrumentation
 */
exports.SIDECAR_CONTAINER_NAME = 'datadog-sidecar';
exports.SIDECAR_IMAGE = 'index.docker.io/datadog/serverless-init:latest';
exports.SIDECAR_PORT = 8126;
exports.DEFAULT_SIDECAR_NAME = 'datadog-sidecar';
exports.DEFAULT_VOLUME_NAME = 'shared-volume';
exports.DEFAULT_VOLUME_PATH = '/shared-volume';
exports.DEFAULT_LOGS_PATH = '/shared-volume/logs/*.log';
exports.DEFAULT_HEALTH_CHECK_PORT = 5555;
/**
 * Regular expression for parsing environment variables in KEY=VALUE format
 */
exports.ENV_VAR_REGEX = /^([\w.]+)=(.*)$/;
