// @datadog/datadog-ci-base/dist/commands/plugin/check.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginCheckCommand = void 0;
const clipanion_1 = require("clipanion");
const message_box_1 = require("../../helpers/message-box");
const plugin_1 = require("../../helpers/plugin");
const __1 = require("../..");
class PluginCheckCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // Positional
        this.packageOrScope = clipanion_1.Option.String();
        this.command = clipanion_1.Option.String({ required: false });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            const packageName = (0, plugin_1.scopeToPackageName)(this.packageOrScope);
            if (this.context.builtinPlugins.includes(packageName)) {
                console.log();
                (0, message_box_1.messageBox)('Built-in plugin 🔌', 'green', [`The plugin ${packageName} is built-in!`]);
                console.log();
                return 0;
            }
            const succeeded = yield (0, plugin_1.checkPlugin)(this.packageOrScope, this.command);
            return succeeded ? 0 : 1;
        });
    }
}
exports.PluginCheckCommand = PluginCheckCommand;
PluginCheckCommand.paths = [['plugin', 'check']];
PluginCheckCommand.usage = clipanion_1.Command.Usage({
    category: 'Plugins',
    description: 'Check and troubleshoot the installation of a plugin.',
    examples: [
        ['Check the plugin by passing its package name', 'datadog-ci plugin check @datadog/datadog-ci-plugin-synthetics'],
        ['Check the plugin by passing its scope', 'datadog-ci plugin check synthetics'],
        ['Check the plugin by passing its scope and one of its commands', 'datadog-ci plugin check synthetics run-tests'],
    ],
});
