// @datadog/datadog-ci-base/dist/commands/plugin/list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginListCommand = void 0;
const chalk_1 = __importDefault(require("chalk"));
const clipanion_1 = require("clipanion");
const logger_1 = require("../../helpers/logger");
const plugin_1 = require("../../helpers/plugin");
const __1 = require("../..");
class PluginListCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // Positional
        this.json = clipanion_1.Option.Boolean('--json', { required: false });
        this.all = clipanion_1.Option.Boolean('-a,--all', { required: false });
        this.logger = new logger_1.Logger((s) => {
            this.context.stdout.write(s);
        }, logger_1.LogLevel.INFO);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            const allPlugins = (0, plugin_1.listAllPlugins)();
            const builtinPlugins = new Set(this.context.builtinPlugins);
            const installablePlugins = allPlugins.filter((name) => !builtinPlugins.has(name));
            const plugins = this.all
                ? [
                    ...installablePlugins.map((name) => ({ name, isBuiltin: false })),
                    ...Array.from(builtinPlugins).map((name) => ({ name, isBuiltin: true })),
                ]
                : installablePlugins.map((name) => ({ name, isBuiltin: false }));
            if (this.json) {
                this.logger.info(JSON.stringify(plugins.map(({ name, isBuiltin }) => ({
                    name,
                    scope: getScope(name),
                    isBuiltin,
                }))));
                return 0;
            }
            this.logger.info(`The following plugins are available:\n`);
            this.logger.info(plugins
                .map(({ name, isBuiltin }) => isBuiltin
                ? ` - ${chalk_1.default.bold('(built-in)')} ${chalk_1.default.magenta(name)}`
                : ` - ${chalk_1.default.bold.magenta(name)} (install with ${chalk_1.default.bold.cyan(`datadog-ci plugin install ${getScope(name)}`)})`)
                .join('\n'));
            return 0;
        });
    }
}
exports.PluginListCommand = PluginListCommand;
PluginListCommand.paths = [['plugin', 'list']];
PluginListCommand.usage = clipanion_1.Command.Usage({
    category: 'Plugins',
    description: 'List the available plugins.',
    details: `
      This command lists the plugins that can be installed with the \`datadog-ci plugin install\` command.
    `,
    examples: [
        ['List the available plugins', 'datadog-ci plugin list'],
        ['List all plugins, including built-in plugins', 'datadog-ci plugin list --all'],
    ],
});
const getScope = (plugin) => plugin.replace('@datadog/datadog-ci-plugin-', '');
