// @datadog/datadog-ci-base/dist/commands/flutter-symbols/helpers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getArchInfoFromFilename = exports.uploadMultipartHelper = exports.getFlutterRequestBuilder = void 0;
const upath_1 = __importDefault(require("upath"));
const base_intake_url_1 = require("@datadog/datadog-ci-base/helpers/base-intake-url");
const upload_1 = require("@datadog/datadog-ci-base/helpers/upload");
const utils_1 = require("@datadog/datadog-ci-base/helpers/utils");
const getFlutterRequestBuilder = (apiKey, cliVersion, site) => (0, utils_1.getRequestBuilder)({
    apiKey,
    baseUrl: (0, base_intake_url_1.getBaseSourcemapIntakeUrl)(site),
    headers: new Map([
        ['DD-EVP-ORIGIN', 'datadog-ci_flutter-symbols'],
        ['DD-EVP-ORIGIN-VERSION', cliVersion],
    ]),
    overrideUrl: 'api/v2/srcmap',
});
exports.getFlutterRequestBuilder = getFlutterRequestBuilder;
// This function exists partially just to make mocking network calls easier.
const uploadMultipartHelper = (requestBuilder, payload, opts) => __awaiter(void 0, void 0, void 0, function* () { return (0, upload_1.upload)(requestBuilder)(payload, opts); });
exports.uploadMultipartHelper = uploadMultipartHelper;
const getArchInfoFromFilename = (filename) => {
    var _a;
    const parsed = upath_1.default.parse(filename);
    const basename = parsed.name;
    const groups = (_a = /^.*\.(?<platform>.*)-(?<arch>.*)$/.exec(basename)) === null || _a === void 0 ? void 0 : _a.groups;
    if (!groups) {
        return undefined;
    }
    const value = {
        arch: groups.arch,
        platform: groups.platform,
    };
    if (!value.platform || !value.arch) {
        return undefined;
    }
    return value;
};
exports.getArchInfoFromFilename = getArchInfoFromFilename;
