// @datadog/datadog-ci-base/dist/commands/aas/common.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AasCommand = exports.getExtensionId = exports.WINDOWS_RUNTIME_EXTENSIONS = exports.resourceIdSegment = exports.renderWebApp = void 0;
const chalk_1 = __importDefault(require("chalk"));
const clipanion_1 = require("clipanion");
const constants_1 = require("../../constants");
const env_1 = require("../../helpers/env");
const fips_1 = require("../../helpers/fips");
const renderer_1 = require("../../helpers/renderer");
const azure_1 = require("../../helpers/serverless/azure");
const constants_2 = require("../../helpers/serverless/constants");
const utils_1 = require("../../helpers/utils");
const __1 = require("../..");
const renderWebApp = ({ name, slot }) => {
    return slot ? `${chalk_1.default.bold(name)} (slot: ${chalk_1.default.bold(slot)})` : `${chalk_1.default.bold(name)}`;
};
exports.renderWebApp = renderWebApp;
/**
 * Returns the final segment of the resource ID after `.../sites/` for a web app
 */
const resourceIdSegment = ({ name, slot }) => {
    return slot ? `${name}/slots/${slot}` : name;
};
exports.resourceIdSegment = resourceIdSegment;
exports.WINDOWS_RUNTIME_EXTENSIONS = {
    node: 'Datadog.AzureAppServices.Node.Apm',
    dotnet: 'Datadog.AzureAppServices.DotNet',
    java: 'Datadog.AzureAppServices.Java.Apm',
};
/**
 * Parses the extension Id from an extension name/id
 * Ex: "myapp/siteextensions/some.extension.id" -> "some.extension.id"
 */
const getExtensionId = (extension) => {
    const parts = extension.split('/');
    return parts[parts.length - 1];
};
exports.getExtensionId = getExtensionId;
class AasCommand extends __1.BaseCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.dryRun = clipanion_1.Option.Boolean('-d,--dry-run', false, {
            description: 'Run the command in dry-run mode, without making any changes',
        });
        this.subscriptionId = clipanion_1.Option.String('-s,--subscription-id', {
            description: 'Azure Subscription ID containing the Web App (or slot)',
        });
        this.resourceGroup = clipanion_1.Option.String('-g,--resource-group', {
            description: 'Name of the Azure Resource Group containing the Web App (or slot)',
        });
        this.aasName = clipanion_1.Option.String('-n,--name', {
            description: 'Name of the Web App to instrument',
        });
        this.slotName = clipanion_1.Option.String('--slot', {
            description: 'Name of the Web App slot to instrument',
        });
        this.resourceIds = clipanion_1.Option.Array('-r,--resource-id', {
            description: 'Full Azure resource IDs to instrument, for example, "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroup}/providers/Microsoft.Web/sites/{aasName}" or "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroup}/providers/Microsoft.Web/sites/{aasName}/slots/{slotName}"',
        });
        this.envVars = clipanion_1.Option.Array('-e,--env-vars', {
            description: 'Additional environment variables to set for the Web App (or slot). Can specify multiple in the form `--env-vars VAR1=VALUE1 --env-vars VAR2=VALUE2`.',
        });
        this.configPath = clipanion_1.Option.String('--config', {
            description: 'Path to the configuration file',
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.fipsConfig = {
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
    }
    get dryRunPrefix() {
        return this.dryRun ? renderer_1.dryRunTag + ' ' : '';
    }
    get additionalConfig() {
        return {};
    }
    enableFips() {
        (0, fips_1.enableFips)(this.fips || this.fipsConfig.fips, this.fipsIgnoreError || this.fipsConfig.fipsIgnoreError);
    }
    ensureConfig() {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            const config = (yield (0, utils_1.resolveConfigFromFile)({
                aas: Object.assign({ subscriptionId: this.subscriptionId, resourceGroup: this.resourceGroup, aasName: this.aasName, slotName: this.slotName, envVars: this.envVars }, this.additionalConfig),
            }, {
                configPath: this.configPath,
                defaultConfigPaths: utils_1.DEFAULT_CONFIG_PATHS,
            })).aas;
            const webApps = {};
            const errors = [];
            if (process.env.DD_API_KEY === undefined) {
                errors.push('DD_API_KEY environment variable is required');
            }
            // Validate that envVars, if provided, are in the format 'key=value'
            if ((_a = config.envVars) === null || _a === void 0 ? void 0 : _a.some((e) => !constants_2.ENV_VAR_REGEX.test(e))) {
                errors.push('All envVars must be in the format `KEY=VALUE`');
            }
            // Validate that extraTags, if provided, comply with the expected format
            if (config.extraTags && !config.extraTags.match(constants_2.EXTRA_TAGS_REG_EXP)) {
                errors.push('Extra tags do not comply with the <key>:<value> array.');
            }
            // Validate musl setting
            if (config.isMusl && !config.isDotnet) {
                errors.push('--musl can only be set if --dotnet is also set, as it is only relevant for containerized .NET applications.');
            }
            if (config.windowsRuntime && !(config.windowsRuntime in exports.WINDOWS_RUNTIME_EXTENSIONS)) {
                errors.push(`--windows-runtime must be one of: ${Object.keys(exports.WINDOWS_RUNTIME_EXTENSIONS).join(', ')}`);
            }
            const specifiedSiteArgs = [config.subscriptionId, config.resourceGroup, config.aasName];
            // all or none of the site args should be specified (except slot)
            const noneSpecified = specifiedSiteArgs.every((arg) => !arg);
            const allSpecified = specifiedSiteArgs.every((arg) => arg);
            if (!(allSpecified || noneSpecified)) {
                errors.push('--subscription-id, --resource-group, and --name must be specified together or not at all');
            }
            else if (!allSpecified && !!config.slotName) {
                errors.push('--slot can only specified if --subscription-id, --resource-group, and --name are specified');
            }
            else if (!noneSpecified) {
                webApps[config.subscriptionId] = { [config.resourceGroup]: [{ name: config.aasName, slot: config.slotName }] };
            }
            if ((_b = this.resourceIds) === null || _b === void 0 ? void 0 : _b.length) {
                for (const resourceId of this.resourceIds) {
                    const parsed = (0, azure_1.parseResourceId)(resourceId);
                    if (!parsed || (!!parsed.subType && parsed.subType !== 'slots')) {
                        errors.push(`Invalid Web App (or Slot) resource ID: ${resourceId}`);
                        continue;
                    }
                    const { subscriptionId, resourceGroup, name, subResourceName: slot } = parsed;
                    if (!webApps[subscriptionId]) {
                        webApps[subscriptionId] = {};
                    }
                    if (!webApps[subscriptionId][resourceGroup]) {
                        webApps[subscriptionId][resourceGroup] = [];
                    }
                    webApps[subscriptionId][resourceGroup].push({ name, slot });
                }
            }
            if (!((_c = this.resourceIds) === null || _c === void 0 ? void 0 : _c.length) && specifiedSiteArgs.every((arg) => !arg)) {
                errors.push('No App Services specified to instrument');
            }
            return [webApps, config, errors];
        });
    }
}
exports.AasCommand = AasCommand;
