// @datadog/datadog-ci-base/dist/commands/aas/instrument.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AasInstrumentCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const common_1 = require("./common");
class AasInstrumentCommand extends common_1.AasCommand {
    constructor() {
        super(...arguments);
        this.service = clipanion_1.Option.String('--service', {
            description: 'The value for the service tag. For example, `my-service`',
        });
        this.environment = clipanion_1.Option.String('--env,--environment', {
            description: 'The value for the env tag. For example, `prod`',
        });
        this.version = clipanion_1.Option.String('--version', {
            description: 'The value for the version tag. For example, `1.0.0`',
        });
        this.isInstanceLoggingEnabled = clipanion_1.Option.Boolean('--instance-logging', false, {
            description: 'When enabled, log collection is automatically configured for an additional file path: /home/LogFiles/*$COMPUTERNAME*.log',
        });
        this.logPath = clipanion_1.Option.String('--log-path', {
            description: 'Where you write your logs. For example, /home/LogFiles/*.log or /home/LogFiles/myapp/*.log',
        });
        this.shouldNotRestart = clipanion_1.Option.Boolean('--no-restart', false, {
            description: 'Do not restart the App Service after applying instrumentation.',
        });
        this.isDotnet = clipanion_1.Option.Boolean('--dotnet,--dotnet-container', false, {
            description: 'Add in required .NET-specific configuration options, is automatically inferred for code runtimes. This should be specified if you are using a containerized .NET app.',
        });
        this.isMusl = clipanion_1.Option.Boolean('--musl', false, {
            description: 'Add in required .NET-specific configuration options for musl-based .NET apps. This should be specified if you are using a containerized .NET app on a musl-based distribution like Alpine Linux.',
        });
        this.sourceCodeIntegration = clipanion_1.Option.Boolean('--source-code-integration,--sourceCodeIntegration', true, {
            description: 'Enable source code integration to add git metadata as tags. Defaults to enabled. Specify `--no-source-code-integration` to disable.',
        });
        this.uploadGitMetadata = clipanion_1.Option.Boolean('--upload-git-metadata,--uploadGitMetadata', true, {
            description: 'Upload git metadata to Datadog. Defaults to enabled. Specify `--no-upload-git-metadata` to disable.',
        });
        this.extraTags = clipanion_1.Option.String('--extra-tags,--extraTags', {
            description: 'Additional tags to add to the service in the format "key1:value1,key2:value2"',
        });
        this.windowsRuntime = clipanion_1.Option.String('--windows-runtime', {
            description: 'Manually specify the Windows runtime (`node`, `dotnet`, or `java`) used by the extension to override automatic detection.',
        });
    }
    get additionalConfig() {
        return {
            service: this.service,
            environment: this.environment,
            version: this.version,
            isInstanceLoggingEnabled: this.isInstanceLoggingEnabled,
            logPath: this.logPath,
            shouldNotRestart: this.shouldNotRestart,
            isDotnet: this.isDotnet,
            isMusl: this.isMusl,
            sourceCodeIntegration: this.sourceCodeIntegration,
            uploadGitMetadata: this.uploadGitMetadata,
            extraTags: this.extraTags,
            windowsRuntime: this.windowsRuntime,
        };
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.AasInstrumentCommand = AasInstrumentCommand;
AasInstrumentCommand.paths = [['aas', 'instrument']];
AasInstrumentCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Apply Datadog instrumentation to an Azure Web App or slot.',
});
