// @datadog/datadog-ci-plugin-sbom/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TARGET_FRAMEWORK_KEY = exports.REACHABLE_SYMBOL_LOCATION_KEY_PREFIX = exports.EXCLUSION_KEY = exports.IS_DEPENDENCY_DEV_ENVIRONMENT_PROPERTY_KEY = exports.IS_DEPENDENCY_DIRECT_PROPERTY_KEY = exports.PACKAGE_MANAGER_PROPERTY_KEY = exports.LEGACY_REACHABLE_SYMBOL_LOCATION_KEY_PREFIX = exports.LEGACY_EXCLUSION_KEY = exports.LEGACY_IS_DEPENDENCY_DEV_ENVIRONMENT_PROPERTY_KEY = exports.LEGACY_IS_DEPENDENCY_DIRECT_PROPERTY_KEY = exports.LEGACY_PACKAGE_MANAGER_PROPERTY_KEY = exports.API_ENDPOINT = void 0;
exports.API_ENDPOINT = 'api/v2/static-analysis-sca/dependencies';
// os-scanner specific SBOM properties
exports.LEGACY_PACKAGE_MANAGER_PROPERTY_KEY = 'osv-scanner:package-manager';
exports.LEGACY_IS_DEPENDENCY_DIRECT_PROPERTY_KEY = 'osv-scanner:is-direct';
exports.LEGACY_IS_DEPENDENCY_DEV_ENVIRONMENT_PROPERTY_KEY = 'osv-scanner:is-dev';
// datadog-sbom-generator specific SBOM properties
exports.LEGACY_EXCLUSION_KEY = 'datadog-sbom-generator:exclusion';
exports.LEGACY_REACHABLE_SYMBOL_LOCATION_KEY_PREFIX = 'datadog-sbom-generator:reachable-symbol-location';
// datadog canonical SBOM properties
exports.PACKAGE_MANAGER_PROPERTY_KEY = 'datadog:package-manager';
exports.IS_DEPENDENCY_DIRECT_PROPERTY_KEY = 'datadog:is-direct';
exports.IS_DEPENDENCY_DEV_ENVIRONMENT_PROPERTY_KEY = 'datadog:is-dev';
exports.EXCLUSION_KEY = 'datadog:exclusion';
exports.REACHABLE_SYMBOL_LOCATION_KEY_PREFIX = 'datadog:reachable-symbol-location';
// datadog-sca specific SBOM properties
exports.TARGET_FRAMEWORK_KEY = 'datadog:target-framework';
