// @datadog/datadog-ci-plugin-sbom/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DependencyLicense = exports.DependencyLanguage = void 0;
var DependencyLanguage;
(function (DependencyLanguage) {
    DependencyLanguage["DOTNET"] = "dotnet";
    DependencyLanguage["NPM"] = "node";
    DependencyLanguage["PYTHON"] = "python";
    DependencyLanguage["PHP"] = "php";
    DependencyLanguage["RUST"] = "rust";
    DependencyLanguage["RUBY"] = "ruby";
    DependencyLanguage["GO"] = "go";
    DependencyLanguage["JVM"] = "jvm";
    DependencyLanguage["C_CPP"] = "c_cpp";
})(DependencyLanguage || (exports.DependencyLanguage = DependencyLanguage = {}));
// List from https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository
var DependencyLicense;
(function (DependencyLicense) {
    DependencyLicense["AFL3"] = "AFL-3.0";
    DependencyLicense["APACHE2"] = "Apache-2.0";
    DependencyLicense["ARTISTIC2"] = "Artistic-2.0";
    DependencyLicense["BSL1"] = "BSL-1.0";
    DependencyLicense["BSD2CLAUSE"] = "BSD-2-Clause";
    DependencyLicense["BSD3CLAUSE"] = "BSD-3-Clause";
    DependencyLicense["BSD3CLAUSECLEAR"] = "BSD-3-Clause-Clear";
    DependencyLicense["BSD4CLAUSE"] = "BSD-4-Clause";
    DependencyLicense["ZEROBSD"] = "0BSD";
    DependencyLicense["CC"] = "CC";
    DependencyLicense["CC0_1_0"] = "CC0-1.0";
    DependencyLicense["CC_BY_4_0"] = "CC-BY-4.0";
    DependencyLicense["CC_BY_SA_4_0"] = "CC-BY-SA-4.0";
    DependencyLicense["WTFPL"] = "WTFPL";
    DependencyLicense["ECL2_0"] = "ECL-2.0";
    DependencyLicense["EPL1_0"] = "EPL-1.0";
    DependencyLicense["EPL2_0"] = "EPL-2.0";
    DependencyLicense["EUPL1_1"] = "EUPL-1.1";
    DependencyLicense["AGPL3_0"] = "AGPL-3.0";
    DependencyLicense["GPL"] = "GPL";
    DependencyLicense["GPL2_0"] = "GPL-2.0";
    DependencyLicense["GPL3_0"] = "GPL-3.0";
    DependencyLicense["LGPL"] = "LGPL";
    DependencyLicense["LGPL2_1"] = "LGPL-2.1";
    DependencyLicense["LGPL3_0"] = "LGPL-3.0";
    DependencyLicense["ISC"] = "ISC";
    DependencyLicense["LPPL_1_3C"] = "LPPL-1.3c";
    DependencyLicense["MS_PL"] = "MS-PL";
    DependencyLicense["MIT"] = "MIT";
    DependencyLicense["MPL_2_0"] = "MPL-2.0";
    DependencyLicense["OSL_3_0"] = "OSL-3.0";
    DependencyLicense["POSTGRESQL"] = "PostgreSQL";
    DependencyLicense["OFL_1_1"] = "OFL-1.1";
    DependencyLicense["NCSA"] = "NCSA";
    DependencyLicense["UNLICENSE"] = "Unlicense";
    DependencyLicense["ZLIB"] = "Zlib";
})(DependencyLicense || (exports.DependencyLicense = DependencyLicense = {}));
