// @datadog/datadog-ci-base/dist/commands/coverage/upload.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoverageUploadCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("@datadog/datadog-ci-base/helpers/plugin");
const __1 = require("../..");
class CoverageUploadCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.reportPaths = clipanion_1.Option.Rest({ required: 1 });
        this.flags = clipanion_1.Option.Array('--flags', {
            description: 'Flags to mark coverage reports for grouping and filtering (e.g., type:unit-tests, jvm-21). Maximum 32 flags per report.',
        });
        this.format = clipanion_1.Option.String('--format', { description: 'The format of the coverage report files' });
        this.uploadGitDiff = clipanion_1.Option.Boolean('--upload-git-diff', true, {
            description: 'If the command is run in a PR context, it will try to upload the PR git diff along with the coverage data',
        });
        this.skipGitMetadataUpload = clipanion_1.Option.Boolean('--skip-git-metadata-upload', false, {
            description: 'Skip the upload of git metadata',
        });
        this.disableFileFixes = clipanion_1.Option.Boolean('--disable-file-fixes', false, {
            description: 'Disable the generation and upload of file fixes for code coverage',
        });
        this.fileFixesSearchPath = clipanion_1.Option.String('--file-fixes-search-path', {
            description: 'Override the root directory used to scan source files for file fixes. By default, the repository root is used.',
        });
        this.gitRepositoryURL = clipanion_1.Option.String('--git-repository-url', {
            description: 'The repository URL to retrieve git metadata from',
        });
        this.basePath = clipanion_1.Option.String('--base-path', { description: 'The base path to the coverage report files' });
        this.ignoredPaths = clipanion_1.Option.String('--ignored-paths', {
            description: 'A comma-separated list of paths that should be excluded from automatic reports discovery (only applicable when `--auto-discovery` is set). Glob patterns are supported.',
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false, { description: 'Enable FIPS mode for the command' });
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false, {
            description: 'Ignore FIPS errors and continue execution',
        });
        this.verbose = clipanion_1.Option.Boolean('--verbose', false, { hidden: true });
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false, {
            description: 'Run the command in dry run mode, without uploading any data to Datadog',
        });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.CoverageUploadCommand = CoverageUploadCommand;
CoverageUploadCommand.paths = [['coverage', 'upload']];
CoverageUploadCommand.usage = clipanion_1.Command.Usage({
    category: 'CI Visibility',
    description: 'Upload code coverage reports files to Datadog.',
    details: `
      This command will upload code coverage report files to Datadog.\n
      See README for details.
    `,
    examples: [
        ['Upload all code coverage report files in current directory and its subfolders', 'datadog-ci coverage upload .'],
        [
            'Upload all code coverage report files in current directory and its subfolders, ignoring src/ignored-module-a and src/ignored-module-b',
            'datadog-ci coverage upload --ignored-paths src/ignored-module-a,src/ignored-module-b .',
        ],
        [
            'Upload all code coverage report files in src/unit-test-coverage and src/acceptance-test-coverage',
            'datadog-ci coverage upload src/unit-test-coverage src/acceptance-test-coverage',
        ],
        [
            'Upload all XML code coverage report files in /coverage/ folders, ignoring src/ignored-module-a',
            'datadog-ci coverage upload **/coverage/*.xml --ignored-paths src/ignored-module-a',
        ],
        ['Upload coverage with flags', 'datadog-ci coverage upload --flags type:unit-tests --flags jvm-21 .'],
        [
            'Upload all code coverage report files in current directory to the datadoghq.eu site',
            'DD_SITE=datadoghq.eu datadog-ci coverage upload .',
        ],
        [
            'Upload all code coverage report files in current directory with extra verbosity',
            'datadog-ci coverage upload --verbose .',
        ],
    ],
});
