// @datadog/datadog-ci-base/dist/commands/lambda/cloudwatch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaCloudwatchCommand = void 0;
const clipanion_1 = require("clipanion");
const t = __importStar(require("typanion"));
const plugin_1 = require("../../helpers/plugin");
const __1 = require("../..");
class LambdaCloudwatchCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.action = clipanion_1.Option.String({ validator: t.isEnum(['disable', 'enable']) });
        this.configPath = clipanion_1.Option.String('--config', {
            description: 'Path to the configuration file',
        });
        this.dryRun = clipanion_1.Option.Boolean('-d,--dry,--dry-run', false, {
            description: 'Preview the changes from running the command',
        });
        this.functions = clipanion_1.Option.Array('-f,--function', [], {
            description: 'The ARN of the Lambda function, or the name of the Lambda function (`--region` must be defined)',
        });
        this.profile = clipanion_1.Option.String('--profile', {
            description: 'Specify the AWS named profile credentials to use. Learn more about AWS named profiles here: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html#using-profiles',
        });
        this.regExPattern = clipanion_1.Option.String('--functions-regex,--functionsRegex', {
            description: 'A regex pattern to match with the Lambda function name',
        });
        this.region = clipanion_1.Option.String('-r,--region', {
            description: 'Default region to use, when `--function` is specified by the function name instead of the ARN',
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.LambdaCloudwatchCommand = LambdaCloudwatchCommand;
LambdaCloudwatchCommand.paths = [['lambda', 'cloudwatch']];
LambdaCloudwatchCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Disable or Re-Enable CloudWatch Logs for Lambda functions by managing IAM deny policies.',
});
