// @datadog/datadog-ci-base/dist/commands/lambda/flare.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaFlareCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const __1 = require("../..");
class LambdaFlareCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.isDryRun = clipanion_1.Option.Boolean('-d,--dry,--dry-run', false, {
            description: 'Preview collected data which would be sent to Datadog support',
        });
        this.withLogs = clipanion_1.Option.Boolean('--with-logs', false, {
            description: 'Collect recent CloudWatch logs for the specified function',
        });
        this.functionName = clipanion_1.Option.String('-f,--function', {
            description: 'The ARN of the Lambda function to gather data for, or the name of the Lambda function (--region must be defined)',
        });
        this.region = clipanion_1.Option.String('-r,--region', {
            description: 'Default region to use, when --function is specified by the function name instead of the ARN',
        });
        this.caseId = clipanion_1.Option.String('-c,--case-id', {
            description: 'The Datadog case ID to send the files to',
        });
        this.email = clipanion_1.Option.String('-e,--email', {
            description: 'The email associated with the specified case ID',
        });
        this.start = clipanion_1.Option.String('--start', {
            description: `Only gather logs within the time range (--with-logs must be included). This argument is a number in milliseconds since Unix Epoch. Must be used with --end`,
        });
        this.end = clipanion_1.Option.String('--end', {
            description: `Only gather logs within the time range (--with-logs must be included). This argument is a number in milliseconds since Unix Epoch. Must be used with --start`,
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.LambdaFlareCommand = LambdaFlareCommand;
LambdaFlareCommand.paths = [['lambda', 'flare']];
LambdaFlareCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Gather config, logs, tags, project files, and more from a Lambda function and sends them to Datadog support.',
});
