// @datadog/datadog-ci-base/dist/commands/lambda/instrument.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaInstrumentCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const lambda_layer_versions_1 = require("../../helpers/serverless/lambda-layer-versions");
const __1 = require("../..");
const LAYER_VERSIONS_HELP_STRING = Object.entries(lambda_layer_versions_1.LAMBDA_LAYER_VERSIONS)
    .filter(([key, _]) => key !== 'extension')
    .map(([key, value]) => `${key} - ${value}`)
    .join(', ');
class LambdaInstrumentCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.apmFlushDeadline = clipanion_1.Option.String('--apm-flush-deadline', {
            description: `Used to determine when to submit spans before a timeout occurs, in milliseconds. When the remaining time in an AWS Lambda invocation is less than the value set, the tracer attempts to submit the current active spans and all finished spans. Supported for NodeJS and Python. Defaults to '100'`,
        });
        this.appsecEnabled = clipanion_1.Option.Boolean('--appsec', false, {
            description: `Enable Application Security Monitoring for the Lambda function. Defaults to 'false'`,
        });
        this.captureLambdaPayload = clipanion_1.Option.String('--capture-lambda-payload,--captureLambdaPayload', {
            description: `Whether to capture and store the payload and response of a lambda invocation. Defaults to 'false'`,
        });
        this.configPath = clipanion_1.Option.String('--config', {
            description: 'Path to the configuration file',
        });
        this.dryRun = clipanion_1.Option.Boolean('-d,--dry,--dry-run', false, {
            description: 'Preview changes running command would apply',
        });
        this.environment = clipanion_1.Option.String('--env', {
            description: `Use --env to separate out your staging, development, and production environments. Learn more about the env tag here: https://docs.datadoghq.com/serverless/troubleshooting/serverless_tagging/#the-env-tag`,
        });
        this.extensionVersion = clipanion_1.Option.String('-e,--extension-version,--extensionVersion', {
            description: `Version of the Datadog Lambda Extension layer to apply. Defaults to 'latest'. Setting this to 'latest' will use version ${lambda_layer_versions_1.LAMBDA_LAYER_VERSIONS['extension']}. When using the extension, make sure to export DATADOG_API_KEY (or if encrypted, DATADOG_KMS_API_KEY or DATADOG_API_KEY_SECRET_ARN) in your environment as well. Mutually exclusive with the forwarder. Set to 'none' to explicitly disable adding the extension layer. Learn more about the Lambda Extension here: https://docs.datadoghq.com/serverless/libraries_integrations/extension`,
        });
        this.extraTags = clipanion_1.Option.String('--extra-tags,--extraTags', {
            description: `Add custom tags to your Lambda function in Datadog. Must be a list of <key>:<value> separated by commas such as: layer:api,team:intake`,
        });
        this.flushMetricsToLogs = clipanion_1.Option.String('--flush-metrics-to-logs,--flushMetricsToLogs', {
            description: `Whether to send metrics via the Datadog Forwarder asynchronously (https://docs.datadoghq.com/serverless/custom_metrics?tab=python#enabling-asynchronous-custom-metrics). If you disable this parameter, it's required to export DATADOG_API_KEY (or if encrypted, DATADOG_KMS_API_KEY or DATADOG_API_KEY_SECRET_ARN). Defaults to 'true'`,
        });
        this.forwarder = clipanion_1.Option.String('--forwarder', {
            description: `The ARN of the datadog forwarder (https://docs.datadoghq.com/logs/guide/forwarder/) to attach this function's LogGroup to`,
        });
        this.functions = clipanion_1.Option.Array('-f,--function', [], {
            description: `The ARN of the Lambda function to be instrumented, or the name of the Lambda function (--region must be defined)`,
        });
        this.interactive = clipanion_1.Option.Boolean('-i,--interactive', false, {
            description: `Allows the user to interactively choose how their function gets instrumented. There is no need to provide any other flags if you choose to use interactive mode since you will be prompted for the information instead`,
        });
        this.layerAWSAccount = clipanion_1.Option.String('-a,--layer-account,--layerAccount', { hidden: true });
        this.layerVersion = clipanion_1.Option.String('-v,--layer-version,--layerVersion', {
            description: `Version of the Datadog Lambda Library layer to apply. Defaults to 'latest'. Setting this to 'latest' will use one of the following versions based on your runtime: ${LAYER_VERSIONS_HELP_STRING}. Setting to 'none' will disable adding the language layer.`,
        });
        this.logging = clipanion_1.Option.String('--logging', {
            description: `Whether to collect logs using the Lambda Extension. Defaults to 'true'`,
        });
        this.logLevel = clipanion_1.Option.String('--log-level,--logLevel', {
            description: `Set to debug to see additional output from the Datadog Lambda Library and/or Lambda Extension for troubleshooting purposes`,
        });
        this.mergeXrayTraces = clipanion_1.Option.String('--merge-xray-traces,--mergeXrayTraces', {
            description: `Whether to join dd-trace traces to AWS X-Ray traces. Useful for tracing API Gateway spans. Defaults to 'false'`,
        });
        this.profile = clipanion_1.Option.String('--profile', {
            description: `Specify the AWS named profile credentials to use to instrument. Learn more about AWS named profiles here: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html#using-profiles`,
        });
        this.regExPattern = clipanion_1.Option.String('--functions-regex,--functionsRegex', {
            description: 'A regex pattern to match with the Lambda function name',
        });
        this.region = clipanion_1.Option.String('-r,--region', {
            description: 'Default region to use, when --function is specified by the function name instead of the ARN',
        });
        this.service = clipanion_1.Option.String('--service', {
            description: `Use --service to group related functions belonging to similar workloads. Learn more about the service tag here: https://docs.datadoghq.com/serverless/troubleshooting/serverless_tagging/#the-service-tag`,
        });
        this.sourceCodeIntegration = clipanion_1.Option.Boolean('-s,--source-code-integration,--sourceCodeIntegration', true, {
            description: `Whether to enable Datadog Source Code Integration (https://docs.datadoghq.com/integrations/guide/source-code-integration). This will tag your lambda(s) with the Git repository URL and the latest commit hash of the current local directory. Note: Git repository must not be ahead of remote, and must not be dirty. Defaults to 'true'`,
        });
        this.uploadGitMetadata = clipanion_1.Option.Boolean('-u,--upload-git-metadata,--uploadGitMetadata', true, {
            description: `Whether to enable Git metadata uploading, as a part of source code integration. Git metadata uploading is only required if you don't have the Datadog Github Integration installed. Defaults to 'true'`,
        });
        this.tracing = clipanion_1.Option.String('--tracing', {
            description: `Whether to enable dd-trace tracing on your Lambda. Defaults to 'true'`,
        });
        this.version = clipanion_1.Option.String('--version', {
            description: `Add the --version tag to correlate spikes in latency, load or errors to new versions. Learn more about the version tag here: https://docs.datadoghq.com/serverless/troubleshooting/serverless_tagging/#the-version-tag`,
        });
        this.llmobs = clipanion_1.Option.String('--llmobs', {
            description: `If specified, enables LLM Observability for the instrumented function(s) with the provided ML application name. Defaults to 'false'`,
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.lambdaFips = clipanion_1.Option.Boolean('--lambda-fips', false, {
            description: `Enable FIPS support in the Lambda functions deployed using this tool. Note that for full FIPS compliance, a FIPS endpoint such as ddog-gov.com is required`,
        });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.LambdaInstrumentCommand = LambdaInstrumentCommand;
LambdaInstrumentCommand.paths = [['lambda', 'instrument']];
LambdaInstrumentCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Apply Datadog instrumentation to a Lambda.',
});
