// @datadog/datadog-ci-base/dist/commands/lambda/uninstrument.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaUninstrumentCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const __1 = require("../..");
class LambdaUninstrumentCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.configPath = clipanion_1.Option.String('--config', {
            description: 'Path to the configuration file',
        });
        this.dryRun = clipanion_1.Option.Boolean('-d,--dry,--dry-run', false, {
            description: 'Preview changes running command would apply',
        });
        this.forwarder = clipanion_1.Option.String('--forwarder', {
            description: `The ARN of the datadog forwarder (https://docs.datadoghq.com/logs/guide/forwarder/) to remove from this function`,
        });
        this.functions = clipanion_1.Option.Array('-f,--function', [], {
            description: `The ARN of the Lambda function to be uninstrumented, or the name of the Lambda function (--region must be defined)`,
        });
        this.interactive = clipanion_1.Option.Boolean('-i,--interactive', false, {
            description: `Allows the user to interactively choose how their function gets uninstrumented. There is no need to provide any other flags if you choose to use interactive mode since you will be prompted for the information instead`,
        });
        this.profile = clipanion_1.Option.String('--profile', {
            description: `Specify the AWS named profile credentials to use to uninstrument. Learn more about AWS named profiles here: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html#using-profiles`,
        });
        this.regExPattern = clipanion_1.Option.String('--functions-regex,--functionsRegex', {
            description: 'A regex pattern to match with the Lambda function name to be uninstrumented',
        });
        this.region = clipanion_1.Option.String('-r,--region', {
            description: 'Default region to use, when --function is specified by the function name instead of the ARN',
        });
        /**
         * Arguments that are not really in use, but to
         * make uninstrumentation easier for the user.
         */
        this.layerVersion = clipanion_1.Option.String('-v,--layer-version,--layerVersion', { hidden: true });
        this.tracing = clipanion_1.Option.String('--tracing', { hidden: true });
        this.logLevel = clipanion_1.Option.String('--log-level,--logLevel', { hidden: true });
        this.service = clipanion_1.Option.String('--service', { hidden: true });
        this.environment = clipanion_1.Option.String('--env', { hidden: true });
        this.version = clipanion_1.Option.String('--version', { hidden: true });
        this.appsecEnabled = clipanion_1.Option.Boolean('--appsec', { hidden: true });
        this.apmFlushDeadline = clipanion_1.Option.String('--apm-flush-deadline', { hidden: true });
        this.extraTags = clipanion_1.Option.String('--extra-tags,--extraTags', { hidden: true });
        this.extensionVersion = clipanion_1.Option.String('-e,--extension-version,--extensionVersion', { hidden: true });
        this.mergeXrayTraces = clipanion_1.Option.String('--merge-xray-traces,--mergeXrayTraces', { hidden: true });
        this.flushMetricsToLogs = clipanion_1.Option.String('--flush-metrics-to-logs,--flushMetricsToLogs', { hidden: true });
        this.captureLambdaPayload = clipanion_1.Option.String('--capture-lambda-payload,--captureLambdaPayload', { hidden: true });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.LambdaUninstrumentCommand = LambdaUninstrumentCommand;
LambdaUninstrumentCommand.paths = [['lambda', 'uninstrument']];
LambdaUninstrumentCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Revert Datadog instrumentation in a Lambda.',
});
