// @datadog/datadog-ci-plugin-sbom/dist/commands/upload.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginCommand = void 0;
const fs_1 = __importDefault(require("fs"));
const process_1 = __importDefault(require("process"));
const upload_1 = require("@datadog/datadog-ci-base/commands/sbom/upload");
const constants_1 = require("@datadog/datadog-ci-base/constants");
const env_1 = require("@datadog/datadog-ci-base/helpers/env");
const fips_1 = require("@datadog/datadog-ci-base/helpers/fips");
const tags_1 = require("@datadog/datadog-ci-base/helpers/tags");
const axios_1 = require("axios");
const api_1 = require("../api");
const payload_1 = require("../payload");
const renderer_1 = require("../renderer");
const validation_1 = require("../validation");
class PluginCommand extends upload_1.SbomUploadCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.config = {
            apiKey: process_1.default.env.DATADOG_API_KEY || process_1.default.env.DD_API_KEY,
            appKey: process_1.default.env.DATADOG_APP_KEY || process_1.default.env.DD_APP_KEY || '',
            env: process_1.default.env.DD_ENV,
            envVarTags: process_1.default.env.DD_TAGS,
            fips: process_1.default.env[constants_1.FIPS_ENV_VAR],
        };
        this.fipsConfig = {
            fips: (_a = (0, env_1.toBoolean)(process_1.default.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process_1.default.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
    }
    /**
     * Execute the command, which means parse the SBOM file, ensure they are
     * compliant with their schema and upload them to datadog.
     */
    execute() {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            (0, fips_1.enableFips)(this.fips || this.fipsConfig.fips, this.fipsIgnoreError || this.fipsConfig.fipsIgnoreError);
            const githubEvent = process_1.default.env.GITHUB_EVENT_NAME;
            const gitlabEvent = process_1.default.env.CI_PIPELINE_SOURCE;
            const azureReason = process_1.default.env.BUILD_REASON;
            if (githubEvent === 'pull_request') {
                // https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-commands#example-setting-an-error-message
                this.context.stdout.write('::error title=Unsupported Trigger::The `pull_request` event is not supported by Datadog Code Security and will cause issues with the product. ' +
                    'To continue using Datadog Code Security, use `push` instead. See: https://docs.datadoghq.com/security/code_security/software_composition_analysis/setup_static/?tab=github#run-via-pipelines-integration for more information.');
                return 1;
            }
            if (gitlabEvent === 'merge_request_event') {
                this.context.stderr.write('The `merge_request_event` pipeline source is not supported by Datadog Code Security and will cause issues with the product. ' +
                    'To continue using Datadog Code Security, use `push` instead. See: https://docs.datadoghq.com/security/code_security/software_composition_analysis/setup_static/?tab=github#run-via-pipelines-integration for more information.');
                return 1;
            }
            if (azureReason === 'PullRequest') {
                // https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#logging-commands-for-build-pipelines
                this.context.stdout.write('##vso[task.logissue type=error]The `PullRequest` build reason is not supported by Datadog Code Security and will cause issues with the product. ' +
                    'To continue using Datadog Code Security, use `push` instead. See: https://docs.datadoghq.com/security/code_security/software_composition_analysis/setup_static/?tab=github#run-via-pipelines-integration for more information.');
                return 1;
            }
            const service = 'datadog-ci';
            const environment = this.env;
            if (!this.basePath || !this.basePath.length) {
                this.context.stderr.write('Missing basePath\n');
                return 1;
            }
            if (!this.config.apiKey) {
                this.context.stderr.write('API key not defined, define the environment variable DD_API_KEY.\n');
                return 1;
            }
            if (!this.config.appKey) {
                this.context.stderr.write('APP key not defined, define the environment variable DD_APP_KEY.\n');
                return 1;
            }
            // Get the API helper to send the payload
            const api = (0, api_1.getApiHelper)(this.config.apiKey, this.config.appKey);
            const tags = yield (0, tags_1.getSpanTags)(this.config, this.tags, !this.noCiTags, this.gitPath);
            // Gather any missing mandatory git fields to display to the user
            const missingGitFields = (0, tags_1.getMissingRequiredGitTags)(tags);
            if (missingGitFields.length > 0) {
                this.context.stdout.write((0, renderer_1.renderMissingTags)(missingGitFields));
                return 1;
            }
            const validator = (0, validation_1.getValidator)();
            const startTimeMs = Date.now();
            const basePath = this.basePath;
            if (this.debug) {
                this.context.stdout.write(`Processing file ${basePath}\n`);
            }
            if (!(0, validation_1.validateSbomFileAgainstSchema)(basePath, validator, !!this.debug)) {
                this.context.stdout.write('SBOM file not fully compliant against CycloneDX 1.4, 1.5 or 1.6 specifications (use --debug to get validation error)\n');
            }
            if (!(0, validation_1.validateFileAgainstToolRequirements)(basePath, !!this.debug)) {
                this.context.stdout.write((0, renderer_1.renderInvalidFile)(basePath));
                return 1;
            }
            const jsonContent = JSON.parse(fs_1.default.readFileSync(basePath).toString('utf8'));
            // Upload content
            try {
                const scaPayload = (0, payload_1.generatePayload)(jsonContent, tags, service, environment);
                if (!scaPayload) {
                    this.context.stdout.write((0, renderer_1.renderInvalidPayload)(basePath));
                    return 1;
                }
                this.context.stdout.write((0, renderer_1.renderPayloadWarning)(scaPayload.dependencies));
                scaPayload.dependencies = (0, validation_1.filterInvalidDependencies)(scaPayload.dependencies);
                this.context.stdout.write((0, renderer_1.renderUploading)(basePath, scaPayload));
                yield api(scaPayload);
                if (this.debug) {
                    this.context.stdout.write(`Upload done for ${basePath}.\n`);
                }
            }
            catch (error) {
                if ((0, axios_1.isAxiosError)(error)) {
                    if (((_a = error.response) === null || _a === void 0 ? void 0 : _a.status) === 409) {
                        const sha = tags[tags_1.GIT_SHA] || 'sha-not-found';
                        const branch = tags[tags_1.GIT_BRANCH] || 'branch-not-found';
                        this.context.stderr.write((0, renderer_1.renderDuplicateUpload)(branch, sha));
                        return 0;
                    }
                    if (((_b = error.response) === null || _b === void 0 ? void 0 : _b.status) === 412) {
                        const repositoryUrl = tags[tags_1.GIT_REPOSITORY_URL] || 'repo-url-not-found';
                        this.context.stderr.write((0, renderer_1.renderNoDefaultBranch)(repositoryUrl));
                        return 1;
                    }
                }
                this.context.stderr.write((0, renderer_1.renderFailedUpload)(basePath, error, !!this.debug));
                return 1;
            }
            const uploadTimeMs = (Date.now() - startTimeMs) / 1000;
            this.context.stdout.write((0, renderer_1.renderSuccessfulCommand)(uploadTimeMs));
            return 0;
        });
    }
}
exports.PluginCommand = PluginCommand;
