// @datadog/datadog-ci-base/dist/commands/sarif/upload.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SarifUploadCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const validation_1 = require("../../helpers/validation");
const __1 = require("../..");
class SarifUploadCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.basePaths = clipanion_1.Option.Rest({ required: 1 });
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false);
        this.env = clipanion_1.Option.String('--env', 'ci');
        this.maxConcurrency = clipanion_1.Option.String('--max-concurrency', '20', { validator: (0, validation_1.isInteger)() });
        this.serviceFromCli = clipanion_1.Option.String('--service');
        this.tags = clipanion_1.Option.Array('--tags');
        this.gitPath = clipanion_1.Option.String('--git-repository');
        this.noVerify = clipanion_1.Option.Boolean('--no-verify', false);
        this.noCiTags = clipanion_1.Option.Boolean('--no-ci-tags', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.SarifUploadCommand = SarifUploadCommand;
SarifUploadCommand.paths = [['sarif', 'upload']];
SarifUploadCommand.usage = clipanion_1.Command.Usage({
    category: 'Static Analysis',
    description: 'Upload SARIF reports files to Datadog.',
    details: `
      This command will upload SARIF reports files to Datadog.\n
      See README for details.
    `,
    examples: [
        ['Upload all SARIF report files in current directory', 'datadog-ci sarif upload .'],
        [
            'Upload all SARIF report files in src/sarif-go-reports and src/sarif-java-reports',
            'datadog-ci sarif upload src/sarif-go-reports src/sarif-java-reports',
        ],
        [
            'Upload all SARIF report files in current directory and add extra tags globally',
            'datadog-ci sarif upload --tags key1:value1 --tags key2:value2 .',
        ],
        [
            'Upload all SARIF report files in current directory to the datadoghq.eu site',
            'DATADOG_SITE=datadoghq.eu datadog-ci sarif upload .',
        ],
    ],
});
