// @datadog/datadog-ci-base/dist/commands/synthetics/deploy-tests.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyntheticsDeployTestsCommand = void 0;
/* eslint-disable @typescript-eslint/member-ordering */
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const utils_1 = require("../../helpers/utils");
const __1 = require("../..");
const datadogDocsBaseUrl = 'https://docs.datadoghq.com';
// BASE COMMAND START
const $B1 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/account_management/api-app-keys`);
const $B2 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/continuous_testing/cicd_integrations/configuration#global-configuration-file`);
const $B3 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/getting_started/site/#access-the-datadog-site`);
// BASE COMMAND END
const $1 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/continuous_testing/cicd_integrations/configuration#test-files`);
class SyntheticsDeployTestsCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // BASE COMMAND START
        this.apiKey = clipanion_1.Option.String('--apiKey', {
            description: `Your Datadog API key. This key is ${$B1 `created in your Datadog organization`} and should be stored as a secret.`,
        });
        this.appKey = clipanion_1.Option.String('--appKey', {
            description: `Your Datadog application key. This key is ${$B1 `created in your Datadog organization`} and should be stored as a secret.`,
        });
        this.configPath = clipanion_1.Option.String('--config', {
            description: `The path to the ${$B2 `global configuration file`} that configures datadog-ci.`,
        });
        this.datadogSite = clipanion_1.Option.String('--datadogSite', {
            description: `Your Datadog site. Possible values are listed ${$B3 `in this table`}.`,
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        // BASE COMMAND END
        this.files = clipanion_1.Option.Array('-f,--files', {
            description: `Glob patterns to detect Synthetic ${$1 `test configuration files`}}.`,
        });
        this.publicIds = clipanion_1.Option.Array('-p,--public-id', { description: 'Public IDs of Synthetic tests to deploy.' });
        this.subdomain = clipanion_1.Option.String('--subdomain', {
            description: 'The custom subdomain to access your Datadog organization. If your URL is `myorg.datadoghq.com`, the custom subdomain is `myorg`.',
        });
        this.excludeFields = clipanion_1.Option.Array('--exclude-field', {
            description: 'Fields to exclude from partial updates, to avoid breaking Main Test Definitions with data specific to Local Test Definitions, like the Start URL. By default, all fields inside `config` are excluded.',
        });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.SyntheticsDeployTestsCommand = SyntheticsDeployTestsCommand;
SyntheticsDeployTestsCommand.paths = [['synthetics', 'deploy-tests']];
SyntheticsDeployTestsCommand.usage = clipanion_1.Command.Usage({
    category: 'Synthetics',
    description: 'Deploy Local Test Definitions as Main Test Definitions in Datadog.',
    details: `
      This command deploys Local Test Definitions as Main Test Definitions in Datadog, usually when a feature branch is merged or during a deployment.
    `,
    examples: [
        [
            'Explicitly specify the local test definitions to deploy',
            'datadog-ci synthetics deploy-tests --public-id pub-lic-id1 --public-id pub-lic-id2',
        ],
        [
            'Override the default glob pattern',
            'datadog-ci synthetics deploy-tests -f ./component-1/**/*.synthetics.json -f ./component-2/**/*.synthetics.json',
        ],
    ],
});
