// @datadog/datadog-ci-base/dist/commands/synthetics/import-tests.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyntheticsImportTestsCommand = void 0;
/* eslint-disable @typescript-eslint/member-ordering */
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const utils_1 = require("../../helpers/utils");
const __1 = require("../..");
const datadogDocsBaseUrl = 'https://docs.datadoghq.com';
const datadogAppBaseUrl = 'https://app.datadoghq.com';
// BASE COMMAND START
const $B1 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/account_management/api-app-keys`);
const $B2 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/continuous_testing/cicd_integrations/configuration#global-configuration-file`);
const $B3 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/getting_started/site/#access-the-datadog-site`);
// BASE COMMAND END
const $1 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/continuous_testing/cicd_integrations/configuration#test-files`);
const $2 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/synthetics/explore/#search`);
const $3 = (0, utils_1.makeTerminalLink)(`${datadogAppBaseUrl}/synthetics/tests`);
class SyntheticsImportTestsCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // BASE COMMAND START
        this.apiKey = clipanion_1.Option.String('--apiKey', {
            description: `Your Datadog API key. This key is ${$B1 `created in your Datadog organization`} and should be stored as a secret.`,
        });
        this.appKey = clipanion_1.Option.String('--appKey', {
            description: `Your Datadog application key. This key is ${$B1 `created in your Datadog organization`} and should be stored as a secret.`,
        });
        this.configPath = clipanion_1.Option.String('--config', {
            description: `The path to the ${$B2 `global configuration file`} that configures datadog-ci.`,
        });
        this.datadogSite = clipanion_1.Option.String('--datadogSite', {
            description: `Your Datadog site. Possible values are listed ${$B3 `in this table`}.`,
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        // BASE COMMAND END
        // TODO: Let's not reuse `files` as it has a different meaning.
        this.files = clipanion_1.Option.Array('-f,--files', {
            description: `The path to the Synthetic ${$1 `test configuration file`} to which to append imported Local Test Definitions.`,
        });
        this.publicIds = clipanion_1.Option.Array('-p,--public-id', { description: 'Public IDs of Synthetic tests to import.' });
        this.testSearchQuery = clipanion_1.Option.String('-s,--search', {
            description: `Use a ${$2 `search query`} to select which Synthetic tests to import. Use the ${$3 `Synthetic Tests list page's search bar`} to craft your query, then copy and paste it.`,
        });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.SyntheticsImportTestsCommand = SyntheticsImportTestsCommand;
SyntheticsImportTestsCommand.paths = [['synthetics', 'import-tests']];
SyntheticsImportTestsCommand.usage = clipanion_1.Command.Usage({
    category: 'Synthetics',
    description: 'Import the Main Test Definition from a Datadog scheduled test as a Local Test Definition.',
    details: `
      This command imports a Main Test Definition from a Datadog scheduled test as a Local Test Definition to be used in local development.
    `,
    examples: [
        [
            'Explicitly specify multiple tests to run',
            'datadog-ci synthetics import-tests --public-id pub-lic-id1 --public-id pub-lic-id2',
        ],
        ['Override the default glob pattern', 'datadog-ci synthetics import-tests -f test-file.synthetics.json'],
    ],
});
