// @datadog/datadog-ci-base/dist/commands/synthetics/run-tests.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyntheticsRunTestsCommand = void 0;
/* eslint-disable @typescript-eslint/member-ordering */
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const utils_1 = require("../../helpers/utils");
const validation = __importStar(require("../../helpers/validation"));
const __1 = require("../..");
const datadogDocsBaseUrl = 'https://docs.datadoghq.com';
const datadogAppBaseUrl = 'https://app.datadoghq.com';
// BASE COMMAND START
const $B1 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/account_management/api-app-keys`);
const $B2 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/continuous_testing/cicd_integrations/configuration#global-configuration-file`);
const $B3 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/getting_started/site/#access-the-datadog-site`);
// BASE COMMAND END
const $1 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/continuous_testing/cicd_integrations/configuration#test-files`);
const $2 = (0, utils_1.makeTerminalLink)(`${datadogAppBaseUrl}/synthetics/settings/continuous-testing`);
const $3 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/synthetics/explore/#search`);
const $4 = (0, utils_1.makeTerminalLink)(`${datadogAppBaseUrl}/synthetics/tests`);
const $5 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/continuous_testing/environments/proxy_firewall_vpn#what-is-the-testing-tunnel`);
const $6 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/synthetics/mobile_app_testing/`);
class SyntheticsRunTestsCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // JUnit options
        this.jUnitReport = clipanion_1.Option.String('-j,--jUnitReport', {
            description: 'The filename for a JUnit report if you want to generate one.',
        });
        this.runName = clipanion_1.Option.String('-n,--runName', {
            description: 'A name for this run, which will be included in the JUnit report.',
        });
        // BASE COMMAND START
        this.apiKey = clipanion_1.Option.String('--apiKey', {
            description: `Your Datadog API key. This key is ${$B1 `created in your Datadog organization`} and should be stored as a secret.`,
        });
        this.appKey = clipanion_1.Option.String('--appKey', {
            description: `Your Datadog application key. This key is ${$B1 `created in your Datadog organization`} and should be stored as a secret.`,
        });
        this.configPath = clipanion_1.Option.String('--config', {
            description: `The path to the ${$B2 `global configuration file`} that configures datadog-ci.`,
        });
        this.datadogSite = clipanion_1.Option.String('--datadogSite', {
            description: `Your Datadog site. Possible values are listed ${$B3 `in this table`}.`,
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        // BASE COMMAND END
        this.batchTimeout = clipanion_1.Option.String('--batchTimeout', {
            description: 'The duration in milliseconds after which the CI batch fails as timed out. This does not affect the outcome of a test run that already started.',
            validator: validation.isInteger(),
        });
        this.failOnCriticalErrors = clipanion_1.Option.Boolean('--failOnCriticalErrors', {
            description: 'Fail the CI job if a critical error that is typically transient occurs, such as rate limits, authentication failures, or Datadog infrastructure issues.',
        });
        this.failOnMissingTests = clipanion_1.Option.Boolean('--failOnMissingTests', {
            description: `Fail the CI job if the list of tests to run is empty or if some explicitly listed tests are missing.`,
        });
        this.failOnTimeout = clipanion_1.Option.Boolean('--failOnTimeout', {
            description: 'A boolean flag that fails the CI job if at least one test exceeds the default test timeout.',
        });
        this.files = clipanion_1.Option.Array('-f,--files', {
            description: `Glob patterns to detect Synthetic ${$1 `test configuration files`}}.`,
        });
        this.mobileApplicationVersion = clipanion_1.Option.String('--mobileApplicationVersion', {
            description: `Override the mobile application version for ${$6 `Synthetic mobile application tests`}. The version must be uploaded and available within Datadog.`,
        });
        this.mobileApplicationVersionFilePath = clipanion_1.Option.String('--mobileApp,--mobileApplicationVersionFilePath', {
            description: `Override the mobile application version for ${$6 `Synthetic mobile application tests`} with a local or recently built application.`,
        });
        this.overrides = clipanion_1.Option.Array('--override', {
            description: 'Override specific test properties.',
        });
        this.publicIds = clipanion_1.Option.Array('-p,--public-id', {
            description: `Public IDs of Synthetic tests to run. If no value is provided, tests are discovered in Synthetic ${$1 `test configuration files`}.`,
        });
        this.selectiveRerun = clipanion_1.Option.Boolean('--selectiveRerun', {
            description: `Whether to only rerun failed tests. If a test has already passed for a given commit, it will not be rerun in subsequent CI batches. By default, your ${$2 `organization's default setting`} is used. Set it to \`false\` to force full runs when your configuration enables it by default.`,
        });
        this.subdomain = clipanion_1.Option.String('--subdomain', {
            description: 'The custom subdomain to access your Datadog organization. If your URL is `myorg.datadoghq.com`, the custom subdomain is `myorg`.',
        });
        this.testSearchQuery = clipanion_1.Option.String('-s,--search', {
            description: `Use a ${$3 `search query`} to select which Synthetic tests to run. Use the ${$4 `Synthetic Tests list page's search bar`} to craft your query, then copy and paste it.`,
        });
        this.tunnel = clipanion_1.Option.Boolean('-t,--tunnel', {
            description: `Use the ${$5 `Continuous Testing tunnel`} to launch tests against internal environments.`,
        });
        this.buildCommand = clipanion_1.Option.String('--buildCommand', {
            description: 'The build command to generate the assets to run the tests against.',
        });
        this.dryRun = clipanion_1.Option.Boolean('--dryRun,--dry-run', {
            description: 'List the tests that would be triggered without actually running them.',
        });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.SyntheticsRunTestsCommand = SyntheticsRunTestsCommand;
SyntheticsRunTestsCommand.paths = [
    ['synthetics', 'run-tests'],
    ['synthetics', 'build-and-test'],
];
SyntheticsRunTestsCommand.usage = clipanion_1.Command.Usage({
    category: 'Synthetics',
    description: 'Run Synthetic tests with Datadog.',
    details: `
      This command will run Synthetic tests with Datadog, wait for the results and report a summary.\n
      https://docs.datadoghq.com/continuous_testing/cicd_integrations
    `,
    examples: [
        [
            'Explicitly specify multiple tests to run',
            'datadog-ci synthetics run-tests --public-id pub-lic-id1 --public-id pub-lic-id2',
        ],
        ['Discover tests with a search query', "datadog-ci synthetics run-tests --search 'tag:e2e-tests'"],
        [
            'Override the default glob pattern to group the tests in suites',
            'datadog-ci synthetics run-tests -f ./component-1/**/*.synthetics.json -f ./component-2/**/*.synthetics.json',
        ],
        [
            'Override existing or inject new local and global variables in tests',
            'datadog-ci synthetics run-tests -f ./component-1/**/*.synthetics.json --override variables.NAME=VALUE',
        ],
    ],
});
