// @datadog/datadog-ci-base/dist/commands/synthetics/upload-application.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyntheticsUploadApplicationCommand = void 0;
/* eslint-disable @typescript-eslint/member-ordering */
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const utils_1 = require("../../helpers/utils");
const __1 = require("../..");
const datadogDocsBaseUrl = 'https://docs.datadoghq.com';
// BASE COMMAND START
const $B1 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/account_management/api-app-keys`);
const $B2 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/continuous_testing/cicd_integrations/configuration#global-configuration-file`);
const $B3 = (0, utils_1.makeTerminalLink)(`${datadogDocsBaseUrl}/getting_started/site/#access-the-datadog-site`);
// BASE COMMAND END
class SyntheticsUploadApplicationCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // BASE COMMAND START
        this.apiKey = clipanion_1.Option.String('--apiKey', {
            description: `Your Datadog API key. This key is ${$B1 `created in your Datadog organization`} and should be stored as a secret.`,
        });
        this.appKey = clipanion_1.Option.String('--appKey', {
            description: `Your Datadog application key. This key is ${$B1 `created in your Datadog organization`} and should be stored as a secret.`,
        });
        this.configPath = clipanion_1.Option.String('--config', {
            description: `The path to the ${$B2 `global configuration file`} that configures datadog-ci.`,
        });
        this.datadogSite = clipanion_1.Option.String('--datadogSite', {
            description: `Your Datadog site. Possible values are listed ${$B3 `in this table`}.`,
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        // BASE COMMAND END
        this.mobileApplicationVersionFilePath = clipanion_1.Option.String('--mobileApp,--mobileApplicationVersionFilePath', {
            description: 'The path to the new version of your mobile application (`.apk` or `.ipa`).',
        });
        this.mobileApplicationId = clipanion_1.Option.String('--mobileApplicationId', {
            description: 'The ID of the application you want to upload the new version to.',
        });
        this.versionName = clipanion_1.Option.String('--versionName', {
            description: 'The name of the new version. It has to be unique.',
        });
        this.latest = clipanion_1.Option.Boolean('--latest', {
            description: 'Mark the new version as `latest`. Any tests that run on the latest version will use this version on their next run.',
        });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.SyntheticsUploadApplicationCommand = SyntheticsUploadApplicationCommand;
SyntheticsUploadApplicationCommand.paths = [['synthetics', 'upload-application']];
SyntheticsUploadApplicationCommand.usage = clipanion_1.Command.Usage({
    category: 'Synthetics',
    description: 'Upload a new version to an existing mobile application in Datadog.',
    details: `
      This command will upload a \`.apk\` or \`.ipa\` file as a new version for a given application, which already exists in Datadog.\n
      https://docs.datadoghq.com/mobile_app_testing/mobile_app_tests
    `,
    examples: [
        [
            'Upload version `example 1.0` and mark it as latest',
            "datadog-ci synthetics upload-application --mobileApplicationId '123-123-123' --mobileApplicationVersionFilePath example/test.apk --versionName 'example 1.0' --latest",
        ],
    ],
});
