// @datadog/datadog-ci-base/dist/commands/terraform/upload.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformUploadCommand = void 0;
const clipanion_1 = require("clipanion");
const t = __importStar(require("typanion"));
const plugin_1 = require("@datadog/datadog-ci-base/helpers/plugin");
const __1 = require("../..");
class TerraformUploadCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // Artifact type: 'plan' or 'state'
        this.artifactType = clipanion_1.Option.String({ required: true, validator: t.isEnum(['plan', 'state']) });
        // File paths to upload (one or more)
        this.filePaths = clipanion_1.Option.Rest({ required: 1 });
        // Optional repo ID override
        this.repoId = clipanion_1.Option.String('--repo-id', {
            description: 'Repository identifier override (e.g., github.com/datadog/my-repo)',
        });
        // Skip git metadata upload
        this.skipGitMetadataUpload = clipanion_1.Option.Boolean('--skip-git-metadata-upload', false, {
            description: 'Skip the upload of git metadata',
        });
        // Verbose logging
        this.verbose = clipanion_1.Option.Boolean('--verbose', false, {
            description: 'Enable verbose logging',
        });
        // FIPS mode
        this.fips = clipanion_1.Option.Boolean('--fips', false, { description: 'Enable FIPS mode for the command' });
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false, {
            description: 'Ignore FIPS errors and continue execution',
        });
        // Dry run mode
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false, {
            description: 'Run the command in dry run mode, without uploading any data to Datadog',
        });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.TerraformUploadCommand = TerraformUploadCommand;
TerraformUploadCommand.paths = [['terraform', 'upload']];
TerraformUploadCommand.usage = clipanion_1.Command.Usage({
    category: 'CI Visibility',
    description: 'Upload Terraform plan or state files to Datadog.',
    details: `
      This command uploads Terraform runtime artifacts (plan or state JSON) to Datadog
      for enhanced cloud-to-code mapping and policy evaluation.\n
      See README for details.
    `,
    examples: [
        ['Upload a Terraform plan file', 'datadog-ci terraform upload plan terraform-plan.json'],
        ['Upload multiple Terraform plan files', 'datadog-ci terraform upload plan plan1.json plan2.json plan3.json'],
        ['Upload a Terraform state file', 'datadog-ci terraform upload state terraform.tfstate'],
        ['Upload with verbose output', 'datadog-ci terraform upload plan terraform-plan.json --verbose'],
        ['Dry run mode', 'datadog-ci terraform upload plan terraform-plan.json --dry-run'],
    ],
});
