// @datadog/datadog-ci-base/dist/commands/sbom/upload.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SbomUploadCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const __1 = require("../..");
class SbomUploadCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // BASE COMMAND START
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        // BASE COMMAND END
        this.basePath = clipanion_1.Option.String();
        this.serviceFromCli = clipanion_1.Option.String('--service');
        this.env = clipanion_1.Option.String('--env', 'ci');
        this.tags = clipanion_1.Option.Array('--tags');
        this.gitPath = clipanion_1.Option.String('--git-repository');
        this.debug = clipanion_1.Option.Boolean('--debug');
        this.noCiTags = clipanion_1.Option.Boolean('--no-ci-tags', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.SbomUploadCommand = SbomUploadCommand;
SbomUploadCommand.paths = [['sbom', 'upload']];
SbomUploadCommand.usage = clipanion_1.Command.Usage({
    category: 'Static Analysis',
    description: 'Upload SBOM files to Datadog.',
    details: `
      This command uploads SBOM files to Datadog for dependency tracking, vulnerability analysis, and compliance auditing.
    `,
    examples: [['Upload the SBOM file: sbom.json', 'datadog-ci sbom upload sbom.json']],
});
