// @datadog/datadog-ci-base/dist/commands/container-app/common.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerAppCommand = void 0;
const clipanion_1 = require("clipanion");
const constants_1 = require("../../constants");
const env_1 = require("../../helpers/env");
const fips_1 = require("../../helpers/fips");
const renderer_1 = require("../../helpers/renderer");
const azure_1 = require("../../helpers/serverless/azure");
const constants_2 = require("../../helpers/serverless/constants");
const utils_1 = require("../../helpers/utils");
const __1 = require("../..");
class ContainerAppCommand extends __1.BaseCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.subscriptionId = clipanion_1.Option.String('-s,--subscription-id', {
            description: 'Subscription ID of the Azure subscription containing the Container App. Must be used with `--resource-group` and `--name`.',
        });
        this.resourceGroup = clipanion_1.Option.String('-g,--resource-group', {
            description: 'Name of the Azure Resource Group containing the Container App. Must be used with `--subscription-id` and `--name`.',
        });
        this.containerAppName = clipanion_1.Option.String('-n,--name', {
            description: 'Name of the Azure Container App to instrument. Must be used with `--subscription-id` and `--resource-group`.',
        });
        this.resourceIds = clipanion_1.Option.Array('-r,--resource-id', {
            description: 'Full Azure resource ID to instrument. Can be specified multiple times. Format: `/subscriptions/<subscription-id>/resourceGroups/<resource-group-name>/providers/Microsoft.App/containerApps/<container-app-name>`.',
        });
        this.envVars = clipanion_1.Option.Array('-e,--env-vars', {
            description: 'Additional environment variables to set for the Container App. Can specify multiple variables in the format `--env-vars VAR1=VALUE1 --env-vars VAR2=VALUE2`.',
        });
        this.configPath = clipanion_1.Option.String('--config', {
            description: 'Path to the configuration file.',
        });
        // eslint-disable-next-line @typescript-eslint/member-ordering -- needed for ordering of arguments in readme
        this.dryRun = clipanion_1.Option.Boolean('-d,--dry-run', false, {
            description: 'Run the command in dry-run mode, without making any changes. Preview the changes that running the command would apply.',
        });
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.fipsConfig = {
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
    }
    get dryRunPrefix() {
        return this.dryRun ? renderer_1.dryRunTag + ' ' : '';
    }
    get additionalConfig() {
        return {};
    }
    enableFips() {
        (0, fips_1.enableFips)(this.fips || this.fipsConfig.fips, this.fipsIgnoreError || this.fipsConfig.fipsIgnoreError);
    }
    ensureConfig() {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            const config = (yield (0, utils_1.resolveConfigFromFile)({
                containerApp: Object.assign({ subscriptionId: this.subscriptionId, resourceGroup: this.resourceGroup, containerAppName: this.containerAppName, envVars: this.envVars }, this.additionalConfig),
            }, {
                configPath: this.configPath,
                defaultConfigPaths: utils_1.DEFAULT_CONFIG_PATHS,
            })).containerApp;
            const containerApps = {};
            const errors = [];
            if (process.env.DD_API_KEY === undefined) {
                errors.push('DD_API_KEY environment variable is required');
            }
            // Validate that envVars, if provided, are in the format 'key=value'
            if ((_a = config.envVars) === null || _a === void 0 ? void 0 : _a.some((e) => !constants_2.ENV_VAR_REGEX.test(e))) {
                errors.push('All envVars must be in the format `KEY=VALUE`');
            }
            // Validate that extraTags, if provided, comply with the expected format
            if (config.extraTags && !config.extraTags.match(constants_2.EXTRA_TAGS_REG_EXP)) {
                errors.push('Extra tags do not comply with the <key>:<value> array.');
            }
            // Validate that logsPath starts with sharedVolumePath
            if (!config.logsPath || !config.sharedVolumePath) {
                errors.push('logsPath and sharedVolumePath must be non-empty when instance logging is enabled');
            }
            else if (!config.logsPath.startsWith(config.sharedVolumePath)) {
                errors.push('logsPath must start with sharedVolumePath when instance logging is enabled');
            }
            const specifiedAppArgs = [config.subscriptionId, config.resourceGroup, config.containerAppName];
            // all or none of the app args should be specified
            if (!(specifiedAppArgs.every((arg) => arg) || specifiedAppArgs.every((arg) => !arg))) {
                errors.push('--subscription-id, --resource-group, and --name must be specified together or not at all');
            }
            else if (specifiedAppArgs.every((arg) => arg)) {
                containerApps[config.subscriptionId] = { [config.resourceGroup]: [config.containerAppName] };
            }
            if ((_b = this.resourceIds) === null || _b === void 0 ? void 0 : _b.length) {
                for (const resourceId of this.resourceIds) {
                    const parsed = (0, azure_1.parseResourceId)(resourceId);
                    if (!parsed || parsed.subType) {
                        errors.push(`Invalid Container App resource ID: ${resourceId}`);
                        continue;
                    }
                    const { subscriptionId, resourceGroup, name } = parsed;
                    if (!containerApps[subscriptionId]) {
                        containerApps[subscriptionId] = {};
                    }
                    if (!containerApps[subscriptionId][resourceGroup]) {
                        containerApps[subscriptionId][resourceGroup] = [];
                    }
                    containerApps[subscriptionId][resourceGroup].push(name);
                }
            }
            if (!((_c = this.resourceIds) === null || _c === void 0 ? void 0 : _c.length) && specifiedAppArgs.every((arg) => !arg)) {
                errors.push('No Container Apps specified to instrument');
            }
            return [containerApps, config, errors];
        });
    }
}
exports.ContainerAppCommand = ContainerAppCommand;
