// @datadog/datadog-ci-base/dist/commands/container-app/instrument.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerAppInstrumentCommand = exports.DEFAULT_SIDECAR_MEMORY = exports.DEFAULT_SIDECAR_CPU = void 0;
const clipanion_1 = require("clipanion");
const typanion_1 = require("typanion");
const plugin_1 = require("../../helpers/plugin");
const constants_1 = require("../../helpers/serverless/constants");
const common_1 = require("./common");
exports.DEFAULT_SIDECAR_CPU = 0.5;
exports.DEFAULT_SIDECAR_MEMORY = 1;
class ContainerAppInstrumentCommand extends common_1.ContainerAppCommand {
    constructor() {
        super(...arguments);
        this.service = clipanion_1.Option.String('--service', {
            description: 'The value for the service tag. Use this to group related Container Apps belonging to similar workloads. For example, `my-service`. If not provided, the Container App name is used.',
        });
        this.environment = clipanion_1.Option.String('--env,--environment', {
            description: 'The value for the env tag. Use this to separate your staging, development, and production environments. For example, `prod`.',
        });
        this.version = clipanion_1.Option.String('--version', {
            description: 'The value for the version tag. Use this to correlate spikes in latency, load, or errors to new versions. For example, `1.0.0`.',
        });
        this.sidecarName = clipanion_1.Option.String('--sidecar-name', constants_1.DEFAULT_SIDECAR_NAME, {
            description: `(Not recommended) The name to use for the sidecar container. Defaults to '${constants_1.DEFAULT_SIDECAR_NAME}'`,
        });
        this.sharedVolumeName = clipanion_1.Option.String('--shared-volume-name', constants_1.DEFAULT_VOLUME_NAME, {
            description: `(Not recommended) Specify a custom shared volume name. Defaults to '${constants_1.DEFAULT_VOLUME_NAME}'`,
        });
        this.sharedVolumePath = clipanion_1.Option.String('--shared-volume-path', constants_1.DEFAULT_VOLUME_PATH, {
            description: `(Not recommended) Specify a custom shared volume path. Defaults to '${constants_1.DEFAULT_VOLUME_PATH}'`,
        });
        this.logsPath = clipanion_1.Option.String('--logs-path', constants_1.DEFAULT_LOGS_PATH, {
            description: `(Not recommended) Specify a custom log file path. Must begin with the shared volume path. Defaults to '${constants_1.DEFAULT_LOGS_PATH}'`,
        });
        this.sidecarCpu = clipanion_1.Option.String('--sidecar-cpu', {
            description: `The number of CPUs to allocate to the sidecar container. Defaults to '${exports.DEFAULT_SIDECAR_CPU}'.`,
            validator: (0, typanion_1.isNumber)(),
        });
        this.sidecarMemory = clipanion_1.Option.String('--sidecar-memory', {
            description: `The amount of memory (in GiB) to allocate to the sidecar container. Defaults to '${exports.DEFAULT_SIDECAR_MEMORY}'.`,
            validator: (0, typanion_1.isNumber)(),
        });
        this.sourceCodeIntegration = clipanion_1.Option.Boolean('--source-code-integration,--sourceCodeIntegration', true, {
            description: 'Whether to enable the Datadog Source Code integration. This tags your service(s) with the Git repository and the latest commit hash of the local directory. Specify `--no-source-code-integration` to disable.',
        });
        this.uploadGitMetadata = clipanion_1.Option.Boolean('--upload-git-metadata,--uploadGitMetadata', true, {
            description: "Whether to enable Git metadata uploading, as a part of the source code integration. Git metadata uploading is only required if you don't have the Datadog GitHub integration installed. Specify `--no-upload-git-metadata` to disable.",
        });
        this.extraTags = clipanion_1.Option.String('--extra-tags,--extraTags', {
            description: 'Additional tags to add to the app in the format "key1:value1,key2:value2".',
        });
    }
    get additionalConfig() {
        return {
            service: this.service,
            environment: this.environment,
            version: this.version,
            sidecarName: this.sidecarName,
            sharedVolumeName: this.sharedVolumeName,
            sharedVolumePath: this.sharedVolumePath,
            logsPath: this.logsPath,
            sidecarCpu: this.sidecarCpu,
            sidecarMemory: this.sidecarMemory,
            sourceCodeIntegration: this.sourceCodeIntegration,
            uploadGitMetadata: this.uploadGitMetadata,
            extraTags: this.extraTags,
        };
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.ContainerAppInstrumentCommand = ContainerAppInstrumentCommand;
ContainerAppInstrumentCommand.paths = [['container-app', 'instrument']];
ContainerAppInstrumentCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Apply Datadog instrumentation to an Azure Container App.',
});
