// @datadog/datadog-ci-base/dist/commands/container-app/uninstrument.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerAppUninstrumentCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("../../helpers/plugin");
const constants_1 = require("../../helpers/serverless/constants");
const common_1 = require("./common");
class ContainerAppUninstrumentCommand extends common_1.ContainerAppCommand {
    constructor() {
        super(...arguments);
        this.sidecarName = clipanion_1.Option.String('--sidecar-name', constants_1.DEFAULT_SIDECAR_NAME, {
            description: `The name of the sidecar container to remove. Specify if you have a different sidecar name. Defaults to '${constants_1.DEFAULT_SIDECAR_NAME}'`,
        });
        this.sharedVolumeName = clipanion_1.Option.String('--shared-volume-name', constants_1.DEFAULT_VOLUME_NAME, {
            description: `The name of the shared volume to remove. Specify if you have a different shared volume name. Defaults to '${constants_1.DEFAULT_VOLUME_NAME}'`,
        });
    }
    get additionalConfig() {
        return {
            sidecarName: this.sidecarName,
            sharedVolumeName: this.sharedVolumeName,
        };
    }
    ensureConfig() {
        const _super = Object.create(null, {
            ensureConfig: { get: () => super.ensureConfig }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const [containerApps, config, errors] = yield _super.ensureConfig.call(this);
            // Remove errors that are specific to instrumentation
            const filteredErrors = errors.filter((error) => !error.includes('DD_API_KEY') &&
                !error.includes('logsPath') &&
                !error.includes('sharedVolumePath') &&
                !error.includes('Extra tags'));
            return [containerApps, config, filteredErrors];
        });
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.ContainerAppUninstrumentCommand = ContainerAppUninstrumentCommand;
ContainerAppUninstrumentCommand.paths = [['container-app', 'uninstrument']];
ContainerAppUninstrumentCommand.usage = clipanion_1.Command.Usage({
    category: 'Serverless',
    description: 'Revert Datadog instrumentation in an Azure Container App.',
});
