// @datadog/datadog-ci-plugin-sarif/dist/commands/upload.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginCommand = void 0;
const fs_1 = __importDefault(require("fs"));
const upload_1 = require("@datadog/datadog-ci-base/commands/sarif/upload");
const constants_1 = require("@datadog/datadog-ci-base/constants");
const concurrency_1 = require("@datadog/datadog-ci-base/helpers/concurrency");
const env_1 = require("@datadog/datadog-ci-base/helpers/env");
const fips_1 = require("@datadog/datadog-ci-base/helpers/fips");
const glob_1 = require("@datadog/datadog-ci-base/helpers/glob");
const retry_1 = require("@datadog/datadog-ci-base/helpers/retry");
const tags_1 = require("@datadog/datadog-ci-base/helpers/tags");
const utils_1 = require("@datadog/datadog-ci-base/helpers/utils");
const chalk_1 = __importDefault(require("chalk"));
const upath_1 = __importDefault(require("upath"));
const api_1 = require("../api");
const renderer_1 = require("../renderer");
const utils_2 = require("../utils");
const validation_1 = require("../validation");
class PluginCommand extends upload_1.SarifUploadCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.config = {
            apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
            env: process.env.DD_ENV,
            envVarTags: process.env.DD_TAGS,
        };
        this.fipsConfig = {
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            (0, fips_1.enableFips)(this.fips || this.fipsConfig.fips, this.fipsIgnoreError || this.fipsConfig.fipsIgnoreError);
            const githubEvent = process.env.GITHUB_EVENT_NAME;
            const gitlabEvent = process.env.CI_PIPELINE_SOURCE;
            const azureReason = process.env.BUILD_REASON;
            if (githubEvent === 'pull_request') {
                // https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-commands#example-setting-an-error-message
                this.context.stdout.write('::error title=Unsupported Trigger::The `pull_request` event is not supported by Datadog Code Security and will cause issues with the product. ' +
                    'To continue using Datadog Code Security, use `push` instead. See: https://docs.datadoghq.com/security/code_security/static_analysis/github_actions/#workflow for more information.');
                return 1;
            }
            if (gitlabEvent === 'merge_request_event') {
                this.context.stderr.write('The `merge_request_event` pipeline source is not supported by Datadog Code Security and will cause issues with the product. ' +
                    'To continue using Datadog Code Security, use `push` instead. See: https://docs.datadoghq.com/security/code_security/static_analysis/generic_ci_providers/ for more information.');
                return 1;
            }
            if (azureReason === 'PullRequest') {
                // https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#logging-commands-for-build-pipelines
                this.context.stdout.write('##vso[task.logissue type=error]The `PullRequest` build reason is not supported by Datadog Code Security and will cause issues with the product. ' +
                    'To continue using Datadog Code Security, use `push` instead. See: https://docs.datadoghq.com/security/code_security/static_analysis/generic_ci_providers/ for more information.');
                return 1;
            }
            if (!this.basePaths || !this.basePaths.length) {
                this.context.stderr.write('Missing basePath\n');
                return 1;
            }
            if (!this.config.env) {
                this.config.env = this.env;
            }
            const api = this.getApiHelper();
            // Normalizing the basePath to resolve .. and .
            this.basePaths = this.basePaths.map((basePath) => upath_1.default.normalize(basePath));
            const spanTags = yield (0, tags_1.getSpanTags)(this.config, this.tags, !this.noCiTags, this.gitPath);
            // Gather any missing mandatory git fields to display to the user
            const missingGitFields = (0, tags_1.getMissingRequiredGitTags)(spanTags);
            if (missingGitFields.length > 0) {
                this.context.stdout.write((0, renderer_1.renderMissingTags)(missingGitFields));
                return 1;
            }
            const payloads = yield this.getMatchingSarifReports(spanTags);
            if (payloads.length === 0) {
                this.context.stdout.write((0, renderer_1.renderFilesNotFound)(this.basePaths));
                return 1;
            }
            const sha = spanTags[tags_1.GIT_SHA] || 'sha-not-found';
            const env = this.config.env || 'env-not-set';
            this.context.stdout.write((0, renderer_1.renderCommandInfo)(this.basePaths, env, sha, this.maxConcurrency, this.dryRun, this.noVerify));
            const upload = (payload) => this.uploadSarifReport(api, payload);
            const initialTime = new Date().getTime();
            yield (0, concurrency_1.doWithMaxConcurrency)(this.maxConcurrency, payloads, upload);
            const totalTimeSeconds = (Date.now() - initialTime) / 1000;
            this.context.stdout.write((0, renderer_1.renderSuccessfulCommand)(payloads.length, totalTimeSeconds));
        });
    }
    uploadSarifReport(api, sarifReport) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.dryRun) {
                this.context.stdout.write((0, renderer_1.renderDryRunUpload)(sarifReport));
                return;
            }
            try {
                yield (0, retry_1.retryRequest)(() => api.uploadSarifReport(sarifReport, this.context.stdout.write.bind(this.context.stdout)), {
                    onRetry: (e, attempt) => {
                        this.context.stderr.write((0, renderer_1.renderRetriedUpload)(sarifReport, e.message, attempt));
                    },
                    retries: 5,
                });
            }
            catch (error) {
                this.context.stderr.write((0, renderer_1.renderFailedUpload)(sarifReport, error));
                throw error;
            }
        });
    }
    getApiHelper() {
        if (!this.config.apiKey) {
            this.context.stdout.write(`Neither ${chalk_1.default.red.bold('DATADOG_API_KEY')} nor ${chalk_1.default.red.bold('DD_API_KEY')} is in your environment.\n`);
            throw new Error('API key is missing');
        }
        return (0, api_1.apiConstructor)((0, utils_2.getBaseIntakeUrl)(), this.config.apiKey);
    }
    getMatchingSarifReports(spanTags) {
        return __awaiter(this, void 0, void 0, function* () {
            const sarifReports = (this.basePaths || []).reduce((acc, basePath) => {
                const isFile = !!upath_1.default.extname(basePath);
                if (isFile) {
                    return acc.concat(fs_1.default.existsSync(basePath) ? [basePath] : []);
                }
                return acc.concat((0, glob_1.globSync)((0, utils_1.buildPath)(basePath, '*.sarif'), { dotRelative: true }));
            }, []);
            const validUniqueFiles = [...new Set(sarifReports)].filter((sarifReport) => {
                if (this.noVerify) {
                    return true;
                }
                const validationErrorMessage = (0, validation_1.validateSarif)(sarifReport);
                if (validationErrorMessage) {
                    this.context.stdout.write((0, renderer_1.renderInvalidFile)(sarifReport, [validationErrorMessage]));
                    return false;
                }
                const potentialErrors = (0, validation_1.checkForError)(sarifReport);
                if (potentialErrors.length > 0) {
                    this.context.stdout.write((0, renderer_1.renderInvalidFile)(sarifReport, potentialErrors));
                    return false;
                }
                return true;
            });
            return validUniqueFiles.map((sarifReport) => {
                return {
                    reportPath: sarifReport,
                    spanTags,
                    service: (0, utils_2.getServiceFromSarifTool)(sarifReport),
                };
            });
        });
    }
}
exports.PluginCommand = PluginCommand;
