// @datadog/datadog-ci-plugin-coverage/dist/file-fixes.js
"use strict";
// File fixes identifies lines that are not executable (comments, blank lines, brackets, etc.)
// so the backend can exclude them from code coverage calculations, reducing false negatives.
// Each file's result is a bitmap where bit i (0-indexed) represents line i+1 (1-indexed).
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateFileFixes = void 0;
const promises_1 = __importDefault(require("fs/promises"));
const concurrency_1 = require("@datadog/datadog-ci-base/helpers/concurrency");
const file_finder_1 = require("@datadog/datadog-ci-base/helpers/file-finder");
const get_git_data_1 = require("@datadog/datadog-ci-base/helpers/git/get-git-data");
const upath_1 = __importDefault(require("upath"));
const MAX_FILE_SIZE = 2 * 1024 * 1024; // 2MB
const MAX_FILES = 200000;
const MAX_OUTPUT_SIZE = 20 * 1024 * 1024; // 20MB estimated serialized size
const CONCURRENCY = 8;
const SUPPORTED_EXTENSIONS = {
    go: ['.go'],
    kotlin: ['.kt', '.kts'],
    'c/cpp/swift/objc': ['.c', '.cpp', '.cc', '.cxx', '.h', '.hpp', '.hxx', '.m', '.mm', '.swift'],
    php: ['.php'],
};
const EMPTY_LINE = /^\s*$/;
const COMMENT_LINE = /^\s*\/\/.*$/;
const BRACKET_LINE = /^\s*[{}]\s*(\/\/.*)?$/;
const PARENTHESIS_LINE = /^\s*[()]\s*(\/\/.*)?$/;
const BLOCK_COMMENT_OPEN = /^\s*\/\*/;
const GO_FUNC_LINE = /^\s*func\s+(?:\([^)]*\)\s*)?[A-Za-z_][A-Za-z0-9_]*(?:\s*\[[^\]]+\])?\s*\(.*\)\s*.*\{\s*(\/\/.*)?$/;
const PHP_END_BRACKET = /^\s*\);\s*(\/\/.*)?$/;
const PHP_HEREDOC_START = /<<<\s*(['"]?)([A-Za-z_][A-Za-z0-9_]*)\1/;
const TRIPLE_QUOTED_STRING_EXTENSIONS = new Set(['.kt', '.kts', '.swift']);
const BASE_PATTERNS = [EMPTY_LINE, COMMENT_LINE, BRACKET_LINE, PARENTHESIS_LINE];
const LANGUAGE_PATTERNS = [
    {
        extensions: SUPPORTED_EXTENSIONS['go'],
        patterns: [...BASE_PATTERNS, GO_FUNC_LINE],
    },
    {
        extensions: SUPPORTED_EXTENSIONS['kotlin'],
        patterns: [...BASE_PATTERNS],
    },
    {
        extensions: SUPPORTED_EXTENSIONS['c/cpp/swift/objc'],
        patterns: [...BASE_PATTERNS],
    },
    {
        extensions: SUPPORTED_EXTENSIONS['php'],
        patterns: [...BASE_PATTERNS, PHP_END_BRACKET],
    },
];
// Build a map from extension to patterns for O(1) lookup
const EXTENSION_TO_PATTERNS = new Map();
for (const lang of LANGUAGE_PATTERNS) {
    for (const ext of lang.extensions) {
        EXTENSION_TO_PATTERNS.set(ext, lang.patterns);
    }
}
/* eslint-disable no-bitwise */
const setBit = (bitmap, index) => {
    bitmap[Math.floor(index / 8)] |= 1 << index % 8;
};
/* eslint-enable no-bitwise */
const getExtension = (filePath) => {
    return upath_1.default.extname(filePath).toLowerCase();
};
const getPatternsForFile = (filePath) => {
    return EXTENSION_TO_PATTERNS.get(getExtension(filePath));
};
const isSupportedFile = (filePath) => {
    return EXTENSION_TO_PATTERNS.has(getExtension(filePath));
};
const countOccurrences = (line, token) => {
    let count = 0;
    let startIndex = 0;
    while (true) {
        const foundIndex = line.indexOf(token, startIndex);
        if (foundIndex === -1) {
            break;
        }
        count++;
        startIndex = foundIndex + token.length;
    }
    return count;
};
const escapeRegExp = (value) => {
    return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
};
const isPhpHeredocTerminator = (line, label) => {
    const escapedLabel = escapeRegExp(label);
    return new RegExp(`^\\s*${escapedLabel};?\\s*$`).test(line);
};
const isLineInMultilineLiteral = (extension, line, state) => {
    if (extension === '.php') {
        if (state.phpHeredocLabel) {
            const heredocLabel = state.phpHeredocLabel;
            if (isPhpHeredocTerminator(line, heredocLabel)) {
                state.phpHeredocLabel = undefined;
            }
            return true;
        }
        const heredocMatch = line.match(PHP_HEREDOC_START);
        if (heredocMatch) {
            state.phpHeredocLabel = heredocMatch[2];
            return true;
        }
        return false;
    }
    if (extension === '.go') {
        const backtickCount = countOccurrences(line, '`');
        const isLiteralLine = state.insideGoRawString || backtickCount > 0;
        if (backtickCount % 2 === 1) {
            state.insideGoRawString = !state.insideGoRawString;
        }
        return isLiteralLine;
    }
    if (TRIPLE_QUOTED_STRING_EXTENSIONS.has(extension)) {
        const tripleQuoteCount = countOccurrences(line, '"""');
        const isLiteralLine = state.insideTripleQuotedString || tripleQuoteCount > 0;
        if (tripleQuoteCount % 2 === 1) {
            state.insideTripleQuotedString = !state.insideTripleQuotedString;
        }
        return isLiteralLine;
    }
    return false;
};
const isOnlyWhitespaceOrLineComment = (value) => {
    const trimmed = value.trim();
    return trimmed === '' || trimmed.startsWith('//');
};
const processFileContent = (filePath, content) => {
    const extension = getExtension(filePath);
    const patterns = getPatternsForFile(filePath);
    if (!patterns) {
        return undefined;
    }
    const lines = content.split(/\r\n|\r|\n/);
    // Remove trailing empty element produced by split when file ends with newline
    if (lines.length > 0 && lines[lines.length - 1] === '') {
        lines.pop();
    }
    const bitmapSize = Math.ceil(lines.length / 8);
    const bitmap = Buffer.alloc(bitmapSize);
    let hasMatch = false;
    let insideBlockComment = false;
    const multilineLiteralState = {
        insideGoRawString: false,
        insideTripleQuotedString: false,
    };
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        let matched = false;
        // Block comment state tracking (takes precedence to avoid string-state updates inside comments)
        if (insideBlockComment) {
            const blockCommentCloseIndex = line.indexOf('*/');
            if (blockCommentCloseIndex === -1) {
                matched = true;
            }
            else {
                const remainingLine = line.slice(blockCommentCloseIndex + 2);
                matched = isOnlyWhitespaceOrLineComment(remainingLine);
                insideBlockComment = false;
            }
        }
        else {
            const lineInMultilineLiteral = isLineInMultilineLiteral(extension, line, multilineLiteralState);
            if (!lineInMultilineLiteral && BLOCK_COMMENT_OPEN.test(line)) {
                const blockCommentCloseIndex = line.indexOf('*/');
                if (blockCommentCloseIndex === -1) {
                    matched = true;
                    insideBlockComment = true;
                }
                else {
                    const remainingLine = line.slice(blockCommentCloseIndex + 2);
                    matched = isOnlyWhitespaceOrLineComment(remainingLine);
                }
            }
            // Regular pattern matching (only if not already matched and not in multiline literal)
            if (!matched && !lineInMultilineLiteral) {
                for (const pattern of patterns) {
                    if (pattern.test(line)) {
                        matched = true;
                        break;
                    }
                }
            }
        }
        if (matched) {
            setBit(bitmap, i);
            hasMatch = true;
        }
    }
    if (!hasMatch) {
        return undefined;
    }
    return { path: filePath, totalLines: lines.length, bitmap };
};
const processFileAsync = (filePath, repoRoot) => __awaiter(void 0, void 0, void 0, function* () {
    const absolutePath = upath_1.default.resolve(repoRoot, filePath);
    // Verify resolved path stays within the repo root
    if (!absolutePath.startsWith(repoRoot + '/') && absolutePath !== repoRoot) {
        return undefined;
    }
    try {
        const stat = yield promises_1.default.lstat(absolutePath);
        if (!stat.isFile() || stat.size > MAX_FILE_SIZE) {
            return undefined;
        }
        const content = yield promises_1.default.readFile(absolutePath, 'utf8');
        return processFileContent(filePath, content);
    }
    catch (_a) {
        return undefined;
    }
});
const listFilesFromFilesystem = (rootPath) => {
    const absolutePaths = (0, file_finder_1.findFiles)([rootPath], true, // searchRecursively
    [], // ignoredPaths (DEFAULT_IGNORED_FOLDERS is applied internally)
    (filePath) => isSupportedFile(filePath), () => undefined, // no validation
    () => undefined // no error rendering
    );
    return absolutePaths.map((absPath) => upath_1.default.relative(rootPath, absPath));
};
const collectResults = (results) => {
    const fileFixes = {};
    let estimatedSize = JSON.stringify(fileFixes).length;
    for (const result of results) {
        const base64 = result.bitmap.toString('base64');
        const entrySize = result.path.length + base64.length;
        if (estimatedSize + entrySize > MAX_OUTPUT_SIZE) {
            break;
        }
        fileFixes[result.path] = { lines: result.totalLines, bitmap: base64 };
        estimatedSize += entrySize;
    }
    return fileFixes;
};
const generateFileFixes = (git, searchPath) => __awaiter(void 0, void 0, void 0, function* () {
    let repoRoot;
    let supportedFiles;
    if (searchPath) {
        // Explicit search path overrides everything
        repoRoot = upath_1.default.resolve(searchPath);
    }
    else if (git) {
        repoRoot = upath_1.default.resolve((yield git.revparse(['--show-toplevel'])).trim());
    }
    else {
        repoRoot = upath_1.default.resolve(process.cwd());
    }
    if (git && !searchPath) {
        // Use git ls-files for tracked files (respects .gitignore)
        const allFiles = yield (0, get_git_data_1.gitTrackedFiles)(git);
        supportedFiles = allFiles.filter(isSupportedFile);
    }
    else {
        // Filesystem walk fallback (when git is not available or search path is overridden)
        supportedFiles = listFilesFromFilesystem(repoRoot);
    }
    if (supportedFiles.length > MAX_FILES) {
        throw new Error(`repository has ${supportedFiles.length} supported files, exceeding the ${MAX_FILES} file limit`);
    }
    // Process files concurrently with bounded parallelism
    const allResults = yield (0, concurrency_1.doWithMaxConcurrency)(CONCURRENCY, supportedFiles, (filePath) => processFileAsync(filePath, repoRoot));
    const results = allResults.filter((r) => r !== undefined);
    return collectResults(results);
});
exports.generateFileFixes = generateFileFixes;
