// @datadog/datadog-ci-plugin-coverage/dist/renderer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderFailedGitDBSync = exports.renderSuccessfulGitDBSync = exports.renderCommandInfo = exports.renderUpload = exports.renderDryRunUpload = exports.renderSuccessfulUploadCommand = exports.renderSuccessfulUpload = exports.renderRetriedUpload = exports.renderFailedUpload = exports.renderInvalidFile = exports.renderFileReadError = void 0;
const chalk_1 = __importDefault(require("chalk"));
const upath_1 = __importDefault(require("upath"));
const utils_1 = require("./utils");
const ICONS = {
    FAILED: '❌',
    SUCCESS: '✅',
    WARNING: '⚠️',
};
const renderFileReadError = (filePath, errorMessage) => {
    const reportPath = `[${chalk_1.default.bold.dim(filePath)}]`;
    return chalk_1.default.red(`${ICONS.FAILED} Error while reading report file ${reportPath}: ${errorMessage}\n`);
};
exports.renderFileReadError = renderFileReadError;
const renderInvalidFile = (filePath, errorMessage) => {
    const reportPath = `[${chalk_1.default.bold.dim(filePath)}]`;
    return chalk_1.default.red(`${ICONS.FAILED} Invalid coverage report file ${reportPath}: ${errorMessage}\n`);
};
exports.renderInvalidFile = renderInvalidFile;
const renderFailedUpload = (payload, errorMessage) => {
    const payloadDetails = `${chalk_1.default.bold.dim(payload.paths)}`;
    return chalk_1.default.red(`${ICONS.FAILED} Upload failed for ${payloadDetails}: ${errorMessage}\n`);
};
exports.renderFailedUpload = renderFailedUpload;
const renderRetriedUpload = (payload, errorMessage, attempt) => {
    const payloadDetails = `${chalk_1.default.bold.dim(payload.paths)}`;
    return chalk_1.default.yellow(`[attempt ${attempt}] Retrying coverage report upload ${payloadDetails}: ${errorMessage}\n`);
};
exports.renderRetriedUpload = renderRetriedUpload;
const renderSuccessfulUpload = (dryRun, fileCount, duration) => {
    return chalk_1.default.green(`${dryRun ? '[DRYRUN] ' : ''}${ICONS.SUCCESS} Uploaded ${fileCount} files in ${duration} seconds.`);
};
exports.renderSuccessfulUpload = renderSuccessfulUpload;
const renderSuccessfulUploadCommand = (spanTags) => {
    const coverageDetailsUrl = (0, utils_1.getCoverageDetailsUrl)(spanTags);
    if (coverageDetailsUrl) {
        let fullStr = '';
        fullStr += chalk_1.default.green('=================================================================================================\n');
        fullStr += chalk_1.default.green('* View detailed coverage report in Datadog (it can take a few minutes to become available)\n');
        fullStr += chalk_1.default.green(`* ${coverageDetailsUrl}\n`);
        fullStr += chalk_1.default.green('=================================================================================================\n');
        return fullStr;
    }
    return '';
};
exports.renderSuccessfulUploadCommand = renderSuccessfulUploadCommand;
const renderDryRunUpload = (payload) => `[DRYRUN] ${(0, exports.renderUpload)(payload)}`;
exports.renderDryRunUpload = renderDryRunUpload;
const renderUpload = (payload) => {
    if (payload.paths && payload.paths.length) {
        let message = `Uploading code coverage report file(s) in ${payload.paths}`;
        if (payload.flags && payload.flags.length > 0) {
            message += ` with flags: ${payload.flags.join(', ')}`;
        }
        return message;
    }
    else {
        return 'No code coverage report paths, doing nothing';
    }
};
exports.renderUpload = renderUpload;
const renderCommandInfo = (reportPaths, dryRun) => {
    let fullStr = '';
    if (dryRun) {
        fullStr += chalk_1.default.yellow(`${ICONS.WARNING} DRY-RUN MODE ENABLED. WILL NOT UPLOAD COVERAGE REPORTS\n`);
    }
    fullStr += chalk_1.default.green(`${new Date().toLocaleString()} - Starting upload. \n`);
    if (!!reportPaths.length) {
        if (reportPaths.length === 1 && !!upath_1.default.extname(reportPaths[0])) {
            fullStr += chalk_1.default.green(`Will upload code coverage report file ${reportPaths[0]}`);
        }
        else {
            fullStr += chalk_1.default.green(`Will look for code coverage report files in ${reportPaths.join(', ')}`);
        }
    }
    return fullStr;
};
exports.renderCommandInfo = renderCommandInfo;
const renderSuccessfulGitDBSync = (dryRun, elapsed) => {
    return chalk_1.default.green(`${dryRun ? '[DRYRUN] ' : ''}${ICONS.SUCCESS} Synced git metadata in ${elapsed} seconds.`);
};
exports.renderSuccessfulGitDBSync = renderSuccessfulGitDBSync;
const renderFailedGitDBSync = (err) => {
    return chalk_1.default.red.bold(`${ICONS.FAILED} Could not sync git metadata: ${err}\n`);
};
exports.renderFailedGitDBSync = renderFailedGitDBSync;
