// @datadog/datadog-ci-base/dist/commands/deployment/correlate-image.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeploymentCorrelateImageCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("@datadog/datadog-ci-base/helpers/plugin");
const __1 = require("../..");
class DeploymentCorrelateImageCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.commitSha = clipanion_1.Option.String('--commit-sha');
        this.repositoryUrl = clipanion_1.Option.String('--repository-url');
        this.image = clipanion_1.Option.String('--image');
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.DeploymentCorrelateImageCommand = DeploymentCorrelateImageCommand;
DeploymentCorrelateImageCommand.paths = [['deployment', 'correlate-image']];
DeploymentCorrelateImageCommand.usage = clipanion_1.Command.Usage({
    category: 'CI Visibility',
    description: 'Correlate images with their source commit.',
    details: 'This command will correlate the image with a commit of the application repository.',
});
