// @datadog/datadog-ci-base/dist/commands/deployment/correlate.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeploymentCorrelateCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("@datadog/datadog-ci-base/helpers/plugin");
const __1 = require("../..");
/**
 * This command collects environment variables and git information to correlate commits from the
 * source code repository to the configuration repository. This allows to connect pipelines triggering
 * changes on the configuration repository to deployments from gitOps CD providers
 */
class DeploymentCorrelateCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.cdProviderParam = clipanion_1.Option.String('--provider');
        this.configurationRepo = clipanion_1.Option.String('--config-repo');
        this.configurationShas = clipanion_1.Option.Array('--config-shas');
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false);
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.DeploymentCorrelateCommand = DeploymentCorrelateCommand;
DeploymentCorrelateCommand.paths = [['deployment', 'correlate']];
DeploymentCorrelateCommand.usage = clipanion_1.Command.Usage({
    category: 'CI Visibility',
    description: 'Correlate GitOps CD deployments with CI pipelines.',
    details: `
      This command will correlate the pipeline with a GitOps CD deployment.\n
      See README for additional details.
    `,
    examples: [
        ['Correlate an Argo CD deployment', 'datadog-ci deployment correlate --provider argocd'],
        [
            'Correlate ArgoCD deployment manually',
            'datadog-ci deployment correlate --provider argocd --config-repo https://github.com/my-manifests-repo --config-shas 92eb0db6926aaf51b9fb223895b6d8d1c0ff1ff4',
        ],
        [
            'Correlate ArgoCD deployment manually to several commits',
            'datadog-ci deployment correlate --provider argocd --config-repo https://github.com/my-manifests-repo --config-shas 92eb0db6926aaf51b9fb223895b6d8d1c0ff1ff4 --config-shas e996e5c30ba1cb4dc7f634ab4a0a59473741c4de',
        ],
    ],
});
