// @datadog/datadog-ci-base/dist/commands/deployment/gate.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeploymentGateCommand = void 0;
const clipanion_1 = require("clipanion");
const t = __importStar(require("typanion"));
const plugin_1 = require("@datadog/datadog-ci-base/helpers/plugin");
const __1 = require("../..");
/**
 * This command allows to evaluate a deployment gate in Datadog.
 * It handles the entire process of requesting a gate evaluation and polling for results
 * The command will exit with status 0 when the gate passes and status 1 otherwise.
 */
class DeploymentGateCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        // Required parameters
        this.service = clipanion_1.Option.String('--service', {
            description: 'The service name (e.g. payments-backend)',
            validator: t.isString(),
        });
        this.env = clipanion_1.Option.String('--env', {
            description: 'The environment name (e.g. prod, staging)',
            validator: t.isString(),
        });
        // Optional parameters
        this.identifier = clipanion_1.Option.String('--identifier', {
            description: 'The deployment identifier (defaults to "default")',
            validator: t.isString(),
        });
        this.version = clipanion_1.Option.String('--version', {
            description: 'The deployment version (required for gates with faulty deployment detection rules)',
            validator: t.isString(),
        });
        this.apmPrimaryTag = clipanion_1.Option.String('--apm-primary-tag', {
            description: 'The APM primary tag (only for gates with faulty deployment detection rules)',
            validator: t.isString(),
        });
        this.timeout = clipanion_1.Option.String('--timeout', '10800', {
            description: 'Maximum amount of seconds to wait for the script execution in seconds (default: 10800 = 3 hours)',
            validator: t.isString(),
        });
        this.failOnError = clipanion_1.Option.Boolean('--fail-on-error', false, {
            description: 'When true, the script will consider the gate as failed when timeout is reached or unexpected errors occur calling the Datadog APIs',
        });
        // monitorsQueryVariable is hidden because it's not available yet
        this.monitorsQueryVariable = clipanion_1.Option.String('--monitors-query-variable', '', {
            validator: t.isString(),
            hidden: true,
        });
        // FIPS options
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.DeploymentGateCommand = DeploymentGateCommand;
DeploymentGateCommand.paths = [['deployment', 'gate']];
DeploymentGateCommand.usage = clipanion_1.Command.Usage({
    category: 'CI Visibility',
    description: 'Evaluate deployment gates in Datadog.',
    details: `
      This command allows to evaluate a deployment gate in Datadog.
      The command will exit with status 0 when the gate passes and status 1 otherwise.
    `,
    examples: [
        [
            'Evaluate a deployment gate for payments-backend service in prod environment',
            'datadog-ci deployment gate --service payments-backend --env prod',
        ],
        [
            'Evaluate a deployment gate with custom timeout',
            'datadog-ci deployment gate --service payments-backend --env prod --timeout 7200',
        ],
        [
            'Evaluate a deployment gate and fail if an error occurs',
            'datadog-ci deployment gate --service payments-backend --env prod --fail-on-error',
        ],
        [
            'Evaluate a deployment gate with version and APM primary tag',
            'datadog-ci deployment gate --service payments-backend --env prod --version 1.2.3 --apm-primary-tag region:us-central-1',
        ],
    ],
});
