// @datadog/datadog-ci-base/dist/commands/deployment/mark.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeploymentMarkCommand = void 0;
const clipanion_1 = require("clipanion");
const plugin_1 = require("@datadog/datadog-ci-base/helpers/plugin");
const __1 = require("../..");
/**
 * This command is a wrapper around the datadog-ci tag command, allowing customers to mark CI jobs
 * as deployments and setting specific properties, like the environment or the revision in a simple way.
 */
class DeploymentMarkCommand extends __1.BaseCommand {
    constructor() {
        super(...arguments);
        this.noFail = clipanion_1.Option.Boolean('--no-fail', false);
        this.isRollback = clipanion_1.Option.Boolean('--is-rollback', false);
        this.env = clipanion_1.Option.String('--env', {
            description: 'Example: prod',
        });
        this.revision = clipanion_1.Option.String('--revision', {
            description: 'Example: 1.0.0',
        });
        this.service = clipanion_1.Option.String('--service', {
            description: 'Example: payment-service',
        });
        this.tags = clipanion_1.Option.Array('--tags');
        this.fips = clipanion_1.Option.Boolean('--fips', false);
        this.fipsIgnoreError = clipanion_1.Option.Boolean('--fips-ignore-error', false);
        this.dryRun = clipanion_1.Option.Boolean('--dry-run', false);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, plugin_1.executePluginCommand)(this);
        });
    }
}
exports.DeploymentMarkCommand = DeploymentMarkCommand;
DeploymentMarkCommand.paths = [['deployment', 'mark']];
DeploymentMarkCommand.usage = clipanion_1.Command.Usage({
    category: 'CI Visibility',
    description: 'Mark a CI job as a deployment.',
    details: `
      This command will mark a CI job as a deployment.\n
      See README for details.
    `,
    examples: [
        ['Mark a CI job as a deployment', 'datadog-ci deployment mark'],
        ['Mark a CI job as a deployment to the staging environment', 'datadog-ci deployment mark --env:staging'],
        ['Mark a CI job as a rollback deployment', 'datadog-ci deployment mark --is-rollback'],
        ['Mark a CI job as a deployment of the v123-456 version', 'datadog-ci deployment mark --revision:v123-456'],
        [
            'Mark a CI job as a deployment for service payment-service',
            'datadog-ci deployment mark --service:payment-service',
        ],
    ],
});
