// @datadog/datadog-ci-base/dist/commands/elf-symbols/elf-constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoteType = exports.ProgramHeaderType = exports.SectionHeaderType = exports.ELF_TYPES_DESCRIPTION = exports.MACHINE_TYPES_DESCRIPTION = exports.MachineType = exports.ElfFileType = exports.ElfClass = void 0;
var ElfClass;
(function (ElfClass) {
    ElfClass[ElfClass["ELFCLASSNONE"] = 0] = "ELFCLASSNONE";
    ElfClass[ElfClass["ELFCLASS32"] = 1] = "ELFCLASS32";
    ElfClass[ElfClass["ELFCLASS64"] = 2] = "ELFCLASS64";
    ElfClass[ElfClass["ELFCLASSNUM"] = 3] = "ELFCLASSNUM";
})(ElfClass || (exports.ElfClass = ElfClass = {}));
var ElfFileType;
(function (ElfFileType) {
    ElfFileType[ElfFileType["ET_NONE"] = 0] = "ET_NONE";
    ElfFileType[ElfFileType["ET_REL"] = 1] = "ET_REL";
    ElfFileType[ElfFileType["ET_EXEC"] = 2] = "ET_EXEC";
    ElfFileType[ElfFileType["ET_DYN"] = 3] = "ET_DYN";
    ElfFileType[ElfFileType["ET_CORE"] = 4] = "ET_CORE";
})(ElfFileType || (exports.ElfFileType = ElfFileType = {}));
var MachineType;
(function (MachineType) {
    MachineType[MachineType["EM_NONE"] = 0] = "EM_NONE";
    MachineType[MachineType["EM_M32"] = 1] = "EM_M32";
    MachineType[MachineType["EM_SPARC"] = 2] = "EM_SPARC";
    MachineType[MachineType["EM_386"] = 3] = "EM_386";
    MachineType[MachineType["EM_68K"] = 4] = "EM_68K";
    MachineType[MachineType["EM_88K"] = 5] = "EM_88K";
    MachineType[MachineType["EM_IAMCU"] = 6] = "EM_IAMCU";
    MachineType[MachineType["EM_860"] = 7] = "EM_860";
    MachineType[MachineType["EM_MIPS"] = 8] = "EM_MIPS";
    MachineType[MachineType["EM_S370"] = 9] = "EM_S370";
    MachineType[MachineType["EM_MIPS_RS3_LE"] = 10] = "EM_MIPS_RS3_LE";
    // Skipping reserved 11-14
    MachineType[MachineType["EM_PARISC"] = 15] = "EM_PARISC";
    // Skipping reserved 16
    MachineType[MachineType["EM_VPP500"] = 17] = "EM_VPP500";
    MachineType[MachineType["EM_SPARC32PLUS"] = 18] = "EM_SPARC32PLUS";
    MachineType[MachineType["EM_960"] = 19] = "EM_960";
    MachineType[MachineType["EM_PPC"] = 20] = "EM_PPC";
    MachineType[MachineType["EM_PPC64"] = 21] = "EM_PPC64";
    MachineType[MachineType["EM_S390"] = 22] = "EM_S390";
    MachineType[MachineType["EM_SPU"] = 23] = "EM_SPU";
    // Skipping reserved 24-35
    MachineType[MachineType["EM_V800"] = 36] = "EM_V800";
    MachineType[MachineType["EM_FR20"] = 37] = "EM_FR20";
    MachineType[MachineType["EM_RH32"] = 38] = "EM_RH32";
    MachineType[MachineType["EM_RCE"] = 39] = "EM_RCE";
    MachineType[MachineType["EM_ARM"] = 40] = "EM_ARM";
    MachineType[MachineType["EM_FAKE_ALPHA"] = 41] = "EM_FAKE_ALPHA";
    MachineType[MachineType["EM_SH"] = 42] = "EM_SH";
    MachineType[MachineType["EM_SPARCV9"] = 43] = "EM_SPARCV9";
    MachineType[MachineType["EM_TRICORE"] = 44] = "EM_TRICORE";
    MachineType[MachineType["EM_ARC"] = 45] = "EM_ARC";
    MachineType[MachineType["EM_H8_300"] = 46] = "EM_H8_300";
    MachineType[MachineType["EM_H8_300H"] = 47] = "EM_H8_300H";
    MachineType[MachineType["EM_H8S"] = 48] = "EM_H8S";
    MachineType[MachineType["EM_H8_500"] = 49] = "EM_H8_500";
    MachineType[MachineType["EM_IA_64"] = 50] = "EM_IA_64";
    MachineType[MachineType["EM_MIPS_X"] = 51] = "EM_MIPS_X";
    MachineType[MachineType["EM_COLDFIRE"] = 52] = "EM_COLDFIRE";
    MachineType[MachineType["EM_68HC12"] = 53] = "EM_68HC12";
    MachineType[MachineType["EM_MMA"] = 54] = "EM_MMA";
    MachineType[MachineType["EM_PCP"] = 55] = "EM_PCP";
    MachineType[MachineType["EM_NCPU"] = 56] = "EM_NCPU";
    MachineType[MachineType["EM_NDR1"] = 57] = "EM_NDR1";
    MachineType[MachineType["EM_STARCORE"] = 58] = "EM_STARCORE";
    MachineType[MachineType["EM_ME16"] = 59] = "EM_ME16";
    MachineType[MachineType["EM_ST100"] = 60] = "EM_ST100";
    MachineType[MachineType["EM_TINYJ"] = 61] = "EM_TINYJ";
    MachineType[MachineType["EM_X86_64"] = 62] = "EM_X86_64";
    MachineType[MachineType["EM_PDSP"] = 63] = "EM_PDSP";
    MachineType[MachineType["EM_PDP10"] = 64] = "EM_PDP10";
    MachineType[MachineType["EM_PDP11"] = 65] = "EM_PDP11";
    MachineType[MachineType["EM_FX66"] = 66] = "EM_FX66";
    MachineType[MachineType["EM_ST9PLUS"] = 67] = "EM_ST9PLUS";
    MachineType[MachineType["EM_ST7"] = 68] = "EM_ST7";
    MachineType[MachineType["EM_68HC16"] = 69] = "EM_68HC16";
    MachineType[MachineType["EM_68HC11"] = 70] = "EM_68HC11";
    MachineType[MachineType["EM_68HC08"] = 71] = "EM_68HC08";
    MachineType[MachineType["EM_68HC05"] = 72] = "EM_68HC05";
    MachineType[MachineType["EM_SVX"] = 73] = "EM_SVX";
    MachineType[MachineType["EM_ST19"] = 74] = "EM_ST19";
    MachineType[MachineType["EM_VAX"] = 75] = "EM_VAX";
    MachineType[MachineType["EM_CRIS"] = 76] = "EM_CRIS";
    MachineType[MachineType["EM_JAVELIN"] = 77] = "EM_JAVELIN";
    MachineType[MachineType["EM_FIREPATH"] = 78] = "EM_FIREPATH";
    MachineType[MachineType["EM_ZSP"] = 79] = "EM_ZSP";
    MachineType[MachineType["EM_MMIX"] = 80] = "EM_MMIX";
    MachineType[MachineType["EM_HUANY"] = 81] = "EM_HUANY";
    MachineType[MachineType["EM_PRISM"] = 82] = "EM_PRISM";
    MachineType[MachineType["EM_AVR"] = 83] = "EM_AVR";
    MachineType[MachineType["EM_FR30"] = 84] = "EM_FR30";
    MachineType[MachineType["EM_D10V"] = 85] = "EM_D10V";
    MachineType[MachineType["EM_D30V"] = 86] = "EM_D30V";
    MachineType[MachineType["EM_V850"] = 87] = "EM_V850";
    MachineType[MachineType["EM_M32R"] = 88] = "EM_M32R";
    MachineType[MachineType["EM_MN10300"] = 89] = "EM_MN10300";
    MachineType[MachineType["EM_MN10200"] = 90] = "EM_MN10200";
    MachineType[MachineType["EM_PJ"] = 91] = "EM_PJ";
    MachineType[MachineType["EM_OPENRISC"] = 92] = "EM_OPENRISC";
    MachineType[MachineType["EM_ARC_COMPACT"] = 93] = "EM_ARC_COMPACT";
    MachineType[MachineType["EM_XTENSA"] = 94] = "EM_XTENSA";
    MachineType[MachineType["EM_VIDEOCORE"] = 95] = "EM_VIDEOCORE";
    MachineType[MachineType["EM_TMM_GPP"] = 96] = "EM_TMM_GPP";
    MachineType[MachineType["EM_NS32K"] = 97] = "EM_NS32K";
    MachineType[MachineType["EM_TPC"] = 98] = "EM_TPC";
    MachineType[MachineType["EM_SNP1K"] = 99] = "EM_SNP1K";
    MachineType[MachineType["EM_ST200"] = 100] = "EM_ST200";
    MachineType[MachineType["EM_IP2K"] = 101] = "EM_IP2K";
    MachineType[MachineType["EM_MAX"] = 102] = "EM_MAX";
    MachineType[MachineType["EM_CR"] = 103] = "EM_CR";
    MachineType[MachineType["EM_F2MC16"] = 104] = "EM_F2MC16";
    MachineType[MachineType["EM_MSP430"] = 105] = "EM_MSP430";
    MachineType[MachineType["EM_BLACKFIN"] = 106] = "EM_BLACKFIN";
    MachineType[MachineType["EM_SE_C33"] = 107] = "EM_SE_C33";
    MachineType[MachineType["EM_SEP"] = 108] = "EM_SEP";
    MachineType[MachineType["EM_ARCA"] = 109] = "EM_ARCA";
    MachineType[MachineType["EM_UNICORE"] = 110] = "EM_UNICORE";
    MachineType[MachineType["EM_EXCESS"] = 111] = "EM_EXCESS";
    MachineType[MachineType["EM_DXP"] = 112] = "EM_DXP";
    MachineType[MachineType["EM_ALTERA_NIOS2"] = 113] = "EM_ALTERA_NIOS2";
    MachineType[MachineType["EM_CRX"] = 114] = "EM_CRX";
    MachineType[MachineType["EM_XGATE"] = 115] = "EM_XGATE";
    MachineType[MachineType["EM_C166"] = 116] = "EM_C166";
    MachineType[MachineType["EM_M16C"] = 117] = "EM_M16C";
    MachineType[MachineType["EM_DSPIC30F"] = 118] = "EM_DSPIC30F";
    MachineType[MachineType["EM_CE"] = 119] = "EM_CE";
    MachineType[MachineType["EM_M32C"] = 120] = "EM_M32C";
    // Skipping reserved 121-130
    MachineType[MachineType["EM_TSK3000"] = 131] = "EM_TSK3000";
    MachineType[MachineType["EM_RS08"] = 132] = "EM_RS08";
    MachineType[MachineType["EM_SHARC"] = 133] = "EM_SHARC";
    MachineType[MachineType["EM_ECOG2"] = 134] = "EM_ECOG2";
    MachineType[MachineType["EM_SCORE7"] = 135] = "EM_SCORE7";
    MachineType[MachineType["EM_DSP24"] = 136] = "EM_DSP24";
    MachineType[MachineType["EM_VIDEOCORE3"] = 137] = "EM_VIDEOCORE3";
    MachineType[MachineType["EM_LATTICEMICO32"] = 138] = "EM_LATTICEMICO32";
    MachineType[MachineType["EM_SE_C17"] = 139] = "EM_SE_C17";
    MachineType[MachineType["EM_TI_C6000"] = 140] = "EM_TI_C6000";
    MachineType[MachineType["EM_TI_C2000"] = 141] = "EM_TI_C2000";
    MachineType[MachineType["EM_TI_C5500"] = 142] = "EM_TI_C5500";
    MachineType[MachineType["EM_TI_ARP32"] = 143] = "EM_TI_ARP32";
    MachineType[MachineType["EM_TI_PRU"] = 144] = "EM_TI_PRU";
    // Skipping reserved 145-159
    MachineType[MachineType["EM_MMDSP_PLUS"] = 160] = "EM_MMDSP_PLUS";
    MachineType[MachineType["EM_CYPRESS_M8C"] = 161] = "EM_CYPRESS_M8C";
    MachineType[MachineType["EM_R32C"] = 162] = "EM_R32C";
    MachineType[MachineType["EM_TRIMEDIA"] = 163] = "EM_TRIMEDIA";
    MachineType[MachineType["EM_QDSP6"] = 164] = "EM_QDSP6";
    MachineType[MachineType["EM_8051"] = 165] = "EM_8051";
    MachineType[MachineType["EM_STXP7X"] = 166] = "EM_STXP7X";
    MachineType[MachineType["EM_NDS32"] = 167] = "EM_NDS32";
    MachineType[MachineType["EM_ECOG1X"] = 168] = "EM_ECOG1X";
    MachineType[MachineType["EM_MAXQ30"] = 169] = "EM_MAXQ30";
    MachineType[MachineType["EM_XIMO16"] = 170] = "EM_XIMO16";
    MachineType[MachineType["EM_MANIK"] = 171] = "EM_MANIK";
    MachineType[MachineType["EM_CRAYNV2"] = 172] = "EM_CRAYNV2";
    MachineType[MachineType["EM_RX"] = 173] = "EM_RX";
    MachineType[MachineType["EM_METAG"] = 174] = "EM_METAG";
    MachineType[MachineType["EM_MCST_ELBRUS"] = 175] = "EM_MCST_ELBRUS";
    MachineType[MachineType["EM_ECOG16"] = 176] = "EM_ECOG16";
    MachineType[MachineType["EM_CR16"] = 177] = "EM_CR16";
    MachineType[MachineType["EM_ETPU"] = 178] = "EM_ETPU";
    MachineType[MachineType["EM_SLE9X"] = 179] = "EM_SLE9X";
    MachineType[MachineType["EM_L10M"] = 180] = "EM_L10M";
    MachineType[MachineType["EM_K10M"] = 181] = "EM_K10M";
    // Skipping reserved 182
    MachineType[MachineType["EM_AARCH64"] = 183] = "EM_AARCH64";
    // Skipping reserved 184
    MachineType[MachineType["EM_AVR32"] = 185] = "EM_AVR32";
    MachineType[MachineType["EM_STM8"] = 186] = "EM_STM8";
    MachineType[MachineType["EM_TILE64"] = 187] = "EM_TILE64";
    MachineType[MachineType["EM_TILEPRO"] = 188] = "EM_TILEPRO";
    MachineType[MachineType["EM_MICROBLAZE"] = 189] = "EM_MICROBLAZE";
    MachineType[MachineType["EM_CUDA"] = 190] = "EM_CUDA";
    MachineType[MachineType["EM_TILEGX"] = 191] = "EM_TILEGX";
    MachineType[MachineType["EM_CLOUDSHIELD"] = 192] = "EM_CLOUDSHIELD";
    MachineType[MachineType["EM_COREA_1ST"] = 193] = "EM_COREA_1ST";
    MachineType[MachineType["EM_COREA_2ND"] = 194] = "EM_COREA_2ND";
    MachineType[MachineType["EM_ARCV2"] = 195] = "EM_ARCV2";
    MachineType[MachineType["EM_OPEN8"] = 196] = "EM_OPEN8";
    MachineType[MachineType["EM_RL78"] = 197] = "EM_RL78";
    MachineType[MachineType["EM_VIDEOCORE5"] = 198] = "EM_VIDEOCORE5";
    MachineType[MachineType["EM_78KOR"] = 199] = "EM_78KOR";
    MachineType[MachineType["EM_56800EX"] = 200] = "EM_56800EX";
    MachineType[MachineType["EM_BA1"] = 201] = "EM_BA1";
    MachineType[MachineType["EM_BA2"] = 202] = "EM_BA2";
    MachineType[MachineType["EM_XCORE"] = 203] = "EM_XCORE";
    MachineType[MachineType["EM_MCHP_PIC"] = 204] = "EM_MCHP_PIC";
    MachineType[MachineType["EM_INTELGT"] = 205] = "EM_INTELGT";
    // Skipping reserved 206-209
    MachineType[MachineType["EM_KM32"] = 210] = "EM_KM32";
    MachineType[MachineType["EM_KMX32"] = 211] = "EM_KMX32";
    MachineType[MachineType["EM_EMX16"] = 212] = "EM_EMX16";
    MachineType[MachineType["EM_EMX8"] = 213] = "EM_EMX8";
    MachineType[MachineType["EM_KVARC"] = 214] = "EM_KVARC";
    MachineType[MachineType["EM_CDP"] = 215] = "EM_CDP";
    MachineType[MachineType["EM_COGE"] = 216] = "EM_COGE";
    MachineType[MachineType["EM_COOL"] = 217] = "EM_COOL";
    MachineType[MachineType["EM_NORC"] = 218] = "EM_NORC";
    MachineType[MachineType["EM_CSR_KALIMBA"] = 219] = "EM_CSR_KALIMBA";
    MachineType[MachineType["EM_Z80"] = 220] = "EM_Z80";
    MachineType[MachineType["EM_VISIUM"] = 221] = "EM_VISIUM";
    MachineType[MachineType["EM_FT32"] = 222] = "EM_FT32";
    MachineType[MachineType["EM_MOXIE"] = 223] = "EM_MOXIE";
    MachineType[MachineType["EM_AMDGPU"] = 224] = "EM_AMDGPU";
    // Skipping reserved 225-242
    MachineType[MachineType["EM_RISCV"] = 243] = "EM_RISCV";
    MachineType[MachineType["EM_BPF"] = 247] = "EM_BPF";
    MachineType[MachineType["EM_CSKY"] = 252] = "EM_CSKY";
})(MachineType || (exports.MachineType = MachineType = {}));
exports.MACHINE_TYPES_DESCRIPTION = {
    [MachineType.EM_NONE]: 'none',
    [MachineType.EM_M32]: 'we32100',
    [MachineType.EM_SPARC]: 'sparc',
    [MachineType.EM_386]: '386',
    [MachineType.EM_68K]: '68k',
    [MachineType.EM_88K]: '88k',
    [MachineType.EM_IAMCU]: 'iamcu',
    [MachineType.EM_860]: '860',
    [MachineType.EM_MIPS]: 'mips',
    [MachineType.EM_S370]: 's370',
    [MachineType.EM_MIPS_RS3_LE]: 'mips_rs3_le',
    [MachineType.EM_PARISC]: 'parisc',
    [MachineType.EM_VPP500]: 'vpp500',
    [MachineType.EM_SPARC32PLUS]: 'sparc32plus',
    [MachineType.EM_960]: '960',
    [MachineType.EM_PPC]: 'ppc',
    [MachineType.EM_PPC64]: 'ppc64',
    [MachineType.EM_S390]: 's390',
    [MachineType.EM_SPU]: 'spu',
    [MachineType.EM_V800]: 'v800',
    [MachineType.EM_FR20]: 'fr20',
    [MachineType.EM_RH32]: 'rh32',
    [MachineType.EM_RCE]: 'rce',
    [MachineType.EM_ARM]: 'arm',
    [MachineType.EM_FAKE_ALPHA]: 'fake_alpha',
    [MachineType.EM_SH]: 'sh',
    [MachineType.EM_SPARCV9]: 'sparcv9',
    [MachineType.EM_TRICORE]: 'tricore',
    [MachineType.EM_ARC]: 'arc',
    [MachineType.EM_H8_300]: 'h8_300',
    [MachineType.EM_H8_300H]: 'h8_300h',
    [MachineType.EM_H8S]: 'h8s',
    [MachineType.EM_H8_500]: 'h8_500',
    [MachineType.EM_IA_64]: 'ia_64',
    [MachineType.EM_MIPS_X]: 'mips_x',
    [MachineType.EM_COLDFIRE]: 'coldfire',
    [MachineType.EM_68HC12]: '68hc12',
    [MachineType.EM_MMA]: 'mma',
    [MachineType.EM_PCP]: 'pcp',
    [MachineType.EM_NCPU]: 'ncpu',
    [MachineType.EM_NDR1]: 'ndr1',
    [MachineType.EM_STARCORE]: 'starcore',
    [MachineType.EM_ME16]: 'me16',
    [MachineType.EM_ST100]: 'st100',
    [MachineType.EM_TINYJ]: 'tinyj',
    [MachineType.EM_X86_64]: 'x86_64',
    [MachineType.EM_PDSP]: 'pdsp',
    [MachineType.EM_PDP10]: 'pdp10',
    [MachineType.EM_PDP11]: 'pdp11',
    [MachineType.EM_FX66]: 'fx66',
    [MachineType.EM_ST9PLUS]: 'st9plus',
    [MachineType.EM_ST7]: 'st7',
    [MachineType.EM_68HC16]: '68hc16',
    [MachineType.EM_68HC11]: '68hc11',
    [MachineType.EM_68HC08]: '68hc08',
    [MachineType.EM_68HC05]: '68hc05',
    [MachineType.EM_SVX]: 'svx',
    [MachineType.EM_ST19]: 'st19',
    [MachineType.EM_VAX]: 'vax',
    [MachineType.EM_CRIS]: 'cris',
    [MachineType.EM_JAVELIN]: 'javelin',
    [MachineType.EM_FIREPATH]: 'firepath',
    [MachineType.EM_ZSP]: 'zsp',
    [MachineType.EM_MMIX]: 'mmix',
    [MachineType.EM_HUANY]: 'huany',
    [MachineType.EM_PRISM]: 'prism',
    [MachineType.EM_AVR]: 'avr',
    [MachineType.EM_FR30]: 'fr30',
    [MachineType.EM_D10V]: 'd10v',
    [MachineType.EM_D30V]: 'd30v',
    [MachineType.EM_V850]: 'v850',
    [MachineType.EM_M32R]: 'm32r',
    [MachineType.EM_MN10300]: 'mn10300',
    [MachineType.EM_MN10200]: 'mn10200',
    [MachineType.EM_PJ]: 'pj',
    [MachineType.EM_OPENRISC]: 'openrisc',
    [MachineType.EM_ARC_COMPACT]: 'arc_compact',
    [MachineType.EM_XTENSA]: 'xtensa',
    [MachineType.EM_VIDEOCORE]: 'videocore',
    [MachineType.EM_TMM_GPP]: 'tmm_gpp',
    [MachineType.EM_NS32K]: 'ns32k',
    [MachineType.EM_TPC]: 'tpc',
    [MachineType.EM_SNP1K]: 'snp1k',
    [MachineType.EM_ST200]: 'st200',
    [MachineType.EM_IP2K]: 'ip2k',
    [MachineType.EM_MAX]: 'max',
    [MachineType.EM_CR]: 'cr',
    [MachineType.EM_F2MC16]: 'f2mc16',
    [MachineType.EM_MSP430]: 'msp430',
    [MachineType.EM_BLACKFIN]: 'blackfin',
    [MachineType.EM_SE_C33]: 'se_c33',
    [MachineType.EM_SEP]: 'sep',
    [MachineType.EM_ARCA]: 'arca',
    [MachineType.EM_UNICORE]: 'unicore',
    [MachineType.EM_EXCESS]: 'excess',
    [MachineType.EM_DXP]: 'dxp',
    [MachineType.EM_ALTERA_NIOS2]: 'altera_nios2',
    [MachineType.EM_CRX]: 'crx',
    [MachineType.EM_XGATE]: 'xgate',
    [MachineType.EM_C166]: 'c166',
    [MachineType.EM_M16C]: 'm16c',
    [MachineType.EM_DSPIC30F]: 'dspic30f',
    [MachineType.EM_CE]: 'ce',
    [MachineType.EM_M32C]: 'm32c',
    [MachineType.EM_TSK3000]: 'tsk3000',
    [MachineType.EM_RS08]: 'rs08',
    [MachineType.EM_SHARC]: 'sharc',
    [MachineType.EM_ECOG2]: 'ecog2',
    [MachineType.EM_SCORE7]: 'score7',
    [MachineType.EM_DSP24]: 'dsp24',
    [MachineType.EM_VIDEOCORE3]: 'videocore3',
    [MachineType.EM_LATTICEMICO32]: 'latticemico32',
    [MachineType.EM_SE_C17]: 'se_c17',
    [MachineType.EM_TI_C6000]: 'ti_c6000',
    [MachineType.EM_TI_C2000]: 'ti_c2000',
    [MachineType.EM_TI_C5500]: 'ti_c5500',
    [MachineType.EM_TI_ARP32]: 'ti_arp32',
    [MachineType.EM_TI_PRU]: 'ti_pru',
    [MachineType.EM_MMDSP_PLUS]: 'mmdsp_plus',
    [MachineType.EM_CYPRESS_M8C]: 'cypress_m8c',
    [MachineType.EM_R32C]: 'r32c',
    [MachineType.EM_TRIMEDIA]: 'trimedia',
    [MachineType.EM_QDSP6]: 'qdsp6',
    [MachineType.EM_8051]: '8051',
    [MachineType.EM_STXP7X]: 'stxp7x',
    [MachineType.EM_NDS32]: 'nds32',
    [MachineType.EM_ECOG1X]: 'ecog1x',
    [MachineType.EM_MAXQ30]: 'maxq30',
    [MachineType.EM_XIMO16]: 'ximo16',
    [MachineType.EM_MANIK]: 'manik',
    [MachineType.EM_CRAYNV2]: 'craynv2',
    [MachineType.EM_RX]: 'rx',
    [MachineType.EM_METAG]: 'metag',
    [MachineType.EM_MCST_ELBRUS]: 'mcst_elbrus',
    [MachineType.EM_ECOG16]: 'ecog16',
    [MachineType.EM_CR16]: 'cr16',
    [MachineType.EM_ETPU]: 'etpu',
    [MachineType.EM_SLE9X]: 'sle9x',
    [MachineType.EM_L10M]: 'l10m',
    [MachineType.EM_K10M]: 'k10m',
    [MachineType.EM_AARCH64]: 'aarch64',
    [MachineType.EM_AVR32]: 'avr32',
    [MachineType.EM_STM8]: 'stm8',
    [MachineType.EM_TILE64]: 'tile64',
    [MachineType.EM_TILEPRO]: 'tilepro',
    [MachineType.EM_MICROBLAZE]: 'microblaze',
    [MachineType.EM_CUDA]: 'cuda',
    [MachineType.EM_TILEGX]: 'tilegx',
    [MachineType.EM_CLOUDSHIELD]: 'cloudshield',
    [MachineType.EM_COREA_1ST]: 'corea_1st',
    [MachineType.EM_COREA_2ND]: 'corea_2nd',
    [MachineType.EM_ARCV2]: 'arcv2',
    [MachineType.EM_OPEN8]: 'open8',
    [MachineType.EM_RL78]: 'rl78',
    [MachineType.EM_VIDEOCORE5]: 'videocore5',
    [MachineType.EM_78KOR]: '78kor',
    [MachineType.EM_56800EX]: '56800ex',
    [MachineType.EM_BA1]: 'ba1',
    [MachineType.EM_BA2]: 'ba2',
    [MachineType.EM_XCORE]: 'xcore',
    [MachineType.EM_MCHP_PIC]: 'mchp_pic',
    [MachineType.EM_INTELGT]: 'intelgt',
    [MachineType.EM_KM32]: 'km32',
    [MachineType.EM_KMX32]: 'kmx32',
    [MachineType.EM_EMX16]: 'emx16',
    [MachineType.EM_EMX8]: 'emx8',
    [MachineType.EM_KVARC]: 'kvarc',
    [MachineType.EM_CDP]: 'cdp',
    [MachineType.EM_COGE]: 'coge',
    [MachineType.EM_COOL]: 'cool',
    [MachineType.EM_NORC]: 'norc',
    [MachineType.EM_CSR_KALIMBA]: 'csr_kalimba',
    [MachineType.EM_Z80]: 'z80',
    [MachineType.EM_VISIUM]: 'visium',
    [MachineType.EM_FT32]: 'ft32',
    [MachineType.EM_MOXIE]: 'moxie',
    [MachineType.EM_AMDGPU]: 'amdgpu',
    [MachineType.EM_RISCV]: 'riscv',
    [MachineType.EM_BPF]: 'bpf',
    [MachineType.EM_CSKY]: 'csky',
};
exports.ELF_TYPES_DESCRIPTION = {
    [ElfFileType.ET_NONE]: 'NONE',
    [ElfFileType.ET_REL]: 'REL',
    [ElfFileType.ET_EXEC]: 'EXEC',
    [ElfFileType.ET_DYN]: 'DYN',
    [ElfFileType.ET_CORE]: 'CORE',
};
var SectionHeaderType;
(function (SectionHeaderType) {
    SectionHeaderType[SectionHeaderType["SHT_NULL"] = 0] = "SHT_NULL";
    SectionHeaderType[SectionHeaderType["SHT_PROGBITS"] = 1] = "SHT_PROGBITS";
    SectionHeaderType[SectionHeaderType["SHT_SYMTAB"] = 2] = "SHT_SYMTAB";
    SectionHeaderType[SectionHeaderType["SHT_STRTAB"] = 3] = "SHT_STRTAB";
    SectionHeaderType[SectionHeaderType["SHT_RELA"] = 4] = "SHT_RELA";
    SectionHeaderType[SectionHeaderType["SHT_HASH"] = 5] = "SHT_HASH";
    SectionHeaderType[SectionHeaderType["SHT_DYNAMIC"] = 6] = "SHT_DYNAMIC";
    SectionHeaderType[SectionHeaderType["SHT_NOTE"] = 7] = "SHT_NOTE";
    SectionHeaderType[SectionHeaderType["SHT_NOBITS"] = 8] = "SHT_NOBITS";
    SectionHeaderType[SectionHeaderType["SHT_REL"] = 9] = "SHT_REL";
    SectionHeaderType[SectionHeaderType["SHT_SHLIB"] = 10] = "SHT_SHLIB";
    SectionHeaderType[SectionHeaderType["SHT_DYNSYM"] = 11] = "SHT_DYNSYM";
    SectionHeaderType[SectionHeaderType["SHT_INIT_ARRAY"] = 14] = "SHT_INIT_ARRAY";
    SectionHeaderType[SectionHeaderType["SHT_FINI_ARRAY"] = 15] = "SHT_FINI_ARRAY";
    SectionHeaderType[SectionHeaderType["SHT_PREINIT_ARRAY"] = 16] = "SHT_PREINIT_ARRAY";
    SectionHeaderType[SectionHeaderType["SHT_GROUP"] = 17] = "SHT_GROUP";
    SectionHeaderType[SectionHeaderType["SHT_SYMTAB_SHNDX"] = 18] = "SHT_SYMTAB_SHNDX";
    SectionHeaderType[SectionHeaderType["SHT_NUM"] = 19] = "SHT_NUM";
    SectionHeaderType[SectionHeaderType["SHT_LOOS"] = 1610612736] = "SHT_LOOS";
    SectionHeaderType[SectionHeaderType["SHT_GNU_ATTRIBUTES"] = 1879048181] = "SHT_GNU_ATTRIBUTES";
    SectionHeaderType[SectionHeaderType["SHT_GNU_HASH"] = 1879048182] = "SHT_GNU_HASH";
    SectionHeaderType[SectionHeaderType["SHT_GNU_LIBLIST"] = 1879048183] = "SHT_GNU_LIBLIST";
    SectionHeaderType[SectionHeaderType["SHT_CHECKSUM"] = 1879048184] = "SHT_CHECKSUM";
    SectionHeaderType[SectionHeaderType["SHT_LOSUNW"] = 1879048186] = "SHT_LOSUNW";
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    SectionHeaderType[SectionHeaderType["SHT_SUNW_move"] = 1879048186] = "SHT_SUNW_move";
    SectionHeaderType[SectionHeaderType["SHT_SUNW_COMDAT"] = 1879048187] = "SHT_SUNW_COMDAT";
    SectionHeaderType[SectionHeaderType["SHT_SUNW_syminfo"] = 1879048188] = "SHT_SUNW_syminfo";
    SectionHeaderType[SectionHeaderType["SHT_GNU_verdef"] = 1879048189] = "SHT_GNU_verdef";
    SectionHeaderType[SectionHeaderType["SHT_GNU_verneed"] = 1879048190] = "SHT_GNU_verneed";
    SectionHeaderType[SectionHeaderType["SHT_GNU_versym"] = 1879048191] = "SHT_GNU_versym";
})(SectionHeaderType || (exports.SectionHeaderType = SectionHeaderType = {}));
var ProgramHeaderType;
(function (ProgramHeaderType) {
    ProgramHeaderType[ProgramHeaderType["PT_NULL"] = 0] = "PT_NULL";
    ProgramHeaderType[ProgramHeaderType["PT_LOAD"] = 1] = "PT_LOAD";
    ProgramHeaderType[ProgramHeaderType["PT_DYNAMIC"] = 2] = "PT_DYNAMIC";
    ProgramHeaderType[ProgramHeaderType["PT_INTERP"] = 3] = "PT_INTERP";
    ProgramHeaderType[ProgramHeaderType["PT_NOTE"] = 4] = "PT_NOTE";
    ProgramHeaderType[ProgramHeaderType["PT_SHLIB"] = 5] = "PT_SHLIB";
    ProgramHeaderType[ProgramHeaderType["PT_PHDR"] = 6] = "PT_PHDR";
    ProgramHeaderType[ProgramHeaderType["PT_TLS"] = 7] = "PT_TLS";
    ProgramHeaderType[ProgramHeaderType["PT_NUM"] = 8] = "PT_NUM";
    ProgramHeaderType[ProgramHeaderType["PT_LOOS"] = 1610612736] = "PT_LOOS";
    ProgramHeaderType[ProgramHeaderType["PT_GNU_EH_FRAME"] = 1685382480] = "PT_GNU_EH_FRAME";
    ProgramHeaderType[ProgramHeaderType["PT_GNU_STACK"] = 1685382481] = "PT_GNU_STACK";
    ProgramHeaderType[ProgramHeaderType["PT_GNU_RELRO"] = 1685382482] = "PT_GNU_RELRO";
    ProgramHeaderType[ProgramHeaderType["PT_GNU_PROPERTY"] = 1685382483] = "PT_GNU_PROPERTY";
    ProgramHeaderType[ProgramHeaderType["PT_LOSUNW"] = 1879048186] = "PT_LOSUNW";
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    ProgramHeaderType[ProgramHeaderType["PT_SUNWBSS"] = 1879048186] = "PT_SUNWBSS";
    ProgramHeaderType[ProgramHeaderType["PT_SUNWSTACK"] = 1879048187] = "PT_SUNWSTACK";
    ProgramHeaderType[ProgramHeaderType["PT_HISUNW"] = 1879048191] = "PT_HISUNW";
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    ProgramHeaderType[ProgramHeaderType["PT_HIOS"] = 1879048191] = "PT_HIOS";
    ProgramHeaderType[ProgramHeaderType["PT_LOPROC"] = 1879048192] = "PT_LOPROC";
    ProgramHeaderType[ProgramHeaderType["PT_HIPROC"] = 2147483647] = "PT_HIPROC";
})(ProgramHeaderType || (exports.ProgramHeaderType = ProgramHeaderType = {}));
var NoteType;
(function (NoteType) {
    NoteType[NoteType["NT_GNU_BUILD_ID"] = 3] = "NT_GNU_BUILD_ID";
    NoteType[NoteType["NT_GO_BUILD_ID"] = 4] = "NT_GO_BUILD_ID";
})(NoteType || (exports.NoteType = NoteType = {}));
