// @datadog/datadog-ci-base/dist/commands/elf-symbols/renderer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderUpload = exports.renderRetriedUpload = exports.renderFailedUpload = exports.renderGeneralizedError = exports.renderMissingBinUtils = exports.renderInvalidSymbolsLocation = exports.renderArgumentMissingError = exports.renderWarning = exports.renderGitWarning = exports.renderMissingBuildId = exports.renderCommandSummary = exports.renderCommandInfo = void 0;
const chalk_1 = __importDefault(require("chalk"));
const formatting_1 = require("@datadog/datadog-ci-base/helpers/formatting");
const upload_1 = require("@datadog/datadog-ci-base/helpers/upload");
const utils_1 = require("@datadog/datadog-ci-base/helpers/utils");
const renderCommandInfo = (dryRun, symbolsLocations) => {
    let fullString = '';
    if (dryRun) {
        fullString += chalk_1.default.yellow(`${formatting_1.ICONS.WARNING} DRY-RUN MODE ENABLED. WILL NOT UPLOAD SYMBOLS\n`);
    }
    const startStr = chalk_1.default.green('Starting upload. \n');
    fullString += startStr;
    fullString += chalk_1.default.green(`Uploading symbols from location(s): ${symbolsLocations.join(' ')}\n`);
    fullString += chalk_1.default.green(`After upload is successful symbol files will be processed and ready to use within the next 5 minutes.\n`);
    return fullString;
};
exports.renderCommandInfo = renderCommandInfo;
const renderCommandSummary = (statuses, duration, dryRun) => {
    const results = new Map();
    statuses.forEach((status) => {
        if (!results.has(status)) {
            results.set(status, 0);
        }
        results.set(status, results.get(status) + 1);
    });
    const output = ['', chalk_1.default.bold('Command summary:')];
    if (results.get(upload_1.UploadStatus.Failure)) {
        output.push(chalk_1.default.red(`${formatting_1.ICONS.FAILED} Some symbol files may not been uploaded correctly.`));
    }
    else if (results.get(upload_1.UploadStatus.Skipped)) {
        output.push(chalk_1.default.yellow(`${formatting_1.ICONS.WARNING}  Some symbol files have been skipped.`));
    }
    else if (results.get(upload_1.UploadStatus.Success)) {
        if (dryRun) {
            output.push(chalk_1.default.green(`${formatting_1.ICONS.SUCCESS} [DRYRUN] Handled symbol ${(0, utils_1.pluralize)(results.get(upload_1.UploadStatus.Success), 'file', 'files')} with success in ${duration} seconds.`));
        }
        else {
            output.push(chalk_1.default.green(`${formatting_1.ICONS.SUCCESS} Uploaded symbol ${(0, utils_1.pluralize)(results.get(upload_1.UploadStatus.Success), 'file', 'files')} in ${duration} seconds.`));
        }
    }
    else {
        output.push(chalk_1.default.yellow(`${formatting_1.ICONS.WARNING} No actions were taken. Did you specify the correct path?`));
    }
    return output.join('\n') + '\n';
};
exports.renderCommandSummary = renderCommandSummary;
const renderMissingBuildId = (path) => chalk_1.default.red(`${formatting_1.ICONS.FAILED} Error: Invalid or missing 'build_id' file. Expected at path ${path}`);
exports.renderMissingBuildId = renderMissingBuildId;
const renderGitWarning = (errorMessage) => chalk_1.default.yellow(`${formatting_1.ICONS.WARNING} An error occurred while invoking git: ${errorMessage}
Make sure the command is running within your git repository to fully leverage Datadog's git integration.
To ignore this warning use the --disable-git flag.\n`);
exports.renderGitWarning = renderGitWarning;
const renderWarning = (errorMessage) => chalk_1.default.yellow(`${formatting_1.ICONS.WARNING} ${errorMessage}.\n`);
exports.renderWarning = renderWarning;
const renderArgumentMissingError = (argumentName) => chalk_1.default.red(`${formatting_1.ICONS.FAILED} Error: parameter "${argumentName}" is required.\n`);
exports.renderArgumentMissingError = renderArgumentMissingError;
const renderInvalidSymbolsLocation = (symbolsDirectory) => chalk_1.default.red(`${formatting_1.ICONS.FAILED} Failed to get symbols files - ${symbolsDirectory} is not a file, nor a directory.\n`);
exports.renderInvalidSymbolsLocation = renderInvalidSymbolsLocation;
const renderMissingBinUtils = () => chalk_1.default.red(`${formatting_1.ICONS.FAILED} Could not find required 'objcopy' program. Please install binutils.\n`);
exports.renderMissingBinUtils = renderMissingBinUtils;
const renderGeneralizedError = (error) => {
    let str = chalk_1.default.red(`${formatting_1.ICONS.FAILED} Error: ${error}\n`);
    str += error.stack;
    return str;
};
exports.renderGeneralizedError = renderGeneralizedError;
const renderFailedUpload = (filePath, errorMessage) => {
    const filePathBold = `[${chalk_1.default.bold.dim(filePath)}]`;
    return chalk_1.default.red(`${formatting_1.ICONS.FAILED} Failed upload for ${filePathBold}: ${errorMessage}\n`);
};
exports.renderFailedUpload = renderFailedUpload;
const renderRetriedUpload = (filePath, errorMessage, attempt) => {
    const sourcemapPathBold = `[${chalk_1.default.bold.dim(filePath)}]`;
    return chalk_1.default.yellow(`[attempt ${attempt}] Retrying upload ${sourcemapPathBold}: ${errorMessage}\n`);
};
exports.renderRetriedUpload = renderRetriedUpload;
const renderUpload = (filePath, metadata) => `Uploading debug info for ELF file ${filePath} (arch:${metadata.arch || 'none'} \
GNUBuildID:${metadata.gnu_build_id || 'none'} \
GoBuildID:${metadata.go_build_id || 'none'} \
FileHash:${metadata.file_hash || 'none'})\n`;
exports.renderUpload = renderUpload;
