// @datadog/datadog-ci-plugin-dora/dist/commands/deployment.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginCommand = void 0;
const deployment_1 = require("@datadog/datadog-ci-base/commands/dora/deployment");
const constants_1 = require("@datadog/datadog-ci-base/constants");
const env_1 = require("@datadog/datadog-ci-base/helpers/env");
const fips_1 = require("@datadog/datadog-ci-base/helpers/fips");
const get_git_data_1 = require("@datadog/datadog-ci-base/helpers/git/get-git-data");
const logger_1 = require("@datadog/datadog-ci-base/helpers/logger");
const retry_1 = require("@datadog/datadog-ci-base/helpers/retry");
const chalk_1 = __importDefault(require("chalk"));
const simple_git_1 = __importDefault(require("simple-git"));
const api_1 = require("../api");
const renderer_1 = require("../renderer");
const nonRetriableErrorCodes = [400, 403];
class PluginCommand extends deployment_1.DoraDeploymentCommand {
    constructor() {
        var _a, _b;
        super(...arguments);
        this.config = {
            apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
            fips: (_a = (0, env_1.toBoolean)(process.env[constants_1.FIPS_ENV_VAR])) !== null && _a !== void 0 ? _a : false,
            fipsIgnoreError: (_b = (0, env_1.toBoolean)(process.env[constants_1.FIPS_IGNORE_ERROR_ENV_VAR])) !== null && _b !== void 0 ? _b : false,
        };
        this.logger = new logger_1.Logger((s) => this.context.stdout.write(s), logger_1.LogLevel.INFO);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            (0, fips_1.enableFips)(this.fips || this.config.fips, this.fipsIgnoreError || this.config.fipsIgnoreError);
            this.logger.setLogLevel(this.verbose ? logger_1.LogLevel.DEBUG : logger_1.LogLevel.INFO);
            this.logger.setShouldIncludeTime(this.verbose);
            if (this.serviceParam) {
                this.service = this.serviceParam;
            }
            else {
                this.logger.error('Missing service. It must be provided with --service or the DD_SERVICE env var');
                return 1;
            }
            if (this.startedAt > (this.finishedAt || new Date())) {
                this.logger.error('--started-at cannot be after --finished-at');
                return 1;
            }
            if (this.skipGit) {
                this.gitInfo = undefined;
            }
            else if (this.gitRepoURL && this.gitCommitSHA) {
                this.gitInfo = { repoURL: this.gitRepoURL, commitSHA: this.gitCommitSHA };
            }
            else {
                this.gitInfo = yield this.getGitInfo();
                this.logger.warn((0, renderer_1.renderGitWarning)(this.gitInfo));
            }
            const api = this.getApiHelper();
            yield this.sendDeploymentEvent(api, this.buildDeploymentEvent());
            if (!this.dryRun) {
                this.logger.info((0, renderer_1.renderSuccessfulRequest)(this.service));
            }
        });
    }
    getGitInfo() {
        return __awaiter(this, void 0, void 0, function* () {
            const git = (0, simple_git_1.default)({
                baseDir: process.cwd(),
                binary: 'git',
                // We are invoking at most 5 git commands at the same time.
                maxConcurrentProcesses: 5,
            });
            const [repoURL, commitSHA] = yield Promise.all([(0, get_git_data_1.gitRepositoryURL)(git), (0, get_git_data_1.gitHash)(git)]);
            return { repoURL, commitSHA };
        });
    }
    getApiHelper() {
        if (!this.config.apiKey) {
            this.logger.error(`Neither ${chalk_1.default.red.bold('DATADOG_API_KEY')} nor ${chalk_1.default.red.bold('DD_API_KEY')} is in your environment.`);
            throw new Error('API key is missing');
        }
        return (0, api_1.apiConstructor)(this.config.apiKey);
    }
    buildDeploymentEvent() {
        const deployment = {
            service: this.service,
            startedAt: this.startedAt,
            finishedAt: this.finishedAt || new Date(),
        };
        if (this.env) {
            deployment.env = this.env;
        }
        if (this.gitInfo) {
            deployment.git = this.gitInfo;
        }
        if (this.version) {
            deployment.version = this.version;
        }
        if (this.team) {
            deployment.team = this.team;
        }
        if (this.customTags) {
            deployment.customTags = this.customTags;
        }
        return deployment;
    }
    sendDeploymentEvent(api, deployment) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.dryRun) {
                this.logger.info((0, renderer_1.renderDryRun)(deployment));
                return;
            }
            try {
                this.logger.info((0, renderer_1.renderRequest)(this.service));
                yield (0, retry_1.retryRequest)(() => api.sendDeploymentEvent(deployment), {
                    onRetry: (e, attempt) => {
                        this.logger.warn((0, renderer_1.renderRetriedRequest)(this.service, e, attempt));
                    },
                    retries: 5,
                });
            }
            catch (error) {
                this.logger.error((0, renderer_1.renderFailedRequest)(this.service, error));
                if (error.response) {
                    // If it's an axios error
                    if (!nonRetriableErrorCodes.includes(error.response.status)) {
                        return;
                    }
                }
                throw error;
            }
        });
    }
}
exports.PluginCommand = PluginCommand;
